---
title: "Stromal annotation"
author: "Lea WÃ¶lbert"
date: '2023-01-35'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### Load libraries, source code 

```{r load,  message = FALSE}
library(DropletUtils, quietly = TRUE) 
library(scater, quietly = TRUE) 
library(scuttle, quietly = TRUE) 
library(scran, quietly = TRUE) 
library(tidyverse, quietly = TRUE) 
library(Seurat, quietly = TRUE) 
library(GOfuncR)
set.seed(37)
```

```{r}
color_tables <- "/omics/odcf/analysis/OE0538_projects/DO-0008/metadata/color_tables"
source(file = "../source/colors.R")
source(file = "../source/plotting.R")
source(file = "../source/sce_functions.R")

sce_str <- readRDS(file = "/omics/odcf/analysis/OE0538_projects/DO-0008/data/sce_objects/10_lcls_fractions/sce_str-10")
fraction_curr <- sce_str$Fraction_ID[1]

```

```{r}

sce_str <- factor_reference_cts(sce_str)

col_cts_baccin <- col_cts[match(levels(sce_str$baccin_celltype_scmapclust), 
                                names(col_cts))]
col_cts_dahlin <- col_cts[match(levels(sce_str$dahlin_celltype_scmapclust), 
                                names(col_cts))]
col_cts_dolgalev <- col_cts[match(levels(sce_str$dolgalev_celltype_scmapclust),
                                  names(col_cts))]

col_batch_exp_day <- col_alp[names(col_alp) %in% sce_str$Batch_exp_day]
col_batch_seq <- col_num[names(col_num) %in% sce_str$Batch_sequencing]
col_ann_age <- col_ann[names(col_ann) %in% sce_str$Age_ID]
col_ann_fraction <- col_ann[names(col_ann) %in% sce_str$Fraction_ID]
col_species <- col_spc[names(col_spc) %in% sce_str$Species_ID]

col_clust_l <- col_num[names(col_num) %in% sce_str$cluster_louvain]

```

# Reference cell types 

Cells were annotated with scmap and three reference datasets.

```{r}

identity_plots <- function(sce){
  
  plot_1 <- umap_base(sce, color_by = "baccin_celltype_scmapclust")+ 
    ggtitle(paste0(fraction_curr))+
    scale_color_manual("baccin_celltype_scmapclust", values = col_cts_baccin)
  plot_1l <- umap_legend(sce, color_by = "baccin_celltype_scmapclust")+
    scale_color_manual("baccin_celltype_scmapclust", values = col_cts_baccin)
  legend_1 <- get_legend(plot_1l)

  plot_2 <- umap_base(sce, color_by = "dahlin_celltype_scmapclust")+
    ggtitle(paste0(fraction_curr))+
    scale_color_manual("dahlin_celltype_scmapclust", values = col_cts_dahlin)
  plot_2l <- umap_legend(sce, color_by = "dahlin_celltype_scmapclust")+
    scale_color_manual("dahlin_celltype_scmapclust", values = col_cts_dahlin)
  legend_2 <- get_legend(plot_2l)
  
  plot_3 <- umap_base(sce, color_by = "dolgalev_celltype_scmapclust")+ 
    ggtitle(paste0(fraction_curr))+
    scale_color_manual("dolgalev_celltype_scmapclust",
                       values = col_cts_dolgalev)
  plot_3l <- umap_legend(sce, color_by = "dolgalev_celltype_scmapclust")+
    scale_color_manual("dolgalev_celltype_scmapclust", 
                       values = col_cts_dolgalev)
  legend_3 <- get_legend(plot_3l)
  
  PLOT_1 <- ggarrange(plot_1, legend_1)
  PLOT_2 <- ggarrange(plot_2, legend_2)
  PLOT_3 <- ggarrange(plot_3, legend_3)

  return(list(PLOT_1, PLOT_2, PLOT_3))
}

identity_plots <- identity_plots(sce_str)

```

```{r fig.width=8, fig.height=4}
identity_plots
```

# Clusters

```{r fig.width=8, fig.height=4}

plot_1 <- umap_base(sce_str, color_by = "cluster_louvain")+ 
  scale_color_manual("cluster (louvain)", values = col_clust_l)
plot_1l <- umap_legend(sce_str, color_by = "cluster_louvain")+
  scale_color_manual("cluster (louvain)", values = col_clust_l)
legend_1 <- get_legend(plot_1l)

plot_2 <- pca_base2(sce_str, color_by = "cluster_louvain")+ 
  scale_color_manual("cluster (louvain)", values = col_clust_l)
plot_2l <- pca_legend(sce_str, color_by = "cluster_louvain")+
  scale_color_manual("cluster (louvain)", values = col_clust_l)
legend_2 <- get_legend(plot_2l)

plot_3 <- pca_base3(sce_str, color_by = "cluster_louvain")+ 
  scale_color_manual("cluster (louvain)", values = col_clust_l)
plot_3l <- pca_legend(sce_str, color_by = "cluster_louvain")+
  scale_color_manual("cluster (louvain)", values = col_clust_l)
legend_3 <- get_legend(plot_3l)

PLOT_1 <- ggarrange(plot_1, legend_1)
PLOT_2 <- ggarrange(plot_2, legend_2)
PLOT_3 <- ggarrange(plot_3, legend_3)
  
PLOT_1
PLOT_2
PLOT_3

```

# Literature marker gene expression

Using marker genes from literature to cell types.

### Functions

```{r}

gene_exp_umaps <- function(gene){
 
  plot_1 <- umap_gene(sce_str, gene)+
  scale_color_continuous(name = gene)
  return(plot_1)

}

gene_exp_bees <- function(gene, ggdf){
  
  plot <- ggplot(ggdf, aes(x = cluster_louvain,
                           y = ggdf[,colnames(ggdf) == gene], 
                           color = cluster_louvain))+
    ggbeeswarm::geom_quasirandom(size = 0.2)+
    theme_classic()+
    scale_color_manual(values = col_clust_l)+
    geom_boxplot(color = "black", alpha = 0)+
    ylab(paste0(gene, " expression"))
  
  return(plot)
}

```


## MSCs

```{r}

msc_marker_genes <- list(
 
)

```

For sources see Marker_list.xlsx

```{r}

plotlist_msc <- lapply(msc_marker_genes, gene_exp_umaps)
plotlist_msc

```

```{r}

ggdf <- as.data.frame(colData(sce_str))
logc <- assays(sce_str)[["logcounts"]]

for(i in msc_marker_genes){
  ggdf$new <- logc[which(rownames(logc) == i)]
  colnames(ggdf)[colnames(ggdf) == "new"] <- i
}

plotlist_msc2 <- lapply(msc_marker_genes, ggdf = ggdf, gene_exp_bees)
plotlist_msc2

```

# Find marker genes

Use seurats wilcoxon test-based method
(see https://www.biorxiv.org/content/10.1101/2022.05.09.490241v1)

```{r}

seurat <- as.Seurat(sce_str, counts = "counts", data = "logcounts",) 
Idents(seurat) <- sce_str$cluster_louvain
tail(match(names(Idents(seurat)), rownames(colData(sce_str))))

hvgs <- modelGeneVar(sce_str)
hvgs <- getTopHVGs(hvgs, n=2000)

clustlist <- as.list(1:length(unique(sce_str$cluster_louvain)))
cluster_markers <- lapply(clustlist, function(x){
  markers <- FindMarkers(seurat, test.use = "wilcox", ident.1 = x, 
                         features = hvgs)
  markers <- markers[order(abs(markers$avg_log2FC), decreasing=TRUE),]
  return(markers)
})

cluster_markers

```

## Visualise

```{r}

c <- 1
marker_plotlist <- lapply(cluster_markers, function(x){
  
  if(length(which(x$avg_log2FC >0))>6){
    markers_pos <- as.list(rownames(x[x$avg_log2FC >0,])[1:6])
  }else{
    markers_pos <- as.list(rownames(x[x$avg_log2FC >0,])) 
  }
  if(length(which(x$avg_log2FC <0))>6){
    markers_neg <- as.list(rownames(x[x$avg_log2FC <0,])[1:6])
  }else{
    markers_neg <- as.list(rownames(x[x$avg_log2FC <0,]))
  }

  umap_list_pos <- lapply(markers_pos, function(y){
    plot <- gene_exp_umaps(y)
    plot <- plot + ggtitle(paste0("cluster ", c, " positive marker"))
    return(plot)
  })
  umap_list_neg <- lapply(markers_neg, function(y){
    plot <- gene_exp_umaps(y)
    plot <- plot + ggtitle(paste0("cluster ", c, " negative marker"))
    return(plot)
  })
  
  ggdf <- as.data.frame(colData(sce_str))
  logc <- assays(sce_str)[["logcounts"]]

  for(i in markers_pos){
    ggdf$new <- logc[which(rownames(logc) == i)]
    colnames(ggdf)[colnames(ggdf) == "new"] <- i
  }
  bee_list_pos <- lapply(markers_pos, function(y){
   plot <- gene_exp_bees(y, ggdf = ggdf) 
   plot <- plot + ggtitle(paste0("cluster ", c, " positive marker"))
   return(plot)
  })
  
  for(i in markers_neg){
    ggdf$new <- logc[which(rownames(logc) == i)]
    colnames(ggdf)[colnames(ggdf) == "new"] <- i
  }
  bee_list_neg <- lapply(markers_neg, function(y){
   plot <- gene_exp_bees(y, ggdf = ggdf) 
   plot <- plot + ggtitle(paste0("cluster ", c, " negative marker"))
   return(plot)
  })
  
  c <- c + 1
  return(list(umap_list_pos, bee_list_pos, umap_list_neg, bee_list_neg))
    
})

marker_plotlist

```

## GO analysis

```{r}

get_go_results <- function(markers){
  
  go_input_df <- data.frame(
    gene_ids = rownames(markers),
    gene_scores =  markers$avg_log2FC
  )
                       
  go_result <- go_enrich(go_input_df, organismDb='Mus.musculus',
                         test='wilcoxon', n_randsets=100)
  return(go_result)  
}

go_results <- lapply(cluster_markers, get_go_results)

go_results_short <- lapply(go_results, function(x){
  results <- x$results[x$results$ontology == "biological_process",]
  return(results)
})

```
