---
title: "HSC annotation"
author: "Lea WÃ¶lbert"
date: '2023-01-10'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### Load libraries, source code 

```{r load,  message = FALSE}
library(DropletUtils, quietly = TRUE) 
library(scater, quietly = TRUE) 
library(scuttle, quietly = TRUE) 
library(scran, quietly = TRUE) 
library(tidyverse, quietly = TRUE) 
library(Seurat, quietly = TRUE) 
library(GOfuncR)
set.seed(37)
```

```{r}
color_tables <- "/omics/odcf/analysis/OE0538_projects/DO-0008/metadata/color_tables"
source(file = "../source/colors.R")
source(file = "../source/plotting.R")
source(file = "../source/sce_functions.R")

sce_hsc <- readRDS(file = "/omics/odcf/analysis/OE0538_projects/DO-0008/data/sce_objects/10_lcls_fractions/sce_hsc-10")
fraction_curr <- sce_hsc$Fraction_ID[1]

```

```{r}

sce_hsc <- factor_reference_cts(sce_hsc)

col_cts_baccin <- col_cts[match(levels(sce_hsc$baccin_celltype_scmapclust), 
                                names(col_cts))]
col_cts_dahlin <- col_cts[match(levels(sce_hsc$dahlin_celltype_scmapclust), 
                                names(col_cts))]
col_cts_dolgalev <- col_cts[match(levels(sce_hsc$dolgalev_celltype_scmapclust),
                                  names(col_cts))]

col_batch_exp_day <- col_alp[names(col_alp) %in% sce_hsc$Batch_exp_day]
col_batch_seq <- col_num[names(col_num) %in% sce_hsc$Batch_sequencing]
col_ann_age <- col_ann[names(col_ann) %in% sce_hsc$Age_ID]
col_ann_fraction <- col_ann[names(col_ann) %in% sce_hsc$Fraction_ID]
col_species <- col_spc[names(col_spc) %in% sce_hsc$Species_ID]

col_clust_l <- col_num[names(col_num) %in% sce_hsc$cluster_louvain]

```

# Reference cell types 

Cells were annotated with scmap and three reference datasets.

```{r}

identity_plots <- function(sce){
  
  plot_1 <- umap_base(sce, color_by = "baccin_celltype_scmapclust")+ 
    ggtitle(paste0(fraction_curr))+
    scale_color_manual("baccin_celltype_scmapclust", values = col_cts_baccin)
  plot_1l <- umap_legend(sce, color_by = "baccin_celltype_scmapclust")+
    scale_color_manual("baccin_celltype_scmapclust", values = col_cts_baccin)
  legend_1 <- get_legend(plot_1l)

  plot_2 <- umap_base(sce, color_by = "dahlin_celltype_scmapclust")+
    ggtitle(paste0(fraction_curr))+
    scale_color_manual("dahlin_celltype_scmapclust", values = col_cts_dahlin)
  plot_2l <- umap_legend(sce, color_by = "dahlin_celltype_scmapclust")+
    scale_color_manual("dahlin_celltype_scmapclust", values = col_cts_dahlin)
  legend_2 <- get_legend(plot_2l)
  
  plot_3 <- umap_base(sce, color_by = "dolgalev_celltype_scmapclust")+ 
    ggtitle(paste0(fraction_curr))+
    scale_color_manual("dolgalev_celltype_scmapclust",
                       values = col_cts_dolgalev)
  plot_3l <- umap_legend(sce, color_by = "dolgalev_celltype_scmapclust")+
    scale_color_manual("dolgalev_celltype_scmapclust", 
                       values = col_cts_dolgalev)
  legend_3 <- get_legend(plot_3l)
  
  PLOT_1 <- ggarrange(plot_1, legend_1)
  PLOT_2 <- ggarrange(plot_2, legend_2)
  PLOT_3 <- ggarrange(plot_3, legend_3)

  return(list(plot_1, PLOT_1, PLOT_2, PLOT_3))
}

identity_plots <- identity_plots(sce_hsc)

```

```{r fig.width=8, fig.height=4}
identity_plots
```

# Clusters

```{r fig.width=8, fig.height=4}

plot_1 <- umap_base(sce_hsc, color_by = "cluster_louvain")+ 
  scale_color_manual("cluster (louvain)", values = col_clust_l)
plot_1l <- umap_legend(sce_hsc, color_by = "cluster_louvain")+
  scale_color_manual("cluster (louvain)", values = col_clust_l)
legend_1 <- get_legend(plot_1l)

plot_2 <- pca_base2(sce_hsc, color_by = "cluster_louvain")+ 
  scale_color_manual("cluster (louvain)", values = col_clust_l)
plot_2l <- pca_legend(sce_hsc, color_by = "cluster_louvain")+
  scale_color_manual("cluster (louvain)", values = col_clust_l)
legend_2 <- get_legend(plot_2l)

plot_3 <- pca_base3(sce_hsc, color_by = "cluster_louvain")+ 
  scale_color_manual("cluster (louvain)", values = col_clust_l)
plot_3l <- pca_legend(sce_hsc, color_by = "cluster_louvain")+
  scale_color_manual("cluster (louvain)", values = col_clust_l)
legend_3 <- get_legend(plot_3l)

PLOT_1 <- ggarrange(plot_1, legend_1)
PLOT_2 <- ggarrange(plot_2, legend_2)
PLOT_3 <- ggarrange(plot_3, legend_3)
  
PLOT_1
PLOT_2
PLOT_3

```

# Literature marker gene expression

Using marker genes from literature to cell types.

### Functions

```{r}

gene_exp_umaps <- function(gene){
 
  plot_1 <- umap_gene(sce_hsc, gene)
  return(plot_1)

}

gene_exp_bees <- function(gene, ggdf){
  
  plot <- ggplot(ggdf, aes(x = cluster_louvain,
                           y = ggdf[,colnames(ggdf) == gene], 
                           color = cluster_louvain))+
    ggbeeswarm::geom_quasirandom(size = 0.2)+
    theme_classic()+
    scale_color_manual(values = col_clust_l)+
    geom_boxplot(color = "black", alpha = 0)+
    ylab(paste0(gene, " expression"))
  
  return(plot)
}

```

# HSCs

### primitive HSC

Not necessarily quiescent or LT.

```{r}

hsc_marker_genes <- list(
 "2410080I02Rik",
 "Aldoc", 
 "Cd48",
 "Cd38",
 "Cd34",
 "Cd82",
 "Cish",
 "Ctnnal1",
 "Cyld",
 "Cyp26b1",
 "Dlk1",
 "Egr1",
 "Adgrl4",
 "Esam",
 "F11r",
 "Fgd5",
 "Fth1",
 "Grb10",
 "Gstm1",
 "Gstm5",
 "H2afy",
 "Hlf",
 "Hmga2",
 "Hoxb2",
 "Hoxb3",
 "Hoxb4",
 "Hoxb5",
 "Hoxb8",
 "Hsp90aa1",
 "Ifitm1",
 "Igf1",
 "Igf2bp2",
 "Igf2r",
 "Inhba",
 "Irgm2",
 "Klf4",
 "Ldhb",
 "Ldhd",
 "Ly6a",
 "Malat1",
 "Mecom",
 "Mecp2",
 "Meg3",
 "Meis1",
 "Mllt3",
 "Mmrn1",
 "Mpl",
 "Mycn",
 "Neo1",
 "Pde1b",
 "Peg3",
 "Procr",
 "Pygm",
 "Robo4",
 "Rarb",
 "Rrm1",
 "Rrm2",
 "Serpinb6a",
 "Slamf1",
 "Stat1",
 "Tek",
 "Tgm2",
 "Tie1",
 "Tinagl1",
 "Trpc6",
 "Txnip"
)

```

```{r}

plotlist_hsc <- lapply(hsc_marker_genes, gene_exp_umaps)
plotlist_hsc

```

```{r}

ggdf <- as.data.frame(colData(sce_hsc))
logc <- assays(sce_hsc)[["logcounts"]]

for(i in hsc_marker_genes){
  ggdf$new <- logc[which(rownames(logc) == i),]
  colnames(ggdf)[colnames(ggdf) == "new"] <- i
}

plot_list_hsc2 <- lapply(hsc_marker_genes, ggdf = ggdf, gene_exp_bees)
plot_list_hsc2

```

```{r}

gene_exp_umaps_fg <- function(gene){
 
  plot_1 <- umap_gene(sce_hsc, gene)+
    facet_grid(cols = vars(sce_hsc$Species_ID))
  return(plot_1)

}

plotlist <- lapply(hsc_marker_genes, gene_exp_umaps_fg)
plotlist

ggdf_temp <- ggdf
ggdf_temp$Species_ID[ggdf_temp$Species_ID == "mcar"] <- "mus caroli"
ggdf_temp$Species_ID[ggdf_temp$Species_ID == "mcas"] <- "mus musculus castaneus"
ggdf_temp$Species_ID[ggdf_temp$Species_ID == "mmus"] <- "C57BL6/J"
ggdf_temp$Species_ID[ggdf_temp$Species_ID == "mspr"] <- "mus spretus"
ggdf_temp$Species_ID <- factor(ggdf_temp$Species_ID, levels = c(
  "C57BL6/J", "mus musculus castaneus",  "mus spretus", "mus caroli")
)

ggplot(ggdf_temp, 
       aes(x = Species_ID, 
           y = logc[rownames(logc) == "Cd48",]))+
  ggbeeswarm::geom_quasirandom(size = 0.2, color = "grey65")+
  theme_classic()+
  geom_boxplot(color = "black", alpha = 0)+
  ylab("Cd48 renormalized logcounts")+
  theme(axis.text.x = element_text(angle = 90),
        axis.title.x = element_blank())

ggplot(ggdf_temp, 
       aes(x = Species_ID,
           y = logc[rownames(logc) == "Cd34",]))+
  ggbeeswarm::geom_quasirandom(size = 0.2, color = "grey65")+
  theme_classic()+
  geom_boxplot(color = "black", alpha = 0)+
  ylab("Cd34 renormalized logcounts")+
  theme(axis.text.x = element_text(angle = 90),
        axis.title.x = element_blank())
  

```

## Ery MK lineage

### MEPs

```{r}

ermk_marker_genes <- list(
  "Col5a1", # MEPs
  "Gata1", # MEPs
  "Gata2", # early MEPs,
  "Gypa", # MEPs
  "Hba-a1", # MEPs
  "Prss50", # MEPs
  "Smim1", # MEPs
  "Snca", # MEPs,
  "Zfpm1" # MEPs
)

```

```{r}

plotlist_ermk <- lapply(ermk_marker_genes, gene_exp_umaps)
plotlist_ermk

```

```{r}

ggdf <- as.data.frame(colData(sce_hsc))
logc <- assays(sce_hsc)[["logcounts"]]

for(i in ermk_marker_genes){
  ggdf$new <- logc[which(rownames(logc) == i),]
  colnames(ggdf)[colnames(ggdf) == "new"] <- i
}

plotlist_ermk2 <- lapply(ermk_marker_genes, ggdf = ggdf, gene_exp_bees)
plotlist_ermk2

```

### Ery priming

```{r}

ery_marker_genes <- list(
  "Apoe", # ery fate
  "Car1", # ery fate
  "Car2", # ery fate
  "Ccna2", # imm late ery progs
  "Ccnd2", # earl imm ery progs
  "Ccne1", # imm late ery progs
  "Cd34", # early ery progs, many other progs
  "Cd48", # eary progs, many other progs
  "Cdk6", # early, imm ery prog
  "Cited4", # ery progs
  "Epor", # ery progs
  "Gfi1b", # ery progs
  "Gypc", # ery progs
  "Hba-a1", # ery late
  "Il17ra", # ery prog
  "Itga2b", # early ery prog
  "Klf1", # ery fate
  "Ldb1", # early to imm ery prog
  "Mcm3", # imm to late ery prog
  "Mcm4", # imm to late ery prog
  "Mcm5", # imm to late ery prog
  "Mcm6", # imm to late ery prog
  "Mcm7", # imm to late ery prog
  "Mt2", # early ery
  "Ryk", # ery prog
  "Tal1", # imm to late ery prog
  "Tfr2", # ery prog
  "Tfrc" # late ery
)

```

```{r}

plotlist_ery <- lapply(ery_marker_genes, gene_exp_umaps)
plotlist_ery

```

```{r}

ggdf <- as.data.frame(colData(sce_hsc))
logc <- assays(sce_hsc)[["logcounts"]]

for(i in ery_marker_genes){
  ggdf$new <- logc[which(rownames(logc) == i),]
  colnames(ggdf)[colnames(ggdf) == "new"] <- i
}

plotlist_ery2 <- lapply(ery_marker_genes, ggdf = ggdf, gene_exp_bees)
plotlist_ery2

```

### MK priming

```{r}

mk_marker_genes <- list(
  "Cd9", # mk prog
  "Clu", # platelet lineage
  "Esam", # platelet lineage
  "Fhl1", # platelet lineage
  "Fli1", # mk fate
  "Gp1bb", # mk progs
  "Itga2b", # mk progs
  "Itgb3", # platelet diff
  "Mef2c", # mk fate
  "Pbx1", # mk fate
  "Pf4", # mk lineage
  "Cavin2", # platelet lineage
  "Selp", # platelet lineage
  "Serpinb1a", # platelet lineage
  "Slamf1", # mk progs
  "Tgm2", # platelet lin
  "Vwf" # platelet lineage
)

```

```{r}

plotlist_mk <- lapply(mk_marker_genes, gene_exp_umaps)
plotlist_mk

```

```{r}

ggdf <- as.data.frame(colData(sce_hsc))
logc <- assays(sce_hsc)[["logcounts"]]

for(i in mk_marker_genes){
  ggdf$new <- logc[which(rownames(logc) == i),]
  colnames(ggdf)[colnames(ggdf) == "new"] <- i
}

plotlist_mk2 <- lapply(mk_marker_genes, ggdf = ggdf, gene_exp_bees)
plotlist_mk2

```

## Lymphoid lineage

```{r}

lym_marker_genes <- list(
  "Ccl5", # lymhpoid, DC prog
  "Cd37", # lymph priming
  "Cd48", # lymph, myeloid, neutro, eo priming
  "Cd74", # BC, DC fate
  "Cd79a", # lymph progenitors
  "Cd79b", # lymph progenitors
  "Dntt", # lymph prog, BC fate
  "Fcrla", # BC fate
  "Flt3", # early progs, lymph progenitors, MPPs,
  "Gata2", # LMPPs, GSC tp MPP transition
  "Gata3", # lympg priming
  "Gzmb", # lymphoid
  "Il7r", # lypmh prog (TC)
  "Ly6d", # lymph progs
  "Nkg7", # innate lymph fate
  "Rag2", # lymph prog
  "Satb1", # lymph prog
  "Sox4", # LSK primed lymph commitment
  "Vpreb1", # BC fate
  "Vpreb3" # BC fate
    
)

```

```{r}

plotlist_lym <- lapply(lym_marker_genes, gene_exp_umaps)
plotlist_lym

```

```{r}

ggdf <- as.data.frame(colData(sce_hsc))
logc <- assays(sce_hsc)[["logcounts"]]

for(i in lym_marker_genes){
  ggdf$new <- logc[which(rownames(logc) == i),]
  colnames(ggdf)[colnames(ggdf) == "new"] <- i
}

plotlist_lym2 <- lapply(lym_marker_genes, ggdf = ggdf, gene_exp_bees)
plotlist_lym2

```

## Myeloid lineage

### Granulocyte-myeloid priming

```{r}

my_marker_genes <- list(
  "Cd52", # myeloid priming, multi-lineage HSCs
  "Cebpa", # myeloid, granulocyte, progenitors, monocyte, neutrophil fate
  "Cebpb", # granulocyte fate
  "Cebpe", # mature granulocytes,  myeloid lineage, myeloid biased LT-HSCs, eos, neutros
  "Ctsg", # myeloid LSKs, monocytes
  "Elane", # granulocytes, GMP
  "Adgrl4", # myeloid priming, progenitors
  "Etv6", # myeloid priming
  "Fcer1g", # myeloid priming
  "Flt3", # early granulocyte progenitors, lymphoid, multilineage, primed HSCs, MPP4, early mono progs
  "Gstm1", # granulocyte fate, early neutros
  "Hdc", # myeloid priming, neutros, mast
  "Irf8", # early granulocyte progenitors, dendritic, monocytic fate
  "Meis2",  # myeloid lineage, myeloid biased LT-HSCs
  "Mpo", # GMP, myelpod fate, progenitors, early and medium neutro progs
  "Nkg7", # myeloid priming, innate lymhpocyte, multilineage hscs
  "Prtn3", # GMP, monocyte progs
  "Spi1", # myeloid lineage, myeloid granulocyte progenitors, early monos
  "Stat3" # granulocyte fate
)

```

```{r}

plotlist_my <- lapply(my_marker_genes, gene_exp_umaps)
plotlist_my

```

```{r}

ggdf <- as.data.frame(colData(sce_hsc))
logc <- assays(sce_hsc)[["logcounts"]]

for(i in my_marker_genes){
  ggdf$new <- logc[which(rownames(logc) == i),]
  colnames(ggdf)[colnames(ggdf) == "new"] <- i
}

plotlist_my2 <- lapply(my_marker_genes, ggdf = ggdf, gene_exp_bees)
plotlist_my2

```

### Monos

```{r}

mono_marker_genes <- list(
  "Ass1", # monocyte program
  "Cebpa", # monocyte fate
  "Csf1r", # monocyte fate, early mono progs
  "Ctsg", # mono progs, mono fate, myeloid prog, eo neutro fate
  "Ctss", # mono program
  "Elane", # mono program, neutro, granulocyte, mast prog
  "Emb", # mono program, multi-lineage HSCs
  "F13a1", # mono fate
  "Flt3", # early mono progs, early progs, multi-lineage HSCs
  "Irf8", # mono and DC progs
  "Ly6c2", # mono fate, neutrophil fate
  "Ly6e", # mono progs
  "Ly86", # mono progs
  "Lyz1", # mono fate, neutrophil fate
  "Lyz2", # mono fate, neutrophil fate
  "Ms4a6c", # mono program
  "Prdx4", # mono program
  "Prtn3", # mono progs, GMPs
  "Rassf4", # mono program
  "S100a4", # late mono program
  "Slpi" # mono program
)

```

```{r}

plotlist_mono <- lapply(mono_marker_genes, gene_exp_umaps)
plotlist_mono

```

```{r}

ggdf <- as.data.frame(colData(sce_hsc))
logc <- assays(sce_hsc)[["logcounts"]]

for(i in mono_marker_genes){
  ggdf$new <- logc[which(rownames(logc) == i),]
  colnames(ggdf)[colnames(ggdf) == "new"] <- i
}

plotlist_mono2 <- lapply(mono_marker_genes, ggdf = ggdf, gene_exp_bees)
plotlist_mono2

```

### Neutros 

```{r}

neutro_marker_genes <- list(
  "Camp", # neutro prog
  "Ccl6", # neutro fate
  "Cd177", # late neutro program
  "Cd34", # neutro progs, multilineage HSCs, HSPCs, MPPs
  "Cd48", # neutro progs, lymphoid prog, HSPCs
  "Cd63", # neutro fate
  "Cebpa", # neutro prog, monos, granus
  "Cebpe", # neutro prog, myeloid, LT-HSCs, eos,
  "Chil1", # late neutro prog
  "Chil3", # late neutro prog
  "Elane", # neutro progs, monos, granus, mast,
  "Fcer1a", # late neutro prog, baso/mast
  "Fcnb", # neutro fate
  "Gfi1", # neutro prog, eos
  "Gstm1", # neutro fate, early neutro program, granus
  "Hdc", # neutro progs, mast, myeloid
  "Ifitm6", # late neutro program
  "Il1b", # neutro diff
  "Itgam", # neutro diff, MPP1
  "Itgb2l", # late neutro program
  "Lcn2", # late/im neutro program
  "Lrg1", # late neutro program
  "Ltf", # late neutrophil program
  "Ly6c2", # neutro fate, monos
  "Lyz1", # neutro fate, monos
  "Lyz2", # neutro fate, monos
  "Mmp8", # neutro diff
  "Mpo", # early neutrophil, myeloid, granus
  "Ngp", # late and im neutro program
  "Orm1", # late neutro program
  "Pglyrp1", # late neutro program
  "S100a8", # late neutro program
  "S100a9", # late neutro program
  "Syne1" # late neutro program
  
)

```

```{r}

plotlist_neutro <- lapply(neutro_marker_genes, gene_exp_umaps)
plotlist_neutro

```

```{r}

ggdf <- as.data.frame(colData(sce_hsc))
logc <- assays(sce_hsc)[["logcounts"]]

for(i in neutro_marker_genes){
  ggdf$new <- logc[which(rownames(logc) == i),]
  colnames(ggdf)[colnames(ggdf) == "new"] <- i
}

plotlist_neutro2 <- lapply(neutro_marker_genes, ggdf = ggdf, gene_exp_bees)
plotlist_neutro2

```

### DCs

```{r}

dc_marker_genes <- list(
  "Ccl5", # DC, lymphoid
  "Ccr7", # DC progs
  "Cd7", # DC progs
  "Cd74", # DCs, BCs fate
  "Id2", # DC prog
  "Irf8", # DC, monos, granus
  "Itgb7", # DC fate
  "Naaa", # DC progs
  "Siglech" # DC fate
)

```

```{r}

plotlist_dc <- lapply(dc_marker_genes, gene_exp_umaps)
plotlist_dc

```

```{r}

ggdf <- as.data.frame(colData(sce_hsc))
logc <- assays(sce_hsc)[["logcounts"]]

for(i in dc_marker_genes){
  ggdf$new <- logc[which(rownames(logc) == i),]
  colnames(ggdf)[colnames(ggdf) == "new"] <- i
}

plotlist_dc2 <- lapply(dc_marker_genes, ggdf = ggdf, gene_exp_bees)
plotlist_dc2

```

### Mast cells

```{r}

mast_marker_genes <- list(
  "Cma1", # mature mast/baso cells
  "Cpa3", # baso/mast progenitor
  "Elane", # mast progenitors, neutro progs, granulocytes, monocyte fate
  "Fcer1a", # baso mast progenitors, myeloid, old neutrophils
  "Gzmb", # mast cell, lymphocyte
  "Hdc", # mast progenitors, myeloid priming, myeloid primed HSCs
  "Mcpt8", # mature mast/basos
  "Ms4a2", #baso/mast progenitors,
  "Ndst2" # mature mast/basos
)

```

```{r}

plotlist_mast <- lapply(mast_marker_genes, gene_exp_umaps)
plotlist_mast

```

```{r}

ggdf <- as.data.frame(colData(sce_hsc))
logc <- assays(sce_hsc)[["logcounts"]]

for(i in mast_marker_genes){
  ggdf$new <- logc[which(rownames(logc) == i),]
  colnames(ggdf)[colnames(ggdf) == "new"] <- i
}

plotlist_mast2 <- lapply(mast_marker_genes, ggdf = ggdf, gene_exp_bees)
plotlist_mast2

```

### Basos

```{r}

baso_marker_genes <- list(
  "Cma1", # mature mast/baso cells
  "Cpa3", # baso/mast progenitor
  "Ctsg", # baso/eo fate, myeloid, monocyte
  "Fcer1a", # baso mast progenitors, myeloid, old neutrophils
  "Ifitm1", # basophil fate, HSCs
  "Lmo4", # baso progs
  "Mcpt8", # mature mast/basos
  "Ms4a2", #baso/mast progenitors,
  "Ndst2", # mature mast/basos
  "Prss34" # baso/baso prog
)

```

```{r}

plotlist_baso <- lapply(baso_marker_genes, gene_exp_umaps)
plotlist_baso

```

```{r}

ggdf <- as.data.frame(colData(sce_hsc))
logc <- assays(sce_hsc)[["logcounts"]]

for(i in baso_marker_genes){
  ggdf$new <- logc[which(rownames(logc) == i),]
  colnames(ggdf)[colnames(ggdf) == "new"] <- i
}

plotlist_baso2 <- lapply(baso_marker_genes, ggdf = ggdf, gene_exp_bees)
plotlist_baso2

```

### Eosinophils

```{r}

eo_marker_genes <- list(
  "Cebpe", #eo, myeloid, neutrophil
  "Ctsg", # baso/eo fate, myeloid, monocyte
  "Gfi1", # eo or neutrophil progenitors
  "Prg2", # basophil fate, HSCs
  "Prg3"
)

```

```{r}

plotlist_eo <- lapply(eo_marker_genes, gene_exp_umaps)
plotlist_eo

```

```{r}

ggdf <- as.data.frame(colData(sce_hsc))
logc <- assays(sce_hsc)[["logcounts"]]

for(i in eo_marker_genes){
  ggdf$new <- logc[which(rownames(logc) == i),]
  colnames(ggdf)[colnames(ggdf) == "new"] <- i
}

plotlist_eo2 <- lapply(eo_marker_genes, ggdf = ggdf, gene_exp_bees)
plotlist_eo2

```

# Find marker genes

Use seurats wilcoxon test-based method
(see https://www.biorxiv.org/content/10.1101/2022.05.09.490241v1)

```{r}

cluster_markers <- readRDS(file = "/omics/odcf/analysis/OE0538_projects/DO-0008/data/sce_objects/11_annotation/results_markers/markers_hsc")

```

## Visualise

```{r}

marker_plotlist <- lapply(cluster_markers, function(x){
  
  c <- x$which_cluster[1]
  if(length(which(x$avg_log2FC >0))>10){
    markers_pos <- as.list(rownames(x[x$avg_log2FC >0,])[1:10])
  }else{
    markers_pos <- as.list(rownames(x[x$avg_log2FC >0,])) 
  }
  if(length(which(x$avg_log2FC <0))>6){
    markers_neg <- as.list(rownames(x[x$avg_log2FC <0,])[1:6])
  }else{
    markers_neg <- as.list(rownames(x[x$avg_log2FC <0,]))
  }

  umap_list_pos <- lapply(markers_pos, function(y){
    plot <- gene_exp_umaps(y)
    plot <- plot + ggtitle(paste0("cluster ", c, " positive marker"))
    return(plot)
  })
  umap_list_neg <- lapply(markers_neg, function(y){
    plot <- gene_exp_umaps(y)
    plot <- plot + ggtitle(paste0("cluster ", c, " negative marker"))
    return(plot)
  })
  
  ggdf <- as.data.frame(colData(sce_hsc))
  logc <- assays(sce_hsc)[["logcounts"]]

  for(i in markers_pos){
    ggdf$new <- logc[which(rownames(logc) == i),]
    colnames(ggdf)[colnames(ggdf) == "new"] <- i
  }
  bee_list_pos <- lapply(markers_pos, function(y){
   plot <- gene_exp_bees(y, ggdf = ggdf) 
   plot <- plot + ggtitle(paste0("cluster ", c, " positive marker"))
   return(plot)
  })
  
  for(i in markers_neg){
    ggdf$new <- logc[which(rownames(logc) == i),]
    colnames(ggdf)[colnames(ggdf) == "new"] <- i
  }
  bee_list_neg <- lapply(markers_neg, function(y){
   plot <- gene_exp_bees(y, ggdf = ggdf) 
   plot <- plot + ggtitle(paste0("cluster ", c, " negative marker"))
   return(plot)
  })
  
  return(list(umap_list_pos, bee_list_pos, umap_list_neg, bee_list_neg))

})

marker_plotlist

```

# NMF program analysis

SCE containing NMF data is slightly subset since Adrien has an older version of
the SCE objects with slightly different retained cells

```{r}
sce_hsc <- readRDS(file = "/omics/odcf/analysis/OE0538_projects/DO-0008/data/sce_objects/11_annotation/sce_nmf/sce_nmf_hsc")
```

Redefine colors since some cells are missing

```{r}

sce_hsc <- factor_reference_cts(sce_hsc)

col_cts_baccin <- col_cts[match(levels(sce_hsc$baccin_celltype_scmapclust), 
                                names(col_cts))]
col_cts_dahlin <- col_cts[match(levels(sce_hsc$dahlin_celltype_scmapclust), 
                                names(col_cts))]
col_cts_dolgalev <- col_cts[match(levels(sce_hsc$dolgalev_celltype_scmapclust),
                                  names(col_cts))]

col_batch_exp_day <- col_alp[names(col_alp) %in% sce_hsc$Batch_exp_day]
col_batch_seq <- col_num[names(col_num) %in% sce_hsc$Batch_sequencing]
col_ann_age <- col_ann[names(col_ann) %in% sce_hsc$Age_ID]
col_ann_fraction <- col_ann[names(col_ann) %in% sce_hsc$Fraction_ID]
col_species <- col_spc[names(col_spc) %in% sce_hsc$Species_ID]

col_clust_l <- col_num[names(col_num) %in% sce_hsc$cluster_louvain]

```

## Usage UMAPs

```{r}

species_list <- as.list(names(metadata(sce_hsc)$usage_list))

plot_list_programs <- lapply(species_list, function(species){
  
  program_list <- as.list(colnames(metadata(sce_hsc)$usage_list[[species]]))
  sce_temp <- sce_hsc[,sce_hsc$Species_ID == species]
  
  plotlist_program <- lapply(program_list, function(program){
  
    plot1 <- umap_program(sce = sce_temp,
                          species = species, 
                          program = program)
  
    plot2 <- ggplot(data.frame(reducedDims(sce_temp)[["UMAP"]]),
                 aes(x = X1, y = X2, 
                     color = as.character(metadata(sce_temp)$cluster_list[[species]][,program])))+
    scale_color_manual(name = paste0(program, " cluster"), values = c("1" = "grey60",
                                                     "2" = "dodgerblue3"))+
    geom_point(size = 0.01)+
    theme_classic()+
    theme(axis.text = element_blank())+
    ylab("UMAP 2")+
    xlab("UMAP 1")
    
    plot_title1 <- plot1+ggtitle(species)
    plot_title2 <- plot2+ggtitle(species)
    return(list(plot_title1, plot_title2))
  })
  return(plotlist_program)
})

plot_list_programs

```

## Usage histograms

```{r}

species_list <- as.list(names(metadata(sce_hsc)$usage_list))
histo_list_programs <- lapply(species_list, function(species){
  
  program_list <- as.list(colnames(metadata(sce_hsc)$usage_list[[species]]))
  clusters <- metadata(sce_hsc)$cluster_list[[species]]
  sce_temp <- sce_hsc[,sce_hsc$Species_ID == species]
  
  plotlist_program <- lapply(program_list, function(program){
  
    plot1 <- ggplot(metadata(sce_temp)$usage_list[[species]],
                 aes(y = metadata(sce_temp)$usage_list[[species]][,program], 
                     x = sce_temp$cluster_louvain,
                     color = sce_temp$cluster_louvain))+
      geom_jitter(size = 0.2)+
      geom_violin(alpha = 0, color ="black", scale = "width")+
      theme_classic()+
      scale_color_manual("cluster (louvain)", values = col_clust_l)+
      ylab(paste0(program, " usage"))+
      xlab("cluster (louvain)")
    
    plot2 <- ggplot(metadata(sce_temp)$usage_list[[species]],
                 aes(y = metadata(sce_temp)$usage_list[[species]][,program], 
                     x = sce_temp$Species_ID))+
      geom_jitter(size = 0.2, color = "grey60")+
      geom_violin(alpha = 0, color ="black", scale = "width")+
      theme_classic()+
      scale_color_manual("cluster (louvain)", values = col_clust_l)+
      ylab(paste0(program, " usage"))+
      theme(axis.title.x = element_blank(),
            axis.text.x = element_blank())   
    
    plot3 <- ggplot(metadata(sce_temp)$usage_list[[species]],
                 aes(y = metadata(sce_temp)$usage_list[[species]][,program], 
                     x = sce_temp$cluster_louvain,
                     color = as.character(clusters[,program])))+
      geom_jitter(size = 0.2)+
      geom_violin(alpha = 0, color ="black", scale = "width")+
      theme_classic()+
      scale_color_manual("cluster program", values = c("1" = "grey60",
                                                       "2" = "dodgerblue3"))+
      ylab(paste0(program, " usage"))+
      xlab("cluster (louvain)")
    
    plot4 <- ggplot(metadata(sce_temp)$usage_list[[species]],
                 aes(y = metadata(sce_temp)$usage_list[[species]][,program], 
                     x = sce_temp$Species_ID,
                     color = as.character(clusters[,program])))+
      geom_jitter(size = 0.2)+
      geom_violin(alpha = 0, color ="black", scale = "width")+
      theme_classic()+
      scale_color_manual("cluster program", values = c("1" = "grey60",
                                                       "2" = "dodgerblue3"))+
      ylab(paste0(program, " usage"))+
      theme(axis.title.x = element_blank(),
            axis.text.x = element_blank()) 
  
    plot_title1 <- plot1+ggtitle(species)
    plot_title2 <- plot2+ggtitle(species)
    plot_title3 <- plot3+ggtitle(species)
    plot_title4 <- plot4+ggtitle(species)
    return(list(plot_title1, plot_title3, plot_title2, plot_title4))
  })
  return(plotlist_program)
})

histo_list_programs

```

