---
title: "1st layer marker gene report"
author: "Lea WÃ¶lbert"
date: '2023-03-07'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### Load libraries, source code 

```{r load,  message = FALSE}
library(DropletUtils, quietly = TRUE) 
library(scater, quietly = TRUE) 
library(scuttle, quietly = TRUE) 
library(scran, quietly = TRUE) 
library(tidyverse, quietly = TRUE) 
set.seed(37)
```

```{r}

# load sources
color_tables <- "/omics/odcf/analysis/OE0538_projects/DO-0008/metadata/color_tables"
source(file = "../source/colors.R")
source(file = "../source/plotting.R")
source(file = "../source/sce_functions.R")

#load SCE objects
sce <- readRDS(file = snakemake@input[["sce_10"]])
sce_cluster <- readRDS(file = snakemake@input[["sce_sep"]])
cluster_curr <- snakemake@wildcards[["cluster"]]
fraction_curr <- sce_cluster$Fraction_ID[1]

# load marker genes and results
results_go <- readRDS(file = snakemake@input[["results_go"]])
results_markers <- readRDS(file = snakemake@input[["results_markers"]])

markers <- results_markers[[as.numeric(cluster_curr)]]
go <- results_go[[as.numeric(cluster_curr)]]


print(cluster_curr)
print(sce)
print(fraction_curr)

#knitr::knit_exit()

```

Prepare color schemes

```{r}

sce <- factor_reference_cts(sce)
sce_cluster <- factor_reference_cts(sce_cluster)

col_cts_baccin <- col_cts[match(levels(sce$baccin_celltype_scmapclust), 
                                names(col_cts))]
col_cts_dahlin <- col_cts[match(levels(sce$dahlin_celltype_scmapclust), 
                                names(col_cts))]
col_cts_dolgalev <- col_cts[match(levels(sce$dolgalev_celltype_scmapclust),
                                  names(col_cts))]

col_batch_exp_day <- col_alp[names(col_alp) %in% sce$Batch_exp_day]
col_batch_seq <- col_num[names(col_num) %in% sce$Batch_sequencing]
col_ann_age <- col_ann[names(col_ann) %in% sce$Age_ID]
col_species <- col_spc[names(col_spc) %in% sce$Species_ID]

col_clust_l <- col_num[names(col_num) %in% sce$cluster_louvain]

```

# Identity plots

# Reference cell types 

Cells were annotated with scmap and three reference datasets.

```{r}

identity_plots <- function(sce){
  
  plot_1 <- umap_base(sce, color_by = "baccin_celltype_scmapclust")+ 
    ggtitle(paste0(fraction_curr))+
    scale_color_manual("baccin_celltype_scmapclust", values = col_cts_baccin)
  plot_1l <- umap_legend(sce, color_by = "baccin_celltype_scmapclust")+
    scale_color_manual("baccin_celltype_scmapclust", values = col_cts_baccin)
  legend_1 <- get_legend(plot_1l)

  plot_2 <- umap_base(sce, color_by = "dahlin_celltype_scmapclust")+
    ggtitle(paste0(fraction_curr))+
    scale_color_manual("dahlin_celltype_scmapclust", values = col_cts_dahlin)
  plot_2l <- umap_legend(sce, color_by = "dahlin_celltype_scmapclust")+
    scale_color_manual("dahlin_celltype_scmapclust", values = col_cts_dahlin)
  legend_2 <- get_legend(plot_2l)
  
  plot_3 <- umap_base(sce, color_by = "dolgalev_celltype_scmapclust")+ 
    ggtitle(paste0(fraction_curr))+
    scale_color_manual("dolgalev_celltype_scmapclust",
                       values = col_cts_dolgalev)
  plot_3l <- umap_legend(sce, color_by = "dolgalev_celltype_scmapclust")+
    scale_color_manual("dolgalev_celltype_scmapclust", 
                       values = col_cts_dolgalev)
  legend_3 <- get_legend(plot_3l)
  
  PLOT_1 <- ggarrange(plot_1, legend_1)
  PLOT_2 <- ggarrange(plot_2, legend_2)
  PLOT_3 <- ggarrange(plot_3, legend_3)

  return(list(PLOT_1, PLOT_2, PLOT_3))
}

identity_plots_full <- identity_plots(sce)
identity_plots <- identity_plots(sce_cluster)

```

```{r fig.width=8, fig.height=4}
identity_plots_full
```

```{r fig.width=8, fig.height=4}
identity_plots
```

# Clusters

```{r fig.width=8, fig.height=4}

plot_1 <- umap_base(sce, color_by = "cluster_louvain")+ 
  scale_color_manual("cluster (louvain)", values = col_clust_l)
plot_1l <- umap_legend(sce, color_by = "cluster_louvain")+
  scale_color_manual("cluster (louvain)", values = col_clust_l)
legend_1 <- get_legend(plot_1l)

plot_2 <- pca_base2(sce, color_by = "cluster_louvain")+ 
  scale_color_manual("cluster (louvain)", values = col_clust_l)
plot_2l <- pca_legend(sce, color_by = "cluster_louvain")+
  scale_color_manual("cluster (louvain)", values = col_clust_l)
legend_2 <- get_legend(plot_2l)

plot_3 <- pca_base3(sce, color_by = "cluster_louvain")+ 
  scale_color_manual("cluster (louvain)", values = col_clust_l)
plot_3l <- pca_legend(sce, color_by = "cluster_louvain")+
  scale_color_manual("cluster (louvain)", values = col_clust_l)
legend_3 <- get_legend(plot_3l)

PLOT_1 <- ggarrange(plot_1, legend_1)
PLOT_2 <- ggarrange(plot_2, legend_2)
PLOT_3 <- ggarrange(plot_3, legend_3)
  
PLOT_1
PLOT_2
PLOT_3
```

# Marker genes

## Found marker genes

```{r}
markers[1:50,1:4]
```

```{r}
markers[order(markers$pct.1, decreasing = TRUE),][1:50,1:4]
```

## GO analysis

```{r}

go_short <- go$results[go$results$ontology == "biological_process",]
go$results[1:50,c(3, 4)]
go_short[1:50,c(3, 4)]

```

## Specific literature markers of interest

Prepare Plots

```{r}

gene_exps <- function(gene){
  
  sce_cluster$Species_ID <- factor(sce_cluster$Species_ID,
                                   levels = c("mmus", "mcas", "mspr", "mcar"))

  sce$Species_ID <- factor(sce$Species_ID,
                           levels = c("mmus", "mcas", "mspr", "mcar"))
  
  # Gene expression UMAP in cluster X
  plot_1 <- umap_gene(sce_cluster, gene)

  # Gene expression UMAP in cluster X, separated by species
  plot_2 <- umap_gene(sce_cluster, gene)+
    facet_grid(cols = vars(sce_cluster$Species_ID))
  
  ggdf <- as.data.frame(colData(sce_cluster))
  logc <- assays(sce_cluster)[["logcounts"]]

  ggdf$Species_ID <- factor(ggdf$Species_ID,
                            levels = c("mmus", "mcas", "mspr", "mcar"))
  
  plot_3 <- ggplot(ggdf, aes(x = cluster_louvain,
                             y = logc[rownames(logc) == gene,]))+
    ggbeeswarm::geom_quasirandom(size = 0.2, color = "grey60")+
    theme_classic()+
    geom_boxplot(color = "black", alpha = 0)+
    ylab(paste0(gene, " expression"))+
    facet_grid(cols = vars(sce_cluster$Species_ID))
  
  ggdf_full <- as.data.frame(colData(sce))
  logc_full <- assays(sce)[["logcounts"]]
  
  plot_4 <- umap_gene(sce, gene)
  
  plot_5 <- ggplot(ggdf_full, aes(x = cluster_louvain,
                           y = logc_full[rownames(logc_full) == gene,], 
                           color = cluster_louvain))+
    ggbeeswarm::geom_quasirandom(size = 0.2)+
    theme_classic()+
    scale_color_manual(values = col_clust_l)+
    geom_boxplot(color = "black", alpha = 0)+
    ylab(paste0(gene, " expression"))

  plot_6 <- umap_gene(sce, gene)+
    facet_grid(cols = vars(sce$Species_ID))
  
  plot_7 <- ggplot(ggdf_full, aes(x = Species_ID,
                           y = logc_full[rownames(logc_full) == gene,]))+
    ggbeeswarm::geom_quasirandom(size = 0.2, color = "grey60")+
    theme_classic()+
    geom_boxplot(color = "black", alpha = 0)+
    ylab(paste0(gene, " expression"))
  
  return(list(plot_1, plot_2, plot_3, plot_4, plot_5, plot_6, plot_7, gene))
}

```

```{r}

found_markers <- rownames(markers)[1:50]

#TODO: TAKE FROM CSV

basic_markers_hsc <- c(
  'Adgrl4',
  'Birc5',
  'Blvrb',
  'Car1',
  'Car2',
  'Cd9',
  'Cd34',
  'Cd52',
  'Cd63',
  'Cd48',
  'Cd74',
  'Cd79a',
  'Cd79b',
  'Cdk1',
  'Cdk4',
  'Cdk6',
  'Cebpa',
  'Cebpb',
  'Cebpe',
  'Cenpa',
  'Cenpe',
  'Cenpf',
  'Cited2',
  'Ctsg',
  'Dntt',
  'Egr1',
  'Elane',
  'Esam',
  'Fli1',
  'Flt3',
  'Gata1',
  'Gata2',
  'Hba-a1',
  'Hlf',
  'Hmga2',
  'Hmgb2',
  'Hmgb3',
  'Il7r',
  'Kit',
  'Klf1',
  'Ly6a',
  'Ly6d',
  'Ly6e',
  'Mecom',
  'Mllt3',
  'Mmrn1',
  'Meis1',
  'Mki67',
  'Mpo',
  'Pbx1',
  'Pf4',
  'Plac8',
  'Prdx4',
  'Procr',
  'S100a8',
  'S100a9',
  'Satb1',
  'Slamf1',
  'Sox4',
  'Spi1',
  'Top2a',
  'Tyrobp'
)

specific_markers_hsc <- c(
  'Alas1',
  'Alcam',
  'Aldh1a1',
  'Alox5ap',
  'Anxa3',
  'Ap3s1',
  'Calr',
  'Ccnd2',
  'Ccne1',
  'Cd7',
  'Cdkn1c',
  'Chil1',
  'Chil3',
  'Clec12a',
  'Clec1a',
  'Clec12a',
  'Cpa3',
  'Crip1',
  'Ebf1',
  'Emb',
  'Epor',
  'Erg',
  'F11r',
  'F13a1',
  'Fabp5',
  'Fcer1g',
  'Fcgr3',
  'Fcnb',
  'Fgd5',
  'Foxo1',
  'Fxyd5',
  'Gfi1',
  'Gfi1b',
  'Gng11',
  'Grb10',
  'Gstm1',
  'Gzmb',
  'Hdc',
  'Hells',
  'Hist1h1e',
  'Hoxa9',
  'Id2',
  'Igf1r',
  'Il17ra',
  'Irf8',
  'Itga2b',
  'Itgam',
  'Itgb2l',
  'Krt18',
  'Lcn2',
  'Ldb1',
  'Ldhb',
  'Lmo4',
  'Lrg1',
  'Lsp1',
  'Lst1',
  'Ltb',
  'Ltf',
  'Ly86',
  'Lyz1',
  'Lyz2',
  'Mcm6',
  'Mcm7',
  'Mcpt8',
  'Mef2c',
  'Mif',
  'Mpl',
  'Ms4a3',
  'Ms4a6c',
  'Myb',
  'Mycn',
  'Ndst2',
  'Ngp',
  'Nme1',
  'Pglyrp1',
  'Pkm',
  'Prdx4',
  'Prss50',
  'Prtn3',
  'Rab27b',
  'Ramp1',
  'Rap1b',
  'Rbp1',
  'Rif1',
  'Runx1',
  'S100a10',
  'Sell',
  'Serpinb6a',
  'Sh3bgrl3',
  'Smim1',
  'Socs2',
  'Sptbn1',
  'Srgn',
  'Syce2',
  'Syne1',
  'Tal1',
  'Tespa1',
  'Tet1',
  'Tgm2',
  'Tie1',
  'Tinagl1',
  'Tmsb10',
  'Tmsb4x',
  'Txnip',
  'Uba7',
  'Ube2c',
  'Ugcg',
  'Vamp5'
)

base_markers_str <- c(
  'Acan',
  'Acta1',
  'Alpl',
  'Angpt1',
  'Bglap',
  'Cd34',
  'Tfrc',
  'Cd74',
  'Cdh5',
  'Cebpb',
  'Cenpa',
  'Col13a1',
  'Col1a1',
  'Cxcl12',
  'Elane',
  'Emcn',
  'Hba-a1',
  'Lepr',
  'Lpl',
  'Ly6a',
  'Mag',
  'Mef2c',
  'Mki67',
  'Pdgfra',
  'Pecam1',
  'Ptprc',
  'Rag1',
  'Rag2',
  'S100a8',
  'S100a9',
  'Kitl',
  'Sox9',
  'Sp7',
  'Spp1',
  'Tagln',
  'Top2a',
  'Vcam1',
  'Vim',
  'Vpreb1',
  'Vpreb3',
  'Wif1',
  'Zfpm1'
)  

if(fraction_curr == "hsc"){
  all_markers <- unique(c(found_markers, basic_markers_hsc, specific_markers_hsc))
}else if(fraction_curr == "str"){
  all_markers <- unique(c(found_markers, base_markers_str))
}

all_markers <- as.list(sort(all_markers))

all_markers[which(!all_markers %in% rownames(sce))]
print(all_markers)

#knitr::knit_exit()
```

## UMAPs

```{r}

plotlist <- lapply(all_markers, gene_exps)
plotlist

```

```{r}
sessionInfo()
```

