---
title: "Signatures summary"
author: "Lea WÃ¶lbert"
date: '2023-11-15'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Report on conserved signatures.

#### Load objects

```{r seed, message = FALSE}
set.seed(37)
```

```{r load, message = FALSE}
library(tidyverse, quietly = TRUE)
library(scater, quietly = TRUE)
library(ComplexUpset, quietly = TRUE)
```

```{r source, message = FALSE}
source(file = snakemake@params[["plotting"]])
```

```{r manual, eval = FALSE, include = FALSE}

signature_list_str <- readRDS("/omics/odcf/analysis/OE0538_projects/DO-0008/data/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_conserved_EMF// ")
signature_list_hsc <- readRDS("/omics/odcf/analysis/OE0538_projects/DO-0008/data/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_conserved_EMF/01_emfs/signature_list_hsc")

signature_list_str <- readRDS("/omics/odcf/analysis/OE0538_projects/DO-0008/data/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_conserved_EMF/01_emfs/signature_list_str")
signature_list_hsc <- readRDS("/omics/odcf/analysis/OE0538_projects/DO-0008/data/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_conserved_EMF/01_emfs/signature_list_hsc")

#colors_path <- "/omics/odcf/analysis/OE0538_projects/DO-0008/metadata/colors/colors.txt"
#source("../../source/colors.R")


# fraction_curr <- "hsc"
#sce <- readRDS("/omics/odcf/analysis/OE0538_projects/DO-0008/data/scRNAseq/main_analysis/sce_objects/02_sce_anno/10_anns/sce_str-10")
#emf_list <- readRDS(paste0("/omics/odcf/analysis/OE0538_projects/DO-0008/data/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_conserved_EMF/01_emfs/emf_list_", fraction_curr))
# sce <- readRDS(paste0("/omics/odcf/analysis/OE0538_projects/DO-0008/data/scRNAseq/main_analysis/sce_objects/02_sce_anno/10_anns/sce_", fraction_curr, "-10"))
#ensembl_hum <- readRDS("/omics/odcf/analysis/OE0538_projects/DO-0008/data/metadata/scRNAseq/03_sce_analysis/reclustering_bm/ensembl_hum")
#ensembl_mus <- readRDS("/omics/odcf/analysis/OE0538_projects/DO-0008/data/metadata/scRNAseq/03_sce_analysis/reclustering_bm/ensembl_mus")

```


```{r load_objects}
signature_list_hsc <- readRDS(snakemake@input[["signature_list_hsc"]])
signature_list_str <- readRDS(snakemake@input[["signature_list_str"]])

fraction_curr <- snakemake@wildcards[["fraction"]]
```

```{r colors, meassage = FALSE}
colors_path <- snakemake@params[["colors_path"]]
source(snakemake@params[["colors"]])
```

# Preparation 

Put all dfs containing info on conservation level per cell type together.

```{r}
cts_str <- as.list(names(signature_list_str))
cts_hsc <- as.list(names(signature_list_hsc))

conserved_df_list_str <- lapply(cts_str, function(c){
  
  signature_list_str[[c]]$conserved_df$cell_type <- c
  return(signature_list_str[[c]]$conserved_df)
})

conserved_df_list_hsc <- lapply(cts_hsc, function(c){
  
  signature_list_hsc[[c]]$conserved_df$cell_type <- c
  return(signature_list_hsc[[c]]$conserved_df)
})

conserved_df_str <- dplyr::bind_rows(conserved_df_list_str)
conserved_df_hsc <- dplyr::bind_rows(conserved_df_list_hsc)
cons_large_df <- BiocGenerics::rbind(conserved_df_str, conserved_df_hsc)
print(head(cons_large_df))
```

# Upset plots

Make one df specifically for upset plot

```{r prepare_upset}
cons_large_df_up <- cons_large_df
cons_large_df_up$mmus[!is.na(cons_large_df_up$mmus)] <- TRUE
cons_large_df_up$mcas[!is.na(cons_large_df_up$mcas)] <- TRUE
cons_large_df_up$mspr[!is.na(cons_large_df_up$mspr)] <- TRUE
cons_large_df_up$mcar[!is.na(cons_large_df_up$mcar)] <- TRUE
cons_large_df_up$mmus[is.na(cons_large_df_up$mmus)] <- FALSE
cons_large_df_up$mcas[is.na(cons_large_df_up$mcas)] <- FALSE
cons_large_df_up$mspr[is.na(cons_large_df_up$mspr)] <- FALSE
cons_large_df_up$mcar[is.na(cons_large_df_up$mcar)] <- FALSE

head(cons_large_df_up)
base::table(cons_large_df_up$conserved_signature)
knitr::knit_exit()
```

```{r upset}
ComplexUpset::upset(cons_large_df_up, 
                    intersect = c("mmus", "mcas", "mspr",  "mcar"))+
  xlab("")+
  ggtitle("all marker genes")

ComplexUpset::upset(cons_large_df_up[cons_large_df_up$ndge == TRUE,], 
                    intersect = c("mmus", "mcas", "mspr",  "mcar"))+
  xlab("")+
  ggtitle("non-differentially expressed marker genes")

ComplexUpset::upset(cons_large_df_up[cons_large_df_up$cons_marker == TRUE,], 
                    intersect = c("mmus", "mcas", "mspr",  "mcar"))+
  xlab("")+
  ggtitle("conserved marker genes")

ComplexUpset::upset(cons_large_df_up[cons_large_df_up$cons_emf == TRUE,], 
                    intersect = c("mmus", "mcas", "mspr",  "mcar"))+
  xlab("")+
  ggtitle("signature genes")

knitr::knit_exit()
```

Make a df with categories of conservation for visualisation.

```{r}
ggdf2 <- cons_emf_df

ggdf2$marker_only <- vector(length = nrow(ggdf2))
ggdf2$ndge_marker <- vector(length = nrow(ggdf2))
ggdf2$conserved_marker <- vector(length = nrow(ggdf2))
ggdf2$conserved_emf <- vector(length = nrow(ggdf2))

ggdf2$marker_only[which(ggdf2$cons_marker == FALSE &
                          ggdf2$cons_emf == FALSE &
                          ggdf2$ndge == FALSE)] <- TRUE

ggdf2$ndge_marker[which(ggdf2$cons_marker == FALSE &
                              ggdf2$cons_emf == FALSE &
                              ggdf2$ndge == TRUE)] <- TRUE

ggdf2$conserved_marker[which(ggdf2$cons_marker == TRUE &
                             ggdf2$cons_emf == FALSE &
                             ggdf2$ndge == FALSE)] <- TRUE

ggdf2$conserved_emf[which(ggdf2$cons_marker == TRUE &
                             ggdf2$cons_emf == TRUE &
                             ggdf2$ndge == TRUE)] <- TRUE

ggdf2 <- as.data.frame(pivot_longer(ggdf2, cols = c("marker_only",
                                                    "ndge_marker",
                                                    "conserved_marker",
                                                    "conserved_emf"), 
                                     names_to = "mode", values_to = "is_true"))

ggdf2 <- as.data.frame(pivot_longer(ggdf2, cols = c("mmus",
                                                    "mcas",
                                                    "mspr",
                                                    "mcar"), 
                                     names_to = "species", values_to = "value"))
ggdf2 <- ggdf2[ggdf2$is_true == TRUE,]
ggdf2 <- ggdf2[!is.na(ggdf2$value),]

ggdf2$cell_type <- factor(ggdf2$cell_type, levels = c(names(col_cts_hsc),
                                                      names(col_cts_str)))
ggdf2$species <- factor(ggdf2$species, levels = names(col_spc))

ggdf2$is_emf <- vector(length = nrow(ggdf2))

ggdf2$is_emf[ggdf2$mode == "marker_only"] <- FALSE
ggdf2$is_emf[ggdf2$mode == "ndge_marker"] <- FALSE
ggdf2$is_emf[ggdf2$mode == "conserved_marker"] <- FALSE
ggdf2$is_emf[ggdf2$mode == "conserved_emf"] <- TRUE

ggdf2$is_emf <- factor(ggdf2$is_emf, levels =c(FALSE, TRUE))

head(ggdf2)
```

expand the df

```{r}
ggdf3 <- ggdf2

species <- names(col_spc)
celltypes <- c(names(col_cts_hsc), names(col_cts_str))

ggdf3$nr_markers_all <- vector(length = nrow(ggdf3))
ggdf3$nr_markers_only <- vector(length = nrow(ggdf3))
ggdf3$nr_ndge_marker <- vector(length = nrow(ggdf3))
ggdf3$nr_conserved_marker <- vector(length = nrow(ggdf3))
ggdf3$nr_cons_emf <- vector(length = nrow(ggdf3))
ggdf3$nr_is_emf<- vector(length = nrow(ggdf3))
ggdf3$nr_is_not_emf <- vector(length = nrow(ggdf3))

for(s in species){
  for(c in celltypes){
    
    pos <- which(ggdf3$species == s & ggdf3$cell_type == c)

    ggdf3[pos,]$nr_markers_all <- length(pos)
   
    ggdf3[pos,]$nr_markers_only <- length(which(ggdf3[pos,]$mode == "marker_only"))
    ggdf3[pos,]$nr_ndge_marker <- length(which(ggdf3[pos,]$mode == "ndge_marker"))
    ggdf3[pos,]$nr_conserved_marker <- length(which(ggdf3[pos,]$mode == "conserved_marker"))
    ggdf3[pos,]$nr_cons_emf <- length(which(ggdf3[pos,]$mode == "conserved_emf"))
    
    ggdf3[pos,]$nr_is_emf <- ggdf3[pos,]$nr_cons_emf
    ggdf3[pos,]$nr_is_not_emf <- length(pos) - ggdf3[pos,]$nr_cons_emf
    
  }
}

ggdf3$prop_markers_only <- ggdf3$nr_markers_only/ggdf3$nr_markers_all
ggdf3$prop_ndge_marker <- ggdf3$nr_ndge_marker/ggdf3$nr_markers_all
ggdf3$prop_conserved_marker <- ggdf3$nr_conserved_marker/ggdf3$nr_markers_all
ggdf3$prop_cons_emf <- ggdf3$nr_cons_emf/ggdf3$nr_markers_all
ggdf3$prop_is_emf <- ggdf3$nr_is_emf/ggdf3$nr_markers_all
ggdf3$prop_is_not_emf <- ggdf3$nr_is_not_emf/ggdf3$nr_markers_all

ggdf3_nr <- pivot_longer(ggdf3, cols = c("nr_markers_only",
                                         "nr_ndge_marker",
                                         "nr_conserved_marker",
                                         "nr_cons_emf"), 
                         values_to = "nrs",
                        names_to = "mode2")

ggdf3_prop <- pivot_longer(ggdf3, cols = c("prop_markers_only",
                                         "prop_ndge_marker",
                                         "prop_conserved_marker",
                                         "prop_cons_emf"), 
                         values_to = "prop",
                        names_to = "mode2")

ggdf3_emf <- pivot_longer(ggdf3, cols = c("nr_is_emf",
                                         "nr_is_not_emf"), 
                         values_to = "nrs",
                        names_to = "mode4")

ggdf3_emf_prop <- pivot_longer(ggdf3, cols = c("prop_is_emf",
                                         "prop_is_not_emf"), 
                         values_to = "prop",
                        names_to = "mode4")

ggdf3_prop$mode2 <- gsub("prop_", "", ggdf3_prop$mode2)
ggdf3_nr$mode2 <- gsub("nr_", "", ggdf3_nr$mode2)
ggdf3_emf$mode4 <- gsub("nr_", "", ggdf3_emf$mode4)
ggdf3_emf_prop$mode4 <- gsub("prop_", "", ggdf3_emf_prop$mode4)

ggdf3_prop$mode2 <- factor(ggdf3_prop$mode2, 
                           levels = c("markers_only", "ndge_marker",
                                      "conserved_marker", "cons_emf"))

ggdf3_nr$mode2 <- factor(ggdf3_nr$mode2, 
                           levels = c("markers_only", "ndge_marker",
                                      "conserved_marker", "cons_emf"))
head(ggdf3_nr)
```

```{r}

ggdf2$mode[ggdf2$mode == "marker_only"] <- "marker only"
ggdf2$mode[ggdf2$mode == "ndge_marker"] <- "non-differentally expressed"
ggdf2$mode[ggdf2$mode == "conserved_marker"] <- "conserved function"
ggdf2$mode[ggdf2$mode == "conserved_emf"] <- "conserved marker gene function and expression"
ggdf2$mode <- factor(ggdf2$mode, 
                           levels = c("marker only", "non-differentally expressed",
                                      "conserved function", "conserved marker gene function and expression"))

col_mode1 <-c("marker only" = "#EDE2D4",
              "non-differentally expressed" = "#ebbf85ff",
              "conserved function" = "#EB9523",
              "conserved marker gene function and expression" = "#DD5C00")


col_mode2 <- c("markers_only" = "#EDE2D4",
               "ndge_marker"= "#ebbf85ff",
               "conserved_marker" = "#f4a036ff",
               "conserved_emf"= "#DD5C00")

col_core <- c("TRUE" = "#DD5C00",
              "FALSE" = "#EDE2D4")

ggdf3_emf$mode4[ggdf3_emf$mode4 == "is_emf"] <- "TRUE"
ggdf3_emf$mode4[ggdf3_emf$mode4 == "is_not_emf"] <- "FALSE"

ggdf3_emf_prop$mode4[ggdf3_emf_prop$mode4 == "is_emf"] <- "TRUE"
ggdf3_emf_prop$mode4[ggdf3_emf_prop$mode4 == "is_not_emf"] <- "FALSE"

ggdf3_emf$mode4 <- factor(ggdf3_emf$mode4, 
                           levels = c("TRUE", "FALSE"))

ggdf3_emf_prop$mode4 <- factor(ggdf3_emf_prop$mode4, 
                           levels = c("TRUE", "FALSE"))

col_mode4 <- c("TRUE" = "#DD5C00",
               "FALSE" = "#EDE2D4")
```

# Important Plots



```{r fig.width = 12, fig.height = 5.5}
exclude_markero <- which(ggdf2$cons_marker == FALSE & ggdf2$ndge == FALSE &
                           ggdf2$cons_emf == FALSE)
  
ggplot(ggdf2[-exclude_markero,], aes(fill = mode, x = species))+
  geom_bar(position = "stack")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Marker genes", values = col_mode1)+
  facet_grid(cols = vars(cell_type))+
  theme(strip.text.x = element_text(angle = 90),
        axis.text.x = element_text(angle = 90),
        legend.position = "bottom")+
  xlab("Species")

ggplot(ggdf2[-exclude_markero,], aes(fill = mode, x = species))+
  geom_bar(position = "fill")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Conservation level", values = col_mode1)+
  facet_grid(cols = vars(cell_type))+
  theme(strip.text.x = element_text(angle = 90),
        axis.text.x = element_text(angle = 90))+
  xlab("Species")

ggplot(ggdf2[-exclude_markero,], aes(fill = mode, x = cell_type))+
  geom_bar(position = "fill")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Conservation level", values = col_mode1)+
  facet_grid(cols = vars(species))+
  theme(strip.text.x = element_text(angle = 90),
        axis.text.x = element_text(angle = 90))+
  xlab("Cell type")
```

```{r fig.width = 5.5, fig.height = 7}
exclude_ndge <- which(ggdf2$cons_marker == FALSE & ggdf2$ndge == TRUE &
                           ggdf2$cons_emf == FALSE)
  

col_mode1_new <- c(col_mode1[c(1, 3, 4)])
col_mode1_new[1] <- "#c8bbacff"
col_mode1_new[2] <- "#eb9523ff"
col_mode1_new[3] <- "#dd5c00ff"

ggdf2_ex <- ggdf2[-exclude_ndge,]
ggdf2_ex$species_rev <- ggdf2_ex$species
ggdf2_ex$species_rev <- factor(ggdf2_ex$species_rev, levels = rev(names(col_spc)))
plot1 <- ggplot(ggdf2_ex[ggdf2_ex$cell_type %in% names(col_cts_hsc),],
       aes(fill = mode, y = species_rev))+
  geom_bar(position = "stack")+
  theme_classic()+
  xlab("Nr of detected marker genes")+
  scale_fill_manual(values = col_mode1_new)+
  facet_grid(rows = vars(cell_type), switch = "y")+
  scale_y_discrete(position = "right")+
  theme(strip.text.y.left = element_text(angle = 0),
        axis.text.y = element_text(angle = 0),
        legend.position = "top", axis.title.y = element_blank(),
        legend.title = element_blank())+
  xlim(c(0, 250))+
  theme(strip.background = element_rect(fill = "grey97",linetype = 0))

plot2 <- ggplot(ggdf2_ex[ggdf2_ex$cell_type %in% names(col_cts_str),],
       aes(fill = mode, y = species_rev))+
  geom_bar(position = "stack")+
  theme_classic()+
  xlab("Nr of detected marker genes")+
  scale_fill_manual(values = col_mode1_new)+
  facet_grid(rows = vars(cell_type), switch = "y")+
  scale_y_discrete(position = "right")+
  theme(strip.text.y.left = element_text(angle = 0),
        axis.text.y = element_text(angle = 0),
        legend.position = "top",
        legend.title = element_blank(), axis.title.y = element_blank())+
  xlim(c(0, 450))+
  theme(strip.background = element_rect(fill = "grey97",linetype = 0))

legend <- cowplot::get_legend(plot1)
```


```{r fig.width = 7, fig.height = 6.5}
plot1+theme(legend.position = "none",
            strip.text = element_text(size = 14),
            axis.text.x = element_text(size = 12),
            axis.title.x = element_text(size = 12),
            axis.text.y = element_blank())
```


```{r fig.width = 7, fig.height = 4.56}
plot2+theme(legend.position = "none",
            strip.text = element_text(size = 14),
            axis.text.x = element_text(size = 12),
            axis.title.x = element_text(size = 14),
            axis.text.y = element_blank(),
            strip.placement = "outside")
plot2+theme(legend.position = "right")
```


```{r fig.width = 13, fig.height = 4}
ggplot(ggdf2, aes(fill = mode, x = species))+
  geom_bar(position = "stack")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Marker genes", values = col_mode1)+
  facet_grid(cols = vars(cell_type))+
  theme(strip.text.x = element_text(angle = 90),
        axis.text.x = element_text(angle = 90))+
  xlab("Species")

ggplot(ggdf2, aes(fill = mode, x = species))+
  geom_bar(position = "fill")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Marker genes", values = col_mode1)+
  facet_grid(cols = vars(cell_type))+
  theme(strip.text.x = element_text(angle = 90),
        axis.text.x = element_text(angle = 90))+
  xlab("Species")

ggplot(ggdf2, aes(fill = mode, x = cell_type))+
  geom_bar(position = "fill")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Marker genes", values = col_mode1)+
  facet_grid(cols = vars(species))+
  theme(strip.text.x = element_text(angle = 90),
        axis.text.x = element_text(angle = 90))+
  xlab("Cell type")
```

```{r, fig.height = 5.5, fig.width=8}

ggdf2_plot <- ggdf2
ggdf2_plot$species <- factor(ggdf2_plot$species, levels = rev(names(col_spc)))

ggplot(ggdf2_plot[ggdf2_plot$assignment == "emitter",], aes(fill = mode, y = species))+
  geom_bar(position = "stack")+
  theme_classic()+
  xlab("Nr of detected marker genes")+
  scale_fill_manual("Marker genes", values = col_mode1)+
  facet_grid(rows = vars(ggdf2_plot[ggdf2_plot$assignment == "emitter",]$cell_type))+
  theme(strip.text.y = element_text(angle = 0, hjust = 0),
        axis.text.y = element_text(angle = 0),
        axis.title.y = element_blank(),
        strip.background = element_rect(fill = "grey95",linetype = 0),
        strip.placement = "outside",
        legend.title = element_text(face = "bold"))

col_mode1_plot1 <-c("marker only" = "#EDE2D4",
              "non-differentally expressed" = "#EDE2D4",
              "conserved function" = "#EDE2D4",
              "conserved marker gene function and expression" = "#EDE2D4")
  
ggplot(ggdf2_plot[ggdf2_plot$assignment == "emitter",], aes(fill = mode, y = species))+
  geom_bar(position = "stack")+
  theme_classic()+
  xlab("Nr of detected marker genes")+
  scale_fill_manual("Marker genes", values = col_mode1_plot1)+
  facet_grid(rows = vars(ggdf2_plot[ggdf2_plot$assignment == "emitter",]$cell_type))+
  theme(strip.text.y = element_text(angle = 0, hjust = 0),
        axis.text.y = element_text(angle = 0),
        axis.title.y = element_blank(),
        strip.background = element_rect(fill = "grey95",linetype = 0),
        strip.placement = "outside",
        legend.title = element_text(face = "bold"))

col_mode1_plot2 <-c("marker only" = "#EDE2D4",
              "non-differentally expressed" = "#ebbf85ff",
              "conserved function" = "#ebbf85ff",
              "conserved marker gene function and expression" = "#ebbf85ff")
  
ggplot(ggdf2_plot[ggdf2_plot$assignment == "emitter",], aes(fill = mode, y = species))+
  geom_bar(position = "stack")+
  theme_classic()+
  xlab("Nr of detected marker genes")+
  scale_fill_manual("Marker genes", values = col_mode1_plot2)+
  facet_grid(rows = vars(ggdf2_plot[ggdf2_plot$assignment == "emitter",]$cell_type))+
  theme(strip.text.y = element_text(angle = 0, hjust = 0),
        axis.text.y = element_text(angle = 0),
        axis.title.y = element_blank(),
        strip.background = element_rect(fill = "grey95",linetype = 0),
        strip.placement = "outside",
        legend.title = element_text(face = "bold"))

col_mode1_plot3 <-c("marker only" = "#EDE2D4",
              "non-differentally expressed" = "#ebbf85ff",
              "conserved function" = "#f4a036ff",
              "conserved marker gene function and expression" = "#f4a036ff")
  
ggplot(ggdf2_plot[ggdf2_plot$assignment == "emitter",], aes(fill = mode, y = species))+
  geom_bar(position = "stack")+
  theme_classic()+
  xlab("Nr of detected marker genes")+
  scale_fill_manual("Marker genes", values = col_mode1_plot3)+
  facet_grid(rows = vars(ggdf2_plot[ggdf2_plot$assignment == "emitter",]$cell_type))+
  theme(strip.text.y = element_text(angle = 0, hjust = 0),
        axis.text.y = element_text(angle = 0),
        axis.title.y = element_blank(),
        strip.background = element_rect(fill = "grey95",linetype = 0),
        strip.placement = "outside",
        legend.title = element_text(face = "bold"))
  
```

```{r, fig.height = 6.5, fig.width=8}
ggplot(ggdf2_plot[ggdf2_plot$assignment == "receiver",], aes(fill = mode, y = species))+
  geom_bar(position = "stack")+
  theme_classic()+
  xlab("Nr of detected marker genes")+
  scale_fill_manual("Marker genes", values = col_mode1)+
  facet_grid(rows = vars(ggdf2_plot[ggdf2_plot$assignment == "receiver",]$cell_type))+
  theme(strip.text.y = element_text(angle = 0, hjust = 0),
        axis.text.y = element_text(angle = 0),
        axis.title.y = element_blank(),
        strip.background = element_rect(fill = "grey95",linetype = 0),
        strip.placement = "outside",
        legend.title = element_text(face = "bold"))
```


```{r fig.width = 5, fig.height = 4}
ggplot(ggdf2[-exclude_markero,][ggdf2[-exclude_markero,]$species == "mmus",],
       aes(fill = mode, x = cell_type))+
  geom_bar(position = "stack")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Marker genes", values = col_mode1)+
  theme(axis.text.x = element_text(angle = 90))+
  ggtitle("BL6 marker genes")+
  xlab("Cell type")

ggplot(ggdf2[-exclude_markero,][ggdf2[-exclude_markero,]$species == "mmus",],
       aes(fill = mode, x = cell_type))+
  geom_bar(position = "fill")+
  theme_classic()+
  ylab("Proportion of detected marker genes")+
  scale_fill_manual("Marker genes", values = col_mode1)+
  theme(axis.text.x = element_text(angle = 90))+
  ggtitle("BL6 marker genes")+
  xlab("Cell type")
```

```{r fig.width = 5, fig.height = 4}
ggplot(ggdf2[ggdf2$species == "mmus",],
       aes(fill = mode, x = cell_type))+
  geom_bar(position = "stack")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Marker genes", values = col_mode1)+
  theme(axis.text.x = element_text(angle = 90))+
  ggtitle("BL6 all marker genes")+
  xlab("Cell type")

ggplot(ggdf2[ggdf2$species == "mmus",],
       aes(fill = mode, x = cell_type))+
  geom_bar(position = "fill")+
  theme_classic()+
  ylab("Proportion of detected marker genes")+
  scale_fill_manual("Marker genes", values = col_mode1)+
  theme(axis.text.x = element_text(angle = 90))+
  ggtitle("BL6 all marker genes")+
  xlab("Cell type")
```

```{r fig.width = 6, fig.height = 4}
ggplot(ggdf2[-exclude_markero,][ggdf2[-exclude_markero,]$species == "mmus",],
       aes(fill = is_emf, x = cell_type))+
  geom_bar(position = "stack")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Core marker gene", values = col_core)+
  theme(axis.text.x = element_text(angle = 90))+
  ggtitle("BL6 marker genes")+
  xlab("Cell type")

ggplot(ggdf2[-exclude_markero,][ggdf2[-exclude_markero,]$species == "mmus",],
       aes(fill = is_emf, x = cell_type))+
  geom_bar(position = "fill")+
  theme_classic()+
  ylab("Proportion of detected marker genes")+
  scale_fill_manual("Core marker gene", values = col_core)+
  theme(axis.text.x = element_text(angle = 90))+
  ggtitle("BL6 marker genes")+
  xlab("Cell type")
```

```{r fig.width = 6, fig.height = 4}
ggplot(ggdf2[ggdf2$species == "mmus",], aes(fill = is_emf, x = cell_type))+
  geom_bar(position = "stack")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Core marker gene", values = col_core)+
  theme(axis.text.x = element_text(angle = 90))+
  ggtitle("BL6 all marker genes")+
  xlab("Cell type")

ggplot(ggdf2[ggdf2$species == "mmus",], aes(fill = is_emf, x = cell_type))+
  geom_bar(position = "fill")+
  theme_classic()+
  ylab("Proportion of detected marker genes")+
  scale_fill_manual("Core marker gene", values = col_core)+
  theme(axis.text.x = element_text(angle = 90))+
  ggtitle("BL6 all marker genes")+
  xlab("Cell type")
```

# QC Plots

```{r fig.width = 13, fig.height = 4}
ggplot(ggdf2, aes(fill = is_emf, x = species))+
  geom_bar(position = "stack")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Core marker gene", values = col_core)+
  facet_grid(cols = vars(cell_type))+
  theme(strip.text.x = element_text(angle = 90),
        axis.text.x = element_text(angle = 90))+
  xlab("Species")

ggplot(ggdf2, aes(fill = is_emf, x = species))+
  geom_bar(position = "fill")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Core marker gene", values = col_core)+
  facet_grid(cols = vars(cell_type))+
  theme(strip.text.x = element_text(angle = 90),
        axis.text.x = element_text(angle = 90))+
  xlab("Species")

ggplot(ggdf2, aes(fill = is_emf, x = cell_type))+
  geom_bar(position = "fill")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Core marker gene", values = col_core)+
  facet_grid(cols = vars(species))+
  theme(strip.text.x = element_text(angle = 90),
        axis.text.x = element_text(angle = 90))+
  xlab("Cell type")

```

```{r}
# for control reasons, compare directly to HSC report
ggplot(ggdf2[ggdf2$cell_type == "HSCs",], aes(fill = mode, x = species))+
  geom_bar(position = "stack")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Conservation level", values = col_mode1)
```

```{r fig.width = 6, fig.height = 4}
pos <- which(ggdf3_nr$mode2 == "markers_only")

ggplot(ggdf3_nr[-pos,], aes(x = species, y = nrs, fill = mode2))+
  geom_boxplot(color = "black", alpha = 0.8)+
  theme_classic()+
  scale_fill_manual("Conservation level", values = col_mode2)+
  xlab("Species")+
  ylab("Nr of marker genes")+
  theme(axis.text.x = element_text(angle = 90))
```

```{r fig.width = 6, fig.height = 4}
ggplot(ggdf3_nr, aes(x = species, y = nrs, fill = mode2))+
  geom_boxplot(color = "black", alpha = 0.8)+
  theme_classic()+
  scale_fill_manual("Conservation level", values = col_mode2)+
  xlab("Species")+
  ylab("Nr of marker genes")+
  theme(axis.text.x = element_text(angle = 90))

ggplot(ggdf3_prop, aes(x = species, y = prop, fill = mode2))+
  geom_boxplot(color = "black", alpha = 0.8)+
  theme_classic()+
  scale_fill_manual("Conservation level", values = col_mode2)+
  xlab("Species")+
  ylab("Proportion of marker genes")+
  theme(axis.text.x = element_text(angle = 90))
```

```{r fig.width = 5, fig.height = 4}
ggplot(ggdf3_emf_prop, aes(x = species, y = prop, fill = mode4))+
  geom_boxplot(color = "black", alpha = 0.8)+
  theme_classic()+
  scale_fill_manual("Core marker gene", values = col_mode4)+
  xlab("Species")+
  ylab("Proportion of all marker genes")+
  theme(axis.text.x = element_text(angle = 90))
  
ggplot(ggdf3_emf, aes(x = species, y = nrs, fill = mode4))+
  geom_boxplot(color = "black", alpha = 0.8)+
  theme_classic()+
  scale_fill_manual("Core marker gene", values = col_mode4)+
  theme(axis.text.x = element_text(angle = 90))+
  xlab("Species")+
  ylab("Nr of all marker genes")
```

```{r fig.width = 5, fig.height = 4}
ggplot(ggdf3_prop, aes(x = mode2, y = prop, fill = species))+
  geom_boxplot(color = "black", alpha = 0.8)+
  theme_classic()+
  scale_fill_manual("Species", values = col_spc)+
  theme(axis.text.x = element_text(angle = 90))+
  xlab("Conservation level")+
  ylab("Proportion of marker genes")

ggplot(ggdf3_emf_prop, aes(x = mode4, y = prop, fill = species))+
  geom_boxplot(color = "black", alpha = 0.8)+
  theme_classic()+
  scale_fill_manual("Species", values = col_spc)+
  theme(axis.text.x = element_text(angle = 90))+
  xlab("Conservation level")+
  ylab("Proportion of marker genes")
```

```{r}
#knitr::knit_exit()
```

```{r}
signature_list_hsc$HSCs$conserved_df
hsc_list <- signature_list_hsc$HSCs$cons_emf

hsc_list_diff <- c("Procr", "Tgm2", "Cavin2", "Itih5", "Nceh1", "Tie1", "Kit", 
                   "Tnip3", "Nupr1, Sult1a1", "Aplp2", "Nrgn", "Kmt2a", "Trim47",
                   "Rock2", "Fcho", "Tsc22d1", "St3gal1", "Ly6a", "Krt18",
                   "Alcam", "Itsn1")

hsc_list_diff <- c("Procr", "Tgm2", "Tie1", "Kit", "Nrgn", "Ly6a", "Krt18", 
                   "Alcam")

sce <- readRDS("/omics/odcf/analysis/OE0538_projects/DO-0008/data/scRNAseq/main_analysis/sce_objects/02_sce_anno/10_anns/sce_hsc-10")
sce_hsc <- sce[,sce$celltypes == "HSCs"]

agg <- aggregateAcrossCells(sce_hsc, id=colData(sce_hsc)[,c("Species_ID")], 
                            statistics = "mean",
                            use.assay.type = "logcounts_before_BC")
```

```{r fig.width = 2.7, fig.height = 2.7}

agg <- aggregateAcrossCells(sce_hsc, id=colData(sce_hsc)[,c("Species_ID")], 
                            statistics = "mean",
                            use.assay.type = "logcounts")

vis_df <- as.data.frame(
  t(assays(agg)$logcounts[
    rownames(assays(agg)$logcounts) %in% hsc_list_diff,]))
vis_df <- rownames_to_column(vis_df, "species")
vis_df <- pivot_longer(vis_df, cols = c(2:ncol(vis_df)), values_to = "expression", names_to = "gene")

vis_df$species <- factor(vis_df$species, levels = names(col_spc))
vis_df$gene <- factor(vis_df$gene, levels = c(
  "Tgm2",
  "Ly6a",
  "Procr", 
  "Alcam",
  "Tie1",
  "Krt18",
  "Nrgn",
  "Kit"
))

ggplot(vis_df, aes(x = species, y = gene, fill = expression))+
  geom_tile()+
  theme_classic()+
  scale_fill_continuous("average logcounts", low = "white", high = "#7a6e62ff",
                        limits = c(0, 3.7))+
  ggtitle("BL6-specific HSC markers")+
  theme(axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        legend.position = "bottom",
        plot.title = element_blank(),
        axis.text.x = element_text(angle = 90, size = 12, hjust = 0.45),
        #plot.title = element_text(size = 14, hjust = 0.45),
        axis.text.y = element_text(size = 12))

ggplot(vis_df, aes(x = species, y = gene, fill = expression))+
  geom_tile()+
  theme_classic()+
  scale_fill_continuous("average logcounts", low = "white", high = col_spc["mmus"],
                        limits = c(0, 3.7))+
  ggtitle("BL6-specific HSC markers")+
  theme(axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        legend.position = "bottom",
        plot.title = element_blank(),
        axis.text.x = element_text(angle = 90, size = 12, hjust = 0.45),
        #plot.title = element_text(size = 14, hjust = 0.45),
        axis.text.y = element_text(size = 12))

ggplot(vis_df, aes(x = species, y = gene, fill = expression))+
  geom_tile()+
  theme_classic()+
  scale_fill_continuous("average logcounts", low = "white", high = "grey21",
                        limits = c(0, 1.5))+
  ggtitle("BL6-specific HSC markers")+
  theme(axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        legend.position = "bottom",
        plot.title = element_blank(),
        axis.text.x = element_text(angle = 90, size = 12, hjust = 0.45),
        #plot.title = element_text(size = 14, hjust = 0.45),
        axis.text.y = element_text(size = 12))

ggplot(vis_df, aes(x = species, y = gene, fill = expression))+
  geom_tile()+
  theme_classic()+
  scale_fill_continuous("mean expr.", low = "white", high = "blue",
                        limits = c(0, 1.5))+
  ggtitle("BL6-specific HSC markers")+
  theme(axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        legend.position = "bottom",
        #plot.title = element_blank(),
        axis.text.x = element_blank(),
        plot.title = element_text(size = 14, hjust = 0.45),
        axis.text.y = element_text(size = 12))

ggplot(vis_df, aes(x = species, y = gene, fill = expression))+
  geom_tile()+
  theme_classic()+
  scale_fill_continuous("mean expr.", low = "white", high = "blue",
                        limits = c(0, 1.5))+
  ggtitle("BL6-specific HSC markers")+
  theme(axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        #plot.title = element_blank(),
        axis.text.x = element_blank(),
        plot.title = element_text(size = 14, hjust = 0.45),
        axis.text.y = element_text(size = 12))
```

```{r fig.width = 3, fig.height = 4.4}

vis_df <- as.data.frame(
  t(assays(agg)$logcounts[
    rownames(assays(agg)$logcounts) %in% hsc_list,]))
vis_df <- rownames_to_column(vis_df, "species")
vis_df <- pivot_longer(vis_df, cols = c(2:ncol(vis_df)), 
                       values_to = "expression", names_to = "gene")

vis_df$species <- factor(vis_df$species, levels = names(col_spc))
vis_df$gene <- factor(vis_df$gene, levels = rev(c(
  "Lmo2",
  "Txnip",
  "Tmem176b",
  "Tmem176a",
  "Pnrc1",
  "Angpt1",
  "Hlf",
  "Msi2",
  "Mecom",
  "Meis1",
  "Ltb",
  "Gng11",
  "Nkx2-3",
  "Mpl"
)))

ggplot(vis_df, aes(x = species, y = gene, fill = expression))+
  geom_tile()+
  theme_classic()+
  scale_fill_continuous("average logcounts", low = "white", high = "#dd5c00ff",
                        limits = c(0, 3))+
  ggtitle("cons. HSC markers")+
  theme(axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        legend.position = "bottom",
        plot.title = element_blank(),
        axis.text.x = element_text(angle = 90, size = 12, hjust = 0.45),
        #plot.title = element_text(size = 14, hjust = 0.5),
        axis.text.y = element_text(size = 12))

ggplot(vis_df, aes(x = species, y = gene, fill = expression))+
  geom_tile()+
  theme_classic()+
  scale_fill_continuous("mean expr.", low = "white", high = "blue",
                        limits = c(0, 3))+
  ggtitle("cons. HSC markers")+
  theme(axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        legend.position = "bottom",
        #plot.title = element_blank(),
        axis.text.x = element_text(angle = 90, size = 12, hjust = 0.45),
        plot.title = element_text(size = 14, hjust = 0.45),
        axis.text.y = element_text(size = 12))

ggplot(vis_df, aes(x = species, y = gene, fill = expression))+
  geom_tile()+
  theme_classic()+
  scale_fill_continuous("mean expr.", low = "white", high = "blue",
                        limits = c(0, 3))+
  ggtitle("cons. HSC markers")+
  theme(axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        #plot.title = element_blank(),
        axis.text.x = element_text(angle = 90, size = 12, hjust = 0.45),
        plot.title = element_text(size = 14, hjust = 0.45),
        axis.text.y = element_text(size = 12))
```


```{r}
adipo_list <- signature_list_str$`Adipo/CARs`$df_core$gene[order(signature_list_str$`Adipo/CARs`$df_core$mmus, decreasing = TRUE)]

adipo_list <- adipo_list[1:15]

sce <- readRDS("/omics/odcf/analysis/OE0538_projects/DO-0008/data/scRNAseq/main_analysis/sce_objects/02_sce_anno/10_anns/sce_str-10")
sce_adipo <- sce[,sce$celltypes == "Adipo/CARs"]


```

```{r fig.width = 3, fig.height = 4.3}

agg <- aggregateAcrossCells(sce_adipo, id=colData(sce_adipo)[,c("Species_ID")], 
                            statistics = "mean",
                            use.assay.type = "logcounts")

vis_df <- as.data.frame(
  t(assays(agg)$logcounts[
    rownames(assays(agg)$logcounts) %in% adipo_list,]))
vis_df <- rownames_to_column(vis_df, "species")
vis_df <- pivot_longer(vis_df, cols = c(2:ncol(vis_df)), 
                       values_to = "expression", names_to = "gene")

vis_df$species <- factor(vis_df$species, levels = names(col_spc))
vis_df$gene <- factor(vis_df$gene, levels = rev(c(
  "Cxcl12", 
  "Gas6",
  "Tpm1",
  "Plpp3",
  adipo_list[!adipo_list %in% c("Cxcl12", "Tpm1", "Plpp3", "Gas6")]
)))

ggplot(vis_df, aes(x = species, y = gene, fill = expression))+
  geom_tile()+
  theme_classic()+
  scale_fill_continuous("average logcounts", low = "white", high = "#dd5c00ff",
                        limits = c(0, max(vis_df$expression)))+
  ggtitle("Top 15 cons. Adipo/CARs markers")+
  theme(axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 90, size = 12, hjust = 0.45),
        plot.title = element_blank(),
        legend.position = "bottom",
        #plot.title = element_text(size = 14, hjust = 0.35),
        axis.text.y = element_text(size = 12))

ggplot(vis_df, aes(x = species, y = gene, fill = expression))+
  geom_tile()+
  theme_classic()+
  scale_fill_continuous("mean expr.", low = "white", high = "blue")+
  ggtitle("Top 15 cons. Adipo/CARs markers")+
  theme(axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        legend.position = "bottom",
        #legend.justification.bottom = "left",
        #legend.justification = "left",
        #legend.title = element_text(hjust = 0),
        #legend.title.position = "top",
        axis.text.x = element_text(angle = 90, size = 12, hjust = 0.45),
        plot.title = element_text(size = 14, hjust = 1),
        axis.text.y = element_text(size = 12))
```

```{r}
sessionInfo()
```