---
title: "Signatures summary"
author: "Lea WÃ¶lbert"
date: '2023-11-15'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Report on conserved signatures.

#### Load objects

```{r seed, message = FALSE}
set.seed(37)
```

```{r load, message = FALSE}
library(tidyverse, quietly = TRUE)
library(scater, quietly = TRUE)
library(ComplexUpset, quietly = TRUE)
```

```{r source, message = FALSE}
source(file = snakemake@params[["plotting"]])
```

```{r manual, eval = FALSE, include = FALSE}

signature_list_hsc <- readRDS("/omics/odcf/analysis/OE0538_projects/DO-0008/data/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/01_sign/signature_list_hsc")
signature_list_str <- readRDS("/omics/odcf/analysis/OE0538_projects/DO-0008/data/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/01_sign/signature_list_str")

sce_hsc <- readRDS("/omics/odcf/analysis/OE0538_projects/DO-0008/data/scRNAseq/main_analysis/sce_objects/02_sce_anno/10_anns/sce_hsc-10")

colors_path <- "/omics/odcf/analysis/OE0538_projects/DO-0008/data/metadata/colors/colors.txt"
source("../../source/colors.R")

```

```{r load_objects}
sce_hsc <- base::readRDS(snakemake@input[["sce_hsc"]])
sce_str <- base::readRDS(snakemake@input[["sce_str"]])

signature_list_hsc <- base::readRDS(snakemake@input[["signature_list_hsc"]])
signature_list_str <- base::readRDS(snakemake@input[["signature_list_str"]])

fraction_curr <- snakemake@wildcards[["fraction"]]
```

```{r colors, meassage = FALSE}
colors_path <- snakemake@params[["colors_path"]]
source(snakemake@params[["colors"]])
```

# Preparation 

Put all dfs containing info on conservation level per cell type together.

```{r}
cts_str <- as.list(names(signature_list_str))
cts_hsc <- as.list(names(signature_list_hsc))

conserved_df_list_str <- lapply(cts_str, function(c){
  
  signature_list_str[[c]]$conserved_df$cell_type <- c
  signature_list_str[[c]]$conserved_df$fraction <- "str"
  return(signature_list_str[[c]]$conserved_df)
})

conserved_df_list_hsc <- lapply(cts_hsc, function(c){
  
  signature_list_hsc[[c]]$conserved_df$cell_type <- c
  signature_list_hsc[[c]]$conserved_df$fraction <- "hsc"
  return(signature_list_hsc[[c]]$conserved_df)
})

conserved_df_str <- dplyr::bind_rows(conserved_df_list_str)
conserved_df_hsc <- dplyr::bind_rows(conserved_df_list_hsc)
cons_large_df <- BiocGenerics::rbind(conserved_df_str, conserved_df_hsc)
print(head(cons_large_df))
```

# Upset plots

Make one df specifically for upset plot

```{r prepare_upset}
cons_large_df_up <- cons_large_df
cons_large_df_up$mmus[which(!is.na(cons_large_df_up$mmus))] <- TRUE
cons_large_df_up$mcas[which(!is.na(cons_large_df_up$mcas))] <- TRUE
cons_large_df_up$mspr[which(!is.na(cons_large_df_up$mspr))] <- TRUE
cons_large_df_up$mcar[which(!is.na(cons_large_df_up$mcar))] <- TRUE
cons_large_df_up$mmus[which(is.na(cons_large_df_up$mmus))] <- FALSE
cons_large_df_up$mcas[which(is.na(cons_large_df_up$mcas))] <- FALSE
cons_large_df_up$mspr[which(is.na(cons_large_df_up$mspr))] <- FALSE
cons_large_df_up$mcar[which(is.na(cons_large_df_up$mcar))] <- FALSE

head(cons_large_df_up)
base::table(cons_large_df_up$conserved_signature)

base::table(base::duplicated(cons_large_df_up[
  cons_large_df_up$ndge == TRUE,]))
base::table(base::duplicated(cons_large_df_up[
  cons_large_df_up$conserved_marker == TRUE,]))
base::table(base::duplicated(cons_large_df_up[
  cons_large_df_up$conserved_signature == TRUE,]))
```

```{r upset}
ComplexUpset::upset(cons_large_df_up[which(
  cons_large_df_up$ndge == TRUE & cons_large_df_up$fraction == "str"),], 
  intersect = c("mmus", "mcas", "mspr",  "mcar"))+
  xlab("")+
  ggplot2::ggtitle("all marker genes, Stromal cells")

ComplexUpset::upset(cons_large_df_up[which(
  cons_large_df_up$ndge == TRUE & cons_large_df_up$fraction == "hsc"),], 
  intersect = c("mmus", "mcas", "mspr",  "mcar"))+
  xlab("")+
  ggplot2::ggtitle("all marker genes, HSPCs")
```

# Categorize

## Total markers

Make a df with categories of conservation for visualisation.

```{r prepare_ggdf_category}
ggdf_category <- cons_large_df

ggdf_category$marker_only <- base::rep(FALSE, nrow(ggdf_category))
ggdf_category$conserved_marker_only <- base::rep(FALSE, nrow(ggdf_category))
ggdf_category$conserved_signature_only <- base::rep(FALSE, nrow(ggdf_category))

# genes that are only non-conserved markers
ggdf_category$marker_only[
  which(ggdf_category$conserved_marker == FALSE &
          ggdf_category$conserved_signature == FALSE)] <- TRUE
base::table(ggdf_category$marker_only)

# genes that are conserved markers, but not part of a conserved signature
# because they were not nDGE
ggdf_category$conserved_marker_only[
  which(ggdf_category$conserved_marker == TRUE &
          ggdf_category$conserved_signature == FALSE)] <- TRUE
base::table(ggdf_category$conserved_marker_only)

# genes that are part of a conserved signature
ggdf_category$conserved_signature_only[
  which(ggdf_category$conserved_marker_only == FALSE &
          ggdf_category$conserved_signature == TRUE)] <- TRUE
base::table(ggdf_category$conserved_signature_only)

# pivot longer by category
ggdf_category_long <- base::as.data.frame(tidyr::pivot_longer(
  ggdf_category,
  cols = c("marker_only",
           "conserved_marker_only",
           "conserved_signature_only"), 
  names_to = "conservation_level", 
  values_to = "is_true"))

# remove unneccessary rows that have been generated by pivot longer
ggdf_category_long <- ggdf_category_long[ggdf_category_long$is_true == TRUE,]

# rename and factor
ggdf_category_long$conservation_level <- base::gsub(
  "_only", "", ggdf_category_long$conservation_level)
ggdf_category_long$conservation_level <- factor(
  ggdf_category_long$conservation_level,
  levels = c("marker",
             "conserved_marker",
             "conserved_signature"))

ggdf_category_long$cell_type_rev <- factor(
  ggdf_category_long$cell_type,
  levels = base::rev(c(names(col_cts_hsc), names(col_cts_str))))
ggdf_category_long$cell_type <- factor(
  ggdf_category_long$cell_type,
  levels = c(names(col_cts_hsc), names(col_cts_str)))
```

```{r plot_category1}
plot_cat_1 <- ggplot2::ggplot(ggdf_category_long,
                              aes(fill = conservation_level, 
                                  y = cell_type_rev))+
  ggplot2::geom_bar(position = "stack")+
  ggplot2::theme_classic()+
  ggplot2::xlab("Nr of detected marker genes")+
  ggplot2::scale_fill_manual("", values = col_mode)+
  ggplot2::theme(strip.text.x = element_text(angle = 90),
                 axis.text.x = element_text(angle = 90),
                 legend.position = "bottom")+
  ggplot2::ylab("Cell type")
plot_cat_1
```

## Per species

```{r prepare_ggdf_category_species}
# pivot longer by species
ggdf_category_long_species <- base::as.data.frame(tidyr::pivot_longer(
  ggdf_category_long,
  cols = c("mmus",
           "mcas",
           "mspr",
           "mcar"), 
  names_to = "species", 
  values_to = "enrichment"))

# remove  unneccessary rows that have been generated by pivot longer
ggdf_category_long_species <- ggdf_category_long_species[
  !is.na(ggdf_category_long_species$enrichment),]

# factorize species
ggdf_category_long_species$species <- factor(
  ggdf_category_long_species$species,
  levels = names(col_spc))
# factorize species
ggdf_category_long_species$species_rev <- factor(
  ggdf_category_long_species$species,
  levels = base::rev(names(col_spc)))
```

```{r plot_category2, fig.width = 5.5, fig.height = 7}
plot_cat2 <- ggplot2::ggplot(ggdf_category_long_species[
  ggdf_category_long_species$cell_type %in% names(col_cts_hsc),],
  aes(fill = conservation_level, y = species_rev))+
  ggplot2::geom_bar(position = "stack")+
  ggplot2::theme_classic()+
  ggplot2::xlab("Nr of detected marker genes")+
  ggplot2::xlim(c(0,450))+
  ggplot2::scale_fill_manual(values = col_mode)+
  ggplot2::facet_grid(rows = vars(cell_type), switch = "y")+
  ggplot2::scale_y_discrete(position = "right")+
  ggplot2::theme(
    legend.position = "top",
    legend.title = element_blank(),
    strip.text.y.left = element_text(angle = 0),
    axis.text.y = element_text(angle = 0),
    axis.title.y = element_blank(),
    strip.text = element_text(size = 14),
    axis.text.x = element_text(size = 12),
    axis.title.x = element_text(size = 12),
    strip.background = element_rect(fill = "grey97", linetype = 0))

plot_cat2
```

```{r, fig.width = 5.5, fig.height = 7}
plot_cat3 <- plot_cat2+
  ggplot2::theme(axis.title.y = element_blank(),
                 axis.text.y = element_blank(),
                 axis.title.x = element_blank(),
                 legend.position = "none") 
plot_cat3
```

```{r plot_category4, fig.width = 5.6, fig.height = 4.5}
plot_cat4 <- ggplot2::ggplot(ggdf_category_long_species[
  ggdf_category_long_species$cell_type %in% names(col_cts_str),],
  aes(fill = conservation_level, y = species_rev))+
  ggplot2::geom_bar(position = "stack")+
  ggplot2::theme_classic()+
  ggplot2::xlab("Nr of detected marker genes")+
  ggplot2::scale_fill_manual(values = col_mode)+
  ggplot2::facet_grid(rows = vars(cell_type), switch = "y")+
  ggplot2::xlim(c(0,450))+
  ggplot2::scale_y_discrete(position = "right")+
  ggplot2::theme(
    legend.position = "none",
    strip.text.y.left = element_text(angle = 0),
    axis.text.y = element_text(angle = 0),
    axis.title.y = element_blank(),
    strip.text = element_text(size = 14),
    axis.text.x = element_text(size = 12),
    axis.title.x = element_text(size = 12),
    strip.background = element_rect(fill = "grey97", linetype = 0))

plot_cat4
```

```{r plot_category5, fig.width = 5.6, fig.height = 4.5}
plot_cat5 <- plot_cat4+
  ggplot2::theme(axis.title.y = element_blank(),
                 axis.text.y = element_blank(),
                 axis.title.x = element_blank(),
                 legend.position = "none") 
plot_cat5
```

## Is it signature or not?

```{r signature_true}
ggdf_issign <- ggdf_category_long_species

ggdf_issign$is_signature <- vector(length = nrow(ggdf_issign))
ggdf_issign$is_signature[
  ggdf_issign$conservation_level == "marker"] <- FALSE
ggdf_issign$is_signature[
  ggdf_issign$conservation_level == "conserved_marker"] <- FALSE
ggdf_issign$is_signature[
  ggdf_issign$conservation_level == "conserved_signature"] <- TRUE

ggdf_issign$is_signature <- factor(ggdf_issign$is_signature, 
                                   levels =c(FALSE, TRUE))
```

```{r plot_signature_true1, fig.width = 6, fig.height = 4}
ggplot2::ggplot(ggdf_issign[ggdf_issign$species == "mmus",],
       aes(fill = is_signature, x = cell_type))+
  ggplot2::geom_bar(position = "stack")+
  ggplot2::theme_classic()+
  ggplot2::ylab("Nr of detected marker genes")+
  ggplot2::scale_fill_manual("Conserved signature", values = col_mode_log)+
  ggplot2::theme(axis.text.x = element_text(angle = 90))+
  ggplot2::ggtitle("BL6 marker genes")+
  ggplot2::xlab("Cell type")

ggplot2::ggplot(ggdf_issign[ggdf_issign$species == "mmus",],
       aes(fill = is_signature, x = cell_type))+
  ggplot2::geom_bar(position = "fill")+
  ggplot2::theme_classic()+
  ggplot2::ylab("Proportion of conserved signature genes")+
  ggplot2::scale_fill_manual("Conserved signature", values = col_mode_log)+
  ggplot2::theme(axis.text.x = element_text(angle = 90))+
  ggplot2::ggtitle("BL6 marker genes")+
  ggplot2::xlab("Cell type")
```

# Signature gene expression

```{r prepare_sces}
agg_hsc <- scuttle::aggregateAcrossCells(
  sce_hsc, 
  id=colData(sce_hsc)[,c("Species_ID")], 
  statistics = "mean",
  use.assay.type = "logcounts") #multibatchnorm

agg_str <- scuttle::aggregateAcrossCells(
  sce_str, 
  id=colData(sce_str)[,c("Species_ID")], 
  statistics = "mean",
  use.assay.type = "logcounts") #multibatchnorm
```

```{r plot_hsc_signatures1, fig.height = 15, fig.width = 4}
cts_hsc <- names(signature_list_hsc)
# activated mpps only have one gene that is not worth to visualise
cts_hsc <- cts_hsc[!cts_hsc == "Activated MPPs"]

for(ct in cts_hsc){
  
  print(ct)
  hsc_genes <- signature_list_hsc[[ct]]$conserved_signature
  
  hsc_genes_nosub <- hsc_genes[!hsc_genes %in%
                                 signature_list_hsc[[ct]]$genes_subclustering]
  
  vis_df <- base::as.data.frame(
    t(SummarizedExperiment::assays(agg_hsc)$logcounts))
  vis_df <- vis_df[,colnames(vis_df) %in% hsc_genes]
  vis_df <- tibble::rownames_to_column(vis_df, "species")
  vis_df <- tidyr::pivot_longer(vis_df, 
                                cols = c(2:ncol(vis_df)),
                                values_to = "expression",
                                names_to = "gene")

  vis_df$species <- factor(vis_df$species, levels = names(col_spc))
  
  # make plots
  plot1 <- ggplot2::ggplot(vis_df, 
                           aes(x = species, 
                               y = reorder(gene, expression), 
                               fill = expression))+
  ggplot2::geom_tile()+
  ggplot2::theme_classic()+
  ggplot2::scale_fill_continuous("average logcounts",
                                 low = "white", high = "blue2",
                                 limits = c(0, max(vis_df$expression)))+
  ggplot2::ggtitle(base::paste(ct, "markers"))+
  ggplot2::theme(axis.title.y = element_blank(),
                 axis.title.x = element_blank(),
                 legend.position = "bottom",
                 axis.text.x = element_text(angle = 90, 
                                            size = 12,
                                            hjust = 0.45),
                 axis.text.y = element_text(size = 12))

  print(plot1)
  
  if(!identical(hsc_genes, hsc_genes_nosub)){
    vis_df <- vis_df[vis_df$gene %in% hsc_genes_nosub,]
    vis_df <- vis_df[order(vis_df$expression),]
    
  plot2 <- ggplot2::ggplot(vis_df, 
                           aes(x = species, 
                               y = reorder(gene, expression), 
                               fill = expression))+
  ggplot2::geom_tile()+
  ggplot2::theme_classic()+
  ggplot2::scale_fill_continuous("average logcounts",
                                 low = "white", high = "blue2",
                                 limits = c(0, max(vis_df$expression)))+
  ggplot2::ggtitle(base::paste(ct, "markers"))+
  ggplot2::theme(axis.title.y = element_blank(),
                 axis.title.x = element_blank(),
                 legend.position = "bottom",
                 axis.text.x = element_text(angle = 90, 
                                            size = 12,
                                            hjust = 0.45),
                 axis.text.y = element_text(size = 12))
  print(plot2)
  }
}
```

```{r plot_str_signatures, fig.height = 18, fig.width = 4}
cts_str <- names(signature_list_str)

for(ct in cts_str){
  
  str_genes <- signature_list_str[[ct]]$conserved_signature
  
  str_genes_nosub <- str_genes[!str_genes %in%
                                 signature_list_str[[ct]]$genes_subclustering]
  
  vis_df <- base::as.data.frame(
    t(SummarizedExperiment::assays(agg_hsc)$logcounts))
  vis_df <- vis_df[,colnames(vis_df) %in% str_genes]
  vis_df <- tibble::rownames_to_column(vis_df, "species")
  vis_df <- tidyr::pivot_longer(vis_df, 
                                cols = c(2:ncol(vis_df)),
                                values_to = "expression",
                                names_to = "gene")

  vis_df$species <- factor(vis_df$species, levels = names(col_spc))
  
  # make plots
  plot1 <- ggplot2::ggplot(vis_df, 
                           aes(x = species, 
                               y = reorder(gene, expression), 
                               fill = expression))+
  ggplot2::geom_tile()+
  ggplot2::theme_classic()+
  ggplot2::scale_fill_continuous("average logcounts",
                                 low = "white", high = "blue2",
                                 limits = c(0, max(vis_df$expression)))+
  ggplot2::ggtitle(base::paste(ct, "markers"))+
  ggplot2::theme(axis.title.y = element_blank(),
                 axis.title.x = element_blank(),
                 legend.position = "bottom",
                 axis.text.x = element_text(angle = 90, 
                                            size = 12,
                                            hjust = 0.45),
                 axis.text.y = element_text(size = 12))

  print(plot1)
  
  if(!identical(str_genes, str_genes_nosub)){
    vis_df <- vis_df[vis_df$gene %in% str_genes_nosub,]
    vis_df <- vis_df[order(vis_df$expression),]
    
  plot2 <- ggplot2::ggplot(vis_df, 
                           aes(x = species, 
                               y = reorder(gene, expression), 
                               fill = expression))+
  ggplot2::geom_tile()+
  ggplot2::theme_classic()+
  ggplot2::scale_fill_continuous("average logcounts",
                                 low = "white", high = "blue2",
                                 limits = c(0, max(vis_df$expression)))+
  ggplot2::ggtitle(base::paste(ct, "markers"))+
  ggplot2::theme(axis.title.y = element_blank(),
                 axis.title.x = element_blank(),
                 legend.position = "bottom",
                 axis.text.x = element_text(angle = 90, 
                                            size = 12,
                                            hjust = 0.45),
                 axis.text.y = element_text(size = 12))
  print(plot2)
  }
}
knitr::knit_exit()

```

## Specific cell types, specific genes

### HSCs

```{r get_hsc_signatures_hsc, fig.width = 4}
hsc_df <- signature_list_hsc$HSCs$conserved_df
hsc_genes <- signature_list_hsc$HSCs$conserved_signature
hsc_genes <- hsc_genes[!hsc_genes %in% 
                         signature_list_hsc$HSCs$genes_subclustering]

print(hsc_df[which(!is.na(hsc_df$mmus) &
                   is.na(hsc_df$mcas) &
                   is.na(hsc_df$mspr) &
                   is.na(hsc_df$mcar)),])

hsc_markers <- hsc_df$gene[which(!is.na(hsc_df$mmus) &
                                   is.na(hsc_df$mcas) &
                                   is.na(hsc_df$mspr) &
                                   is.na(hsc_df$mcar))]

hsc_markers_short <- hsc_markers[
  hsc_markers %in% c("Ly6a", "Procr", "Tie1", "Alcam", "Kit", "Nrgn")
]

vis_df_hsc <- base::as.data.frame(
  t(SummarizedExperiment::assays(agg_hsc)$logcounts))
vis_df_hsc <- vis_df_hsc[,colnames(vis_df_hsc) %in% hsc_markers]
vis_df_hsc <- tibble::rownames_to_column(vis_df_hsc, "species")
vis_df_hsc <- tidyr::pivot_longer(vis_df_hsc, 
                                  cols = c(2:ncol(vis_df_hsc)),
                                  values_to = "expression",
                                  names_to = "gene")

vis_df_hsc$species <- factor(vis_df_hsc$species, levels = names(col_spc))

plot3 <- ggplot2::ggplot(vis_df_hsc, 
                           aes(x = species, 
                               y = reorder(gene, expression), 
                               fill = expression))+
  ggplot2::geom_tile()+
  ggplot2::theme_classic()+
  ggplot2::scale_fill_continuous("average logcounts",
                                 low = "white", high = "blue2",
                                 limits = c(0, max(vis_df_hsc$expression)))+
  ggplot2::ggtitle("BL6-specific HSC markers")+
  ggplot2::theme(axis.title.y = element_blank(),
                 axis.title.x = element_blank(),
                 legend.position = "bottom",
                 axis.text.x = element_text(angle = 90, 
                                            size = 12,
                                            hjust = 0.45),
                 axis.text.y = element_text(size = 12))
plot3

plot4 <- ggplot2::ggplot(vis_df_hsc[vis_df_hsc$gene %in% hsc_markers_short,], 
                           aes(x = species, 
                               y = reorder(gene, expression), 
                               fill = expression))+
  ggplot2::geom_tile()+
  ggplot2::theme_classic()+
  ggplot2::scale_fill_continuous("average logcounts",
                                 low = "white", high = "blue2")+
  ggplot2::ggtitle("BL6-specific HSC markers")+
  ggplot2::theme(axis.title.y = element_blank(),
                 axis.title.x = element_blank(),
                 legend.position = "bottom",
                 axis.text.x = element_text(angle = 90, 
                                            size = 12,
                                            hjust = 0.45),
                 axis.text.y = element_text(size = 12))
plot4
```

```{r sessioninfo}
utils::sessionInfo()
```