---
title: "Differentially mapped gene report per sample"
author: "Lea WÃ¶lbert"
date: '2023-03-31'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Report QC for all samples separately (Object_ID).
Check if both mappings result in similar quality.

#### Load libraries, source code 

```{r seed,  message = FALSE}
set.seed(37)
```

```{r libraries,  message = FALSE}
library(scuttle, quietly = TRUE) 
library(scater, quietly = TRUE) 
library(scran, quietly = TRUE) 
library(pheatmap, quietly = TRUE) 
```

```{r source, message = FALSE}
source(file = snakemake@params[["plotting"]])
```

```{r load_objects, message = FALSE}
sce_og <- readRDS(file = snakemake@input[["sce_input"]])
sce_fg <- readRDS(file = snakemake@input[["sce_fg"]])

dmg_list <- readRDS(file = snakemake@input[["dmg_list"]])
dmgs <- readRDS(file = snakemake@input[["dmgs"]])

ensembl_list_mspr <- readRDS(snakemake@input[["ensembl_list_mspr"]])

name_curr <- colData(sce_og)$Object_ID[1] 
nr_hvgs <- snakemake@params[["nr_hvgs"]]
species_curr <- snakemake@wildcards[["species"]]
```

```{r susbet}
# subset fg to sce_og
sce_fg <- sce_fg[,which(!is.na(match(colnames(sce_fg), colnames(sce_og))))]
sce_og
sce_fg
```

MSPR objects don't have symbols.
This part was taken from Perrine's script on how to download the relevant data 
from BioMart.

```{r mspr_ensembl}

if(grepl("mspr", name_curr)){
  
  intersect_IDs <- intersect(
    rowData(sce_fg)$Symbol, ensembl_list_mspr$ensembl_gene_id)

  ensembl_list_mspr <- ensembl_list_mspr[
    ensembl_list_mspr$ensembl_gene_id %in% intersect_IDs,]
  sce_fg <- sce_fg[rowData(sce_fg)$Symbol %in% intersect_IDs,]

  ensembl_list_mspr <- ensembl_list_mspr[
    !duplicated(ensembl_list_mspr$ensembl_gene_id),]
  sce_fg <- sce_fg[!duplicated(rowData(sce_fg)$Symbol),]

  rowData(sce_fg)$Symbol[
    match(
      ensembl_list_mspr$ensembl_gene_id,
      rownames(sce_fg))] <- ensembl_list_mspr$mmusculus_homolog_associated_gene_name

  rowData(sce_fg)$Symbol[
    rowData(sce_fg)$Symbol == ""] <- rowData(sce_fg)$ID[
      rowData(sce_fg)$Symbol == ""]
}
```

Prepare combination of both SCE.
RowData first, then colData.

```{r prep_comb_row}

head(rowData(sce_og))
head(rowData(sce_fg))

shared_genes <- intersect(rowData(sce_og)$Symbol, rowData(sce_fg)$Symbol)
length(shared_genes)

sce_og_shared <- sce_og[rowData(sce_og)$Symbol %in% shared_genes,]
sce_fg_shared <- sce_fg[rowData(sce_fg)$Symbol %in% shared_genes,]

sce_og_shared <- sce_og_shared[order(rowData(sce_og_shared)$Symbol),]
sce_fg_shared <- sce_fg_shared[order(rowData(sce_fg_shared)$Symbol),]

sce_og_shared <- sce_og_shared[!duplicated(rowData(sce_og_shared)$Symbol),]
sce_fg_shared <- sce_fg_shared[!duplicated(rowData(sce_fg_shared)$Symbol),]
```

```{r prep_comp_col}

colnames(sce_og_shared) <- paste0(colnames(sce_og_shared), "_og")
sce_og_shared$Genome <- rep("OneGenome", ncol(sce_og_shared))

colnames(sce_fg_shared) <- paste0(colnames(sce_fg_shared), "_fg")
sce_fg_shared$Genome <- rep("FourGenomes", ncol(sce_fg_shared))

colnames(rowData(sce_og_shared))[
  colnames(rowData(sce_og_shared)) == "ID"] <- "ID_og"
colnames(rowData(sce_fg_shared))[
  colnames(rowData(sce_fg_shared)) == "ID"] <- "ID_fg"

rownames(sce_og_shared) <- rowData(sce_og_shared)$Symbol 
rownames(sce_fg_shared) <- rowData(sce_fg_shared)$Symbol 

colData(sce_og_shared) <- colData(sce_og_shared)[,-20]
```

```{r combine}
# combine
sce_tog <- cbind(sce_og_shared, sce_fg_shared)
sce_tog
```

## General quality

```{r qc_metrics}
qcdf_og <- perCellQCMetrics(sce_og_shared)
qcdf_fg <- perCellQCMetrics(sce_fg_shared)
```

```{r qc_metrics_plot1}

qcdf_gg <- as.data.frame(qcdf_og)
qcdf_gg <- pivot_longer(qcdf_gg, 
                        names_to = "metric", cols = c(1, 2, 3))

ggplot(qcdf_gg[qcdf_gg$metric == "sum",], aes(y = value, x = metric))+
  theme_classic()+
  ggbeeswarm::geom_quasirandom(size = 0.1)+
  geom_boxplot(alpha = 0)+
  scale_color_manual("Removed", 
                     values = c("FALSE" = "grey80", "TRUE" = "orange"))+
  ylab("Library size")+
  ggtitle("One genome")+
  theme(axis.title.x = element_blank(), axis.text.x = element_blank())
```

```{r qc_metrics_plot2}

qcdf_gg <- as.data.frame(qcdf_fg)
qcdf_gg <- pivot_longer(qcdf_gg, 
                        names_to = "metric", cols = c(1, 2, 3))

ggplot(qcdf_gg[qcdf_gg$metric == "sum",], aes(y = value, x = metric))+
  theme_classic()+
  ggbeeswarm::geom_quasirandom(size = 0.1)+
  geom_boxplot(alpha = 0)+
  scale_color_manual("Removed", 
                     values = c("FALSE" = "grey80", "TRUE" = "orange"))+
  ylab("Library size")+
  ggtitle("Four genomes")+
  theme(axis.title.x = element_blank(), axis.text.x = element_blank())
```

## Markergenes

Calculate logcounts together for visualisation.
DMGs = differentially mapped genes.

```{r marker_genes}
quick_clust <- scran::quickCluster(sce_tog) # to accelerate calculation
sce_tog <- scran::computeSumFactors(sce_tog, cluster = quick_clust)
sce_tog <- scuttle::logNormCounts(sce_tog) # log-transformation
```

### Sample heatmap

```{r sample_heatmap}

sce_tog_temp <- sce_tog[rowSums(assays(sce_tog)$counts) > 5,]
# randomly subset to 2000 cells if necessary 
# first half and second half of object should be subset identically
if(ncol(sce_tog_temp) > 2000){
  random_cols <- sample(1:(ncol(sce_tog_temp)/2), 1000)
  random_cols_firsthalf <- random_cols
  random_cols_secondhalf <- random_cols + (ncol(sce_tog_temp)/2)
  sce_tog_temp <- sce_tog_temp[,c(random_cols_firsthalf, random_cols_secondhalf)]
}

mat <- as.matrix(
  assays(sce_tog_temp)$logcounts
)

anno_col <- data.frame(
  features = colnames(sce_tog_temp),
  Genome = colData(sce_tog_temp)$Genome)

anno_col <- column_to_rownames(anno_col, var = "features")

pheatmap::pheatmap(
  mat,
  cluster_cols = FALSE,
  treeheight_row = 0.1,  
  annotation_col = anno_col,
  show_colnames = FALSE,
  show_rownames = FALSE
)
```

### Sample-specific DMGs:

```{r subset_dmgs}
dmgs[,c(2, 3, 5)]
if(nrow(dmgs) > 200){dmgs <- dmgs[1:200,]}
```

```{r heatmap, fig.height=14, fig.width=7}

sce_tog_temp <- sce_tog[which(rownames(sce_tog) %in% rownames(dmgs)),]

mat <- as.matrix(
  assays(sce_tog_temp)$logcounts
)

anno_col <- data.frame(
  features = colnames(sce_tog_temp),
  Genome = colData(sce_tog_temp)$Genome)

anno_col <- column_to_rownames(anno_col, var = "features")

pheatmap::pheatmap(
  mat,
  cluster_cols = FALSE,
  annotation_col = anno_col,
  show_colnames = FALSE
)
```

### All combined DMGs

```{r all_dmgs}
print(dmg_list)
```

```{r heatmap_all}

sce_tog_temp <- sce_tog[which(rownames(sce_tog) %in% dmg_list),]

mat <- as.matrix(
  assays(sce_tog_temp)$logcounts
)

anno_col <- data.frame(
  features = colnames(sce_tog_temp),
  Genome = colData(sce_tog_temp)$Genome)

anno_col <- column_to_rownames(anno_col, var = "features")

pheatmap::pheatmap(
  mat,
  treeheight_row = 0.1,
  cluster_cols = FALSE,
  cluster_rows = FALSE,
  annotation_col = anno_col,
  show_colnames = FALSE,
  show_rownames = FALSE
)
```

```{r session_info}
sessionInfo()
```