---
title: "Figure 2"
date: '2024-09-04'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Generate Figure 2, current title:
"Cell type conserved signature genes are robust across mouse species and age."

#### Load objects

```{r seed, message = FALSE}
base::RNGkind(kind = "Mersenne-Twister")
set.seed(37)
```

```{r load_packages, message = FALSE}
library(tidyverse, quietly = TRUE)
library(SingleCellExperiment, quietly = TRUE)
library(ggpubr, quietly = TRUE)
library(scuttle, quietly = TRUE)
```

```{r load_manual, eval = FALSE, include = FALSE}

# load data for manual use here

base_path <- "/omics/odcf/analysis/OE0538_projects/DO-0008/data/"

sce_hsc <- base::readRDS(base::paste0(
  base_path, 
  "/scRNAseq/main_analysis/sce_objects/02_sce_anno/10_anns/sce_hsc-10"))

sce_hsc_all_pseudotime <- base::readRDS(base::paste0(
  base_path, 
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/05_cellrank/04_psce/sce_hsc_pseudotime"))

# list of genesets (conserved signature and others)
geneset_list_hsc <- base::readRDS(base::paste0(
  base_path, 
  "scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/01_reclustering_own/01_gens/geneset_list_hsc"))
  
colors_path <- base::paste0(
  base_path, 
  "/metadata/colors/colors.txt")

source("../source/colors.R")
```

```{r load_snakemake}

# data during snakemake run is loaded here
# @Veronica I can add all paths from you here

# annotated SCE objects
sce_hsc <- base::readRDS(snakemake@input[["sce_hsc"]])
sce_hsc_all_pseudotime <- base::readRDS(snakemake@input[["sce_pt"]])

# gene list
geneset_list_hsc <- base::readRDS(snakemake@input[["geneset_list_hsc"]])

# colors
colors_path <- snakemake@params[["colors_path"]]
colors_source <- snakemake@params[["colors"]]
source(colors_source)

#knitr::knit_exit()

```

```{r params}
# global params
source("determine_params.R")

# specific params
# TODO: put in colors later
colors_z_score <- c("blue3", "white", "red3")

celltype <- "HSCs"
```

## A

Model of methods used to extract conserved signature genes for each cell type. 
Conserved signature genes are non-differentially expressed conserved markers
within the same cell type across all species.

@Lea

## B

Expression of conserved signature genes in all mouse species across HSPC
cell types (scaled by z-score per gene).

@Lea

TODO: CHECK IF WITH OR WITHOUT SUBCLUSTERING GENES!!

#### Plot Components

```{r b_prep1}
# get all unique signature genes
all_sign_genes <- vector()
for(i in 1:length(geneset_list_hsc)){
  sign_genes <- geneset_list_hsc[[i]]$conserved_signature
  all_sign_genes <- c(all_sign_genes, sign_genes)
}
all_sign_genes_hsc <- base::unique(all_sign_genes)
```

```{r b_prep2}
# aggregate per cell type and species
cts_hsc <- as.list(names(geneset_list_hsc))

# separate SCE objects by cell type, then aggregate the counts to obtain
# the average counts per species for each cell type separately
# use the multi-batchnorm-corrected logcounts 

agg_hsc_list <- lapply(cts_hsc, function(ct){
  
  sce_ct <- sce_hsc[,sce_hsc$celltypes == ct]
  agg_hsc <- scuttle::aggregateAcrossCells(
    sce_ct, 
    id=colData(sce_ct)[,c("Species_ID")], 
    statistics = "mean",
    use.assay.type = "logcounts") #multibatchnorm
  
  return(agg_hsc)
})
names(agg_hsc_list) <- names(geneset_list_hsc)
```

```{r b_prep3}
expr_df_list_hsc <- lapply(agg_hsc_list, function(agg_hsc){
  
  # get current cell type and relevant signature genes
  ct_curr <- agg_hsc$celltypes[1]
  sign_genes_ct <- geneset_list_hsc[[ct_curr]]$conserved_signature
  
  # get a df from average logcounts per ct and species
  vis_df <- base::as.data.frame(
    t(SummarizedExperiment::assays(agg_hsc)$logcounts))
  
  # subset to all signature genes of this tissue
  vis_df <- vis_df[,colnames(vis_df) %in% all_sign_genes_hsc]
  
  # change df format 
  vis_df <- tibble::rownames_to_column(vis_df, "species")
  vis_df <- tidyr::pivot_longer(vis_df, 
                                cols = c(2:ncol(vis_df)),
                                values_to = "expression",
                                names_to = "gene")
  
  # add ct info
  vis_df$cell_type <- base::rep(ct_curr)

  # add info on genes that are signature genes of this ct
  vis_df$is_sign_gene <- vector(length = nrow(vis_df))
  vis_df$is_sign_gene[vis_df$gene %in% sign_genes_ct] <- TRUE

  # scale expression of all genes per cell type 
  # for ordering of genes ONLY
  vis_df$scaled_expression_per_ct <- base::scale(vis_df$expression)
  vis_df <- vis_df[base::order(vis_df$scaled_expression_per_ct, 
                               decreasing = TRUE),]

  return(vis_df)
})

# bind dfs 
agg_hsc_df <- dplyr::bind_rows(expr_df_list_hsc)
```

```{r b_prep4}
# scale expression per gene using z-scores
agg_hsc_df$zscore_gene <- vector(length = nrow(agg_hsc_df), mode = "numeric")
for(g in base::unique(agg_hsc_df$gene)){
  
  expr_vector_g <- agg_hsc_df[agg_hsc_df$gene == g,]$expression

  # calculate z score per gene
  z_score_g <- (expr_vector_g - base::mean(expr_vector_g))/
    stats::sd(expr_vector_g)
  agg_hsc_df$zscore_gene[agg_hsc_df$gene == g] <- z_score_g
  
}

```

```{r b_prep5}
# get a well ordered list of unique genes for factorization (after order above)
all_sign_genes_for_factor <- base::unique(agg_hsc_df$gene[
  agg_hsc_df$is_sign_gene == TRUE])

# factorize
agg_hsc_df$gene <- factor(agg_hsc_df$gene,
                          levels = base::rev(all_sign_genes_for_factor))
agg_hsc_df$species <- factor(agg_hsc_df$species,
                             levels = names(col_spc))
```

```{r b_base_plot}
b_base_plot <- ggplot2::ggplot(
  agg_hsc_df,
  aes(x = species, y = gene, fill = zscore_gene))+
  ggplot2::geom_tile()+
  ggplot2::facet_grid(cols = vars(agg_hsc_df$cell_type))
```

```{r b_theme_plot}
b_theme_plot <- b_base_plot+
  ggplot2::theme_classic()+
  ggplot2::theme(
    axis.text.y = element_blank(),
    axis.title.y = element_text(
      face = axis_title_face,
      size = axis_title_size,
      color = axis_title_color),
    axis.title.x = element_blank(),
    axis.text.x = element_text(
      angle = 90,
      face = axis_text_face,
      size = axis_text_size,
      color = axis_text_color,
      vjust = 0.5,
      hjust = 0),
    strip.text.x = element_text(
      angle = 90,
      face = axis_text_face,
      size = axis_text_size,
      color = axis_text_color,
      vjust = 0.5,
      hjust = 0),
    strip.background.x = element_rect(
      color = strip_background_color,
      fill = strip_background_fill),
    legend.title = element_text(
      face = legend_title_face,
      size = legend_title_size,
      color = legend_title_color),
    legend.text = element_text(
      face = legend_text_face,
      size = legend_text_size,
      color = legend_text_color),
    axis.ticks.y = element_blank())+
  ggplot2::ylab("Conserved signature genes")+
  ggplot2::scale_fill_gradientn(
    "z-score per gene", 
    colours = colors_z_score,
    breaks = c(-5, -2.5,  0, 2.5, 5), 
    limits = c(-5,5))

b_legend <- ggpubr::get_legend(b_theme_plot)
```

```{r b_rm}
#rm(agg_hsc_list)
```

### Construct

```{r b_construct, fig.width = 8.5, fig.height = 6}
b_plot_theme <- ggpubr::ggarrange(
  b_theme_plot + theme(legend.position = "none")
)
b_plot_theme
```

```{r b_construct_legend, fig.width = 4, fig.height = 4}
b_legend_theme <- ggpubr::ggarrange(
  b_legend 
)
b_legend_theme
```

## C 

HSPCs in UMAP space, colored by diffusion pseudotime.

@Lea

#### Plot components

```{r c_prep}

# transfer pseudotime
sce_hsc$pseudotime <- sce_hsc_all_pseudotime$dpt_pseudotime

hsc_umap <- base::as.data.frame(
  SingleCellExperiment::reducedDim(
    sce_hsc, type = "UMAP"))
hsc_umap$color_by <- sce_hsc$pseudotime
hsc_umap <- hsc_umap[base::sample(
  1:nrow(hsc_umap),
  nrow(hsc_umap),
  replace = FALSE),]
```

```{r c_base_plot}
c_base_plot <- ggplot2::ggplot(
  hsc_umap, 
  aes(x = X1,
      y = X2,
      color = color_by))+
  ggplot2::geom_point(
    size = umap_point_size,
    alpha = umap_point_alpha)+
  ggplot2::xlim(c(base::min(hsc_umap$X1)-1, base::max(hsc_umap$X1))+1)+
  ggplot2::ylim(c(base::min(hsc_umap$X2)-1, base::max(hsc_umap$X2))+1)
```

```{r c_theme_plot}
c_theme_plot <- c_base_plot+
  cowplot::theme_nothing()+
  ggplot2::scale_color_gradientn(
    "Pseudotime",
    colors = mycolors_to1
  )
```

```{r c_legend_base_plot}
c_legend_base_plot <- ggplot2::ggplot(
  hsc_umap, 
  aes(x = X1,
      y = X2,
      color = color_by))+
  ggplot2::geom_point()
```

```{r c_legend_theme_plot}
c_legend_theme_plot <- c_legend_base_plot+
  ggplot2::scale_color_gradientn(
    "Pseudotime",
    colors = mycolors_to1,
    breaks = c(0, 0.5, 1))+
  ggplot2::theme(
    legend.title = element_text(
      size = legend_title_size,
      face = legend_title_face,
      color = legend_title_color,
      margin = margin(r = 10),
      vjust = 0.85),
    legend.text = element_text(
      size = legend_text_size,
      face = legend_text_face,
      color = legend_text_color),
    legend.box.just = "left",
    legend.position = "bottom")
      #legend.box = "vertical", # Stack legends vertically

c_legend <- ggpubr::get_legend(c_legend_theme_plot)
```

#### Construct

```{r c_construct, fig.width = 3.7, fig.height = 3.7}
c_plot_theme <- ggpubr::ggarrange(
  c_theme_plot
)
c_plot_theme
```

```{r c_construct_legend, fig.width = 3, fig.height = 1}
c_legend_theme <- ggpubr::ggarrange(
  c_legend
)
c_legend_theme
```

## D

Min-max scaled expression of HSC (top), Mk prog. (middle) or Erythroid prog.
(bottom) conserved signature genes along diffusion pseudotime trajectory within 
the erythroid branch. Cells were partitioned into branches based on their 
lineage differentiation probability calculated by cellrank.

@Veronica

## E 

Scaled expression of species-specific marker genes in all mouse species within
HSCs (scaled by z-score per gene).

@Lea

```{r e_prep}
# get ct specific data
hsc_cons_df_ct <- geneset_list_hsc$HSCs$conserved_df

hsc_mmus_specific <- hsc_cons_df_ct$gene[which(
  !is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]

hsc_mcas_specific <- hsc_cons_df_ct$gene[which(
  is.na(hsc_cons_df_ct$mmus) &
    !is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]

hsc_mspr_specific <- hsc_cons_df_ct$gene[which(
  is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    !is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]

hsc_mcar_specific <- hsc_cons_df_ct$gene[which(
  is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    !is.na(hsc_cons_df_ct$mcar))]
  
hsc_genes <- c(hsc_mmus_specific,
               hsc_mcas_specific,
               hsc_mspr_specific,
               hsc_mcar_specific)

stopifnot(!duplicated(hsc_genes))
```

```{r e_prep2}
agg_hsc_ct <- agg_hsc_list[[celltype]]

# make visualisation DF
vis_df_hsc <- as.data.frame(
  t(SummarizedExperiment::assays(agg_hsc_ct)$logcounts)) # batchnorm, aggregated
vis_df_hsc <- vis_df_hsc[,colnames(vis_df_hsc) %in% hsc_genes]
vis_df_hsc <- tibble::rownames_to_column(vis_df_hsc, "species")
vis_df_hsc <- tidyr::pivot_longer(
  vis_df_hsc, 
  cols = c(2:ncol(vis_df_hsc)),
  values_to = "expression",
  names_to = "gene")

vis_df_hsc$species <- factor(vis_df_hsc$species, levels = names(col_spc))
vis_df_hsc$gene <- factor(vis_df_hsc$gene, levels = base::rev(hsc_genes))

vis_df_hsc$zscore_gene <- vector(length = nrow(vis_df_hsc), mode = "numeric")

for(g in base::unique(vis_df_hsc$gene)){
  
  expr_vector_g <- vis_df_hsc[vis_df_hsc$gene == g,]$expression

  # calculate z score per gene
  z_score_g <- (expr_vector_g - base::mean(expr_vector_g))/stats::sd(expr_vector_g)
  vis_df_hsc$zscore_gene[vis_df_hsc$gene == g] <- z_score_g
  
}
```

```{r e_base_plot}
e_base_plot <- ggplot2::ggplot(
  vis_df_hsc, 
  aes(x = species, 
      y = gene, 
      fill = zscore_gene))+
  ggplot2::geom_tile()
```

```{r e_theme_plot}
e_theme_plot <- e_base_plot +
  ggplot2::theme_classic()+
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_text(
      face = axis_title_face,
      size = axis_title_size,
      color = axis_title_color),
    axis.ticks.y = element_blank(),
    axis.text.y = element_blank(),
    axis.text.x = element_text(
      angle = 90,
      face = axis_text_face,
      size = axis_text_size,
      color = axis_text_color,
      vjust = 0.5,
      hjust = 0),
    legend.title = element_text(
      face = legend_title_face,
      size = legend_title_size,
      color = legend_title_color),
    legend.text = element_text(
      face = legend_text_face,
      size = legend_text_size,
      color = legend_text_color),
)+
  ggplot2::ylab("Species-specific marker genes")+
  ggplot2::scale_fill_gradientn(
    "z-score per gene", 
    colours = colors_z_score,
    breaks = c(-1.5, 0, 1.5), 
    #breaks = c(-5, -2.5, 0, 2.5, 5), 
    #limits = c(-5,5)
    limits = c(-2,2)
    )+
  scale_x_discrete(expand = c(0, 0))

e_legend <- ggpubr::get_legend(e_theme_plot)
```

#### Construct

```{r e_construct, fig.width = 3, fig.height = 4}
e_plot_theme <- ggpubr::ggarrange(
  e_theme_plot+theme(legend.position = "none")
)
e_plot_theme
```

```{r e_construct_legend, fig.width = 2, fig.height = 2}
e_legend_theme <- ggpubr::ggarrange(
  e_legend
)
e_legend_theme
```

## F 

Number of marker genes per HSPC cell type identified when all, only old, 
or only young mouse samples are included.

@Veronica

## G

G: Scaled expression of signature genes [geneX], [geneY], and [geneZ] across
mouse species and ages (z-score per species and age).

@Veronica