---
title: "Figure 2"
date: '2024-09-04'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Generate Figure 2, current title:
"Cell type conserved signature genes are robust across mouse species and age."

Because Figure 2 requires so much data, it is ordered by how big each chunk of 
data is.

- Expression Heatmap signature genes 
- Expression Heatmap species-specific genes 
- Expression heatmaps condition 
- Number of marker genes 

Anything using erythroid or pseudotime data comes later:
Relevant data is also loaded later to save space before that.

- Pseudotime and Erythroid Branch UMAPs
- Pseudotime ranking and scaling 

#### Load objects

```{r seed, message = FALSE}
RNGkind("L'Ecuyer-CMRG") 
set.seed(37)
```

```{r load_packages, message = FALSE}
library(tidyverse, quietly = TRUE)
library(SingleCellExperiment, quietly = TRUE)
library(ggpubr, quietly = TRUE)
library(scuttle, quietly = TRUE)
library(Seurat, quietly = TRUE)
library(SeuratObject, quietly = TRUE) 
library(ggpattern, quietly = TRUE) 
```

```{r base_path}
base_path <- "/omics/odcf/analysis/OE0538_projects/DO-0008/data"
```

```{r load_manual}
# fully annotated SCE objects
sce_hsc <- base::readRDS(base::paste0(
  base_path, 
  "/scRNAseq/main_analysis/sce_objects/02_sce_anno/10_anns/sce_hsc-10"))

# list of gene sets for HSPCs (signature genes, conserved markers, etc.)
geneset_list_hsc <- base::readRDS(base::paste0(
  base_path, 
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/01_reclustering_own/01_gens/geneset_list_hsc"))

#conserved markers per cell type per age (markers_conservation_hsc)
base::load(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/03_marker_conservation/02_age/cons_markers_hsc.RData"))

colors_path <- base::paste0(
  base_path, 
  "/metadata/colors/colors.txt")

source("../source/colors.R")
```

```{r params}
# global params
source("determine_params.R")

# specific params

pattern_density_val <- 0.015
pattern_spacing_val <- 0.03

celltype <- "HSC"

colors_z_score_ct <- c("steelblue3", "white", "red3")
colors_z_score_cond <- c("purple3", "white", "orange")
```

```{r export_paths}

output_pdf_pt_umap3_plot <- base::paste0(
  base_path, "/manuscript1/figure2/figure2_pt_umap3.pdf")

output_pdf_pt_umap_legend3 <- base::paste0(
  base_path, "/manuscript1/figure2/figure2_pt_umap_legend_3.pdf")

output_pdf_sign_heatmap_plot <- base::paste0(
  base_path, "/manuscript1/figure2/figure2_sign_heatmap_plot.pdf")
output_pdf_sign_heatmap_legend <- base::paste0(
  base_path, "/manuscript1/figure2/figure2_sign_heatmap_legend.pdf")

output_pdf_heatmap_spec_plot <- base::paste0(
  base_path, "/manuscript1/figure2/figure2_heatmap_spec_plot.pdf")

output_pdf_pt_scaled <- base::paste0(
  base_path, "/manuscript1/figure2/figure2_pt_scaled_plot.pdf")
output_pdf_heatmap_conds_stacked_plot <- base::paste0(
  base_path, "/manuscript1/figure2/figure2_cond_plot.pdf")
output_pdf_heatmap_conds_stacked_legend <- base::paste0(
  base_path, "/manuscript1/figure2/figure2_cond_legend.pdf")

output_pdf_aging_theme_plot <- base::paste0(
  base_path, "/manuscript1/figure2/figure2_aging_plot.pdf")
output_pdf_aging_theme_legend <- base::paste0(
  base_path, "/manuscript1/figure2/figure2_aging_legend.pdf")
```

## Obtain data

```{r obtain_data}
# nr of detected transcribed genes
# remove lots of unnecessary stuff to avoid overloading the environment
sce_hsc_met <- sce_hsc
rowData(sce_hsc_met) <- rowData(sce_hsc_met)[,c(2, 3, 4)]

print(length(which(rowSums(assays(sce_hsc)$counts_before_BC) > 0)))

rm(sce_hsc_met)
gc()

# nr of conserved signature genes identified

conserved_signature_genes <- lapply(geneset_list_hsc, function(geneset_list){
  conserved_signature_genes <- geneset_list$conserved_signature
  return(conserved_signature_genes)
})
all_conserved_signature_genes <- unique(unlist(conserved_signature_genes))
length(all_conserved_signature_genes)

conserved_signature_genes$HSC
```

## Expression heatmap signature

Expression of signature genes in all mouse species across HSPC
cell types (scaled by z-score per gene).
For Figure 2b.

#### PREP

```{r sign_heatmap0}
sce_hsc$species_pub <- vector(length = ncol(sce_hsc))
sce_hsc$species_pub[sce_hsc$Species_ID == "mmus"] <- "BL6"
sce_hsc$species_pub[sce_hsc$Species_ID == "mcas"] <- "CAST"
sce_hsc$species_pub[sce_hsc$Species_ID == "mspr"] <- "SPRET"
sce_hsc$species_pub[sce_hsc$Species_ID == "mcar"] <- "CAROLI"
sce_hsc$species_pub <- factor(sce_hsc$species_pub, 
                              levels = names(col_spc_pub))
```

```{r sign_heatmap1}
# get all unique signature genes
all_sign_genes <- vector()
for(i in 1:length(geneset_list_hsc)){
  sign_genes <- geneset_list_hsc[[i]]$conserved_signature
  all_sign_genes <- c(all_sign_genes, sign_genes)
}
all_sign_genes_hsc <- base::unique(all_sign_genes)
```

```{r sign_heatmap2}
# aggregate per cell type and species
cts_hsc <- as.list(names(geneset_list_hsc))

# separate SCE objects by cell type, then aggregate the counts to obtain
# the average counts per species for each cell type separately
# use the multi-batchnorm-corrected logcounts 

agg_hsc_list <- lapply(cts_hsc, function(ct){
  
  sce_ct <- sce_hsc[,sce_hsc$celltypes == ct]
  agg_hsc <- scuttle::aggregateAcrossCells(
    sce_ct, 
    id=colData(sce_ct)[,c("species_pub")], 
    statistics = "mean",
    use.assay.type = "logcounts") #multibatchnorm
  
  return(agg_hsc)
})
names(agg_hsc_list) <- names(geneset_list_hsc)
```

```{r sign_heatmap3}
expr_df_list_hsc <- lapply(agg_hsc_list, function(agg_hsc){
  
  # get current cell type and relevant signature genes
  ct_curr <- agg_hsc$celltypes[1]
  sign_genes_ct <- geneset_list_hsc[[ct_curr]]$conserved_signature
  
  # get a df from average logcounts per ct and species
  vis_df <- base::as.data.frame(
    t(SummarizedExperiment::assays(agg_hsc)$logcounts))
  
  # subset to all signature genes of this tissue
  vis_df <- vis_df[,colnames(vis_df) %in% all_sign_genes_hsc]
  
  # change df format 
  vis_df <- tibble::rownames_to_column(vis_df, "species")
  vis_df <- tidyr::pivot_longer(vis_df, 
                                cols = c(2:ncol(vis_df)),
                                values_to = "expression",
                                names_to = "gene")
  
  # add ct info
  vis_df$cell_type <- base::rep(ct_curr)

  # add info on genes that are signature genes of this ct
  vis_df$is_sign_gene <- vector(length = nrow(vis_df))
  vis_df$is_sign_gene[vis_df$gene %in% sign_genes_ct] <- TRUE

  # scale expression of all genes per cell type 
  # for ordering of genes ONLY
  vis_df$scaled_expression_per_ct <- base::scale(vis_df$expression)
  vis_df <- vis_df[base::order(vis_df$scaled_expression_per_ct, 
                               decreasing = TRUE),]

  return(vis_df)
})

# bind dfs 
agg_hsc_df <- dplyr::bind_rows(expr_df_list_hsc)
```

```{r sign_heatmap4}
# scale expression per gene using z-scores
agg_hsc_df$zscore_gene <- vector(length = nrow(agg_hsc_df), mode = "numeric")
for(g in base::unique(agg_hsc_df$gene)){
  
  expr_vector_g <- agg_hsc_df[agg_hsc_df$gene == g,]$expression

  # calculate z score per gene
  z_score_g <- (expr_vector_g - base::mean(expr_vector_g))/
    stats::sd(expr_vector_g)
  agg_hsc_df$zscore_gene[agg_hsc_df$gene == g] <- z_score_g
}
```

```{r sign_heatmap5}
# get a well ordered list of unique genes for factorization (after order above)
all_sign_genes_for_factor <- base::unique(agg_hsc_df$gene[
  agg_hsc_df$is_sign_gene == TRUE])

# factorize
agg_hsc_df$gene <- factor(agg_hsc_df$gene,
                          levels = all_sign_genes_for_factor)
agg_hsc_df$species <- factor(agg_hsc_df$species,
                             levels = base::rev(names(col_spc_pub)))
```

```{r sign_heatmap6}
interest_genes <- c("Msi2",
                    "Angpt1",
                    "Cd34",
                    "Mpo",
                    "Tcf4", 
                    "Itga2b",
                    "Klf1",
                    "Ccnb2")

agg_hsc_df$is_gene <- vector(length = nrow(agg_hsc_df))
agg_hsc_df$is_gene[agg_hsc_df$gene %in% interest_genes] <- "TRUE"
agg_hsc_df$is_gene <- as.logical(agg_hsc_df$is_gene)
```

#### PLOT

```{r sign_heatmap_base_plot}
sign_heatmap_base_plot <- ggplot2::ggplot(
  agg_hsc_df,
  aes(y = species, x = gene, fill = zscore_gene))+
  ggplot2::geom_tile()+
  ggplot2::facet_grid(rows = vars(agg_hsc_df$cell_type))
```

```{r sign_heatmap_theme_plot}
sign_heatmap_theme_plot <- sign_heatmap_base_plot+
  theme_all+
  ggplot2::theme(
    legend.position = "bottom",
    legend.title.position = "top",
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    plot.title = element_text(
      vjust = 0.5,
      hjust = 0.5
      ),
    axis.title.y = element_blank(),
    axis.line.x = element_blank(),
    axis.text.y = element_text(
      hjust = 0),
    strip.text.y = element_text(
      size = axis_text_size,
      face = axis_text_face,
      color = axis_text_color,
      angle = 0,
      vjust = 0.5,
      hjust = 0),
    strip.background.y = element_blank())+
  ggplot2::ggtitle("Signature genes")+
  ggplot2::scale_fill_gradientn(
    "z-score per gene", 
    colours = colors_z_score_ct,
    breaks = c(-5, 0, 5), 
    limits = c(-5,5))+
  ggplot2::scale_x_discrete(
    position =  "bottom",
    breaks = agg_hsc_df$gene[agg_hsc_df$is_gene])

sign_heatmap_legend_theme <- ggpubr::get_legend(sign_heatmap_theme_plot)
```

#### Assemble

```{r sign_heatmap_assemble, fig.width = 6.5, fig.height = 8}
sign_heatmap_plot <- ggpubr::ggarrange(
  sign_heatmap_theme_plot + ggplot2::theme(legend.position = "none")
)
sign_heatmap_plot
```

#### LEGEND

```{r sign_heatmap_legend, fig.width = 4, fig.height = 4}
sign_heatmap_legend <- ggpubr::ggarrange(sign_heatmap_legend_theme)
```

#### EXPORT

```{r output_pdf_sign_heatmap_plot}
pdf(output_pdf_sign_heatmap_plot, width = 6.5, height = 8)
sign_heatmap_plot
dev.off()
```

```{r output_sign_heatmap_legend}
pdf(output_pdf_sign_heatmap_legend, width = 4, height = 2)
sign_heatmap_legend
dev.off()
```

```{r sign_heatmap_remove}
rm(expr_df_list_hsc)
gc()
```

## Heatmap species-specific marker genes

Scaled expression of species-specific marker genes in all mouse species within
HSCs (scaled by z-score per gene).
For Figure 2c.

#### PREP

```{r heatmap_spec1}
# get ct specific data
hsc_cons_df_ct <- geneset_list_hsc$HSC$conserved_df

hsc_mmus_specific <- hsc_cons_df_ct$gene[which(
  !is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]

hsc_mcas_specific <- hsc_cons_df_ct$gene[which(
  is.na(hsc_cons_df_ct$mmus) &
    !is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]

hsc_mspr_specific <- hsc_cons_df_ct$gene[which(
  is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    !is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]

hsc_mcar_specific <- hsc_cons_df_ct$gene[which(
  is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    !is.na(hsc_cons_df_ct$mcar))]
  
hsc_genes <- c(hsc_mmus_specific,
               hsc_mcas_specific,
               hsc_mspr_specific,
               hsc_mcar_specific)

stopifnot(!duplicated(hsc_genes))
```

```{r}
hsc_mmus_specific
hsc_mcas_specific
hsc_mspr_specific
hsc_mcar_specific
```

```{r print_ct_bl6specific}
hsc_cons_df_ct <- geneset_list_hsc$`Early MPP`$conserved_df
hsc_mmus_specific <- hsc_cons_df_ct$gene[which(
  !is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]
hsc_mmus_specific

hsc_cons_df_ct <- geneset_list_hsc$`Early Myeloid`$conserved_df
hsc_mmus_specific <- hsc_cons_df_ct$gene[which(
  !is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]
hsc_mmus_specific

hsc_cons_df_ct <- geneset_list_hsc$Lymphoid$conserved_df
hsc_mmus_specific <- hsc_cons_df_ct$gene[which(
  !is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]
hsc_mmus_specific

hsc_cons_df_ct <- geneset_list_hsc$`Mk prog`$conserved_df
hsc_mmus_specific <- hsc_cons_df_ct$gene[which(
  !is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]
hsc_mmus_specific

hsc_cons_df_ct <- geneset_list_hsc$`Mk/Ery prog`$conserved_df
hsc_mmus_specific <- hsc_cons_df_ct$gene[which(
  !is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]
hsc_mmus_specific

hsc_cons_df_ct <- geneset_list_hsc$`Mono prog`$conserved_df
hsc_mmus_specific <- hsc_cons_df_ct$gene[which(
  !is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]
hsc_mmus_specific

hsc_cons_df_ct <- geneset_list_hsc$`Neutro prog`$conserved_df
hsc_mmus_specific <- hsc_cons_df_ct$gene[which(
  !is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]
hsc_mmus_specific

hsc_cons_df_ct <- geneset_list_hsc$`BM prog`$conserved_df
hsc_mmus_specific <- hsc_cons_df_ct$gene[which(
  !is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]
hsc_mmus_specific

hsc_cons_df_ct <- geneset_list_hsc$Erythroid$conserved_df
hsc_mmus_specific <- hsc_cons_df_ct$gene[which(
  !is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]
hsc_mmus_specific
```


```{r heatmap_spec2}
agg_hsc_ct <- agg_hsc_list[[celltype]]

rm(agg_hsc_list)
gc()

# make visualisation DF
vis_df_hsc <- as.data.frame(
  t(SummarizedExperiment::assays(agg_hsc_ct)$logcounts)) # batchnorm, aggregated
vis_df_hsc <- vis_df_hsc[,colnames(vis_df_hsc) %in% hsc_genes]
vis_df_hsc <- tibble::rownames_to_column(vis_df_hsc, "species")
vis_df_hsc <- tidyr::pivot_longer(
  vis_df_hsc, 
  cols = c(2:ncol(vis_df_hsc)),
  values_to = "expression",
  names_to = "gene")

vis_df_hsc$species <- factor(vis_df_hsc$species, 
                             levels = base::rev(names(col_spc_pub)))
vis_df_hsc$gene <- factor(vis_df_hsc$gene, levels = hsc_genes)

vis_df_hsc$zscore_gene <- vector(length = nrow(vis_df_hsc), mode = "numeric")

for(g in base::unique(vis_df_hsc$gene)){
  
  expr_vector_g <- vis_df_hsc[vis_df_hsc$gene == g,]$expression

  # calculate z score per gene
  z_score_g <- (expr_vector_g - base::mean(expr_vector_g))/stats::sd(expr_vector_g)
  vis_df_hsc$zscore_gene[vis_df_hsc$gene == g] <- z_score_g
  
}
levels(vis_df_hsc$gene)
```

```{r heatmap_spec3}
interest_genes2 <- c(
  "Procr",
  "Ly6a",
  "Cited2",
  "Cd74",
  "Krt18",
  "Socs2",
  "Cd34",
  "Cd63",
  "Pbx1",
  "Plek"
)

vis_df_hsc$is_gene <- vector(length = nrow(vis_df_hsc))
vis_df_hsc$is_gene[vis_df_hsc$gene %in% interest_genes2] <- "TRUE"
vis_df_hsc$is_gene <- as.logical(vis_df_hsc$is_gene)
print(levels(vis_df_hsc$gene))
```

#### PLOT

```{r heatmap_spec_base_plot}
heatmap_spec_base_plot <- ggplot2::ggplot(
  vis_df_hsc, 
  aes(y = species, 
      x = gene, 
      fill = zscore_gene))+
  ggplot2::geom_tile()
```

```{r heatmap_spec_theme_plot}
heatmap_spec_theme_plot <- heatmap_spec_base_plot+
  theme_all+
  ggplot2::theme(
    axis.title = element_blank(),
    plot.title = element_text(
      hjust = 0.5),
    legend.position = "none",
    axis.text.x = element_blank(),
    axis.text.y = element_text(
      hjust = 0),
    axis.line.x = element_blank())+
  ggplot2::ggtitle("Species-specific HSC marker genes")+
  ggplot2::scale_fill_gradientn(
    "z-score per gene", 
    colours = colors_z_score_ct,
    breaks = c(-5, -2.5, 0, 2.5, 5), 
    limits = c(-5,5))+
  ggplot2::scale_y_discrete(expand = c(0, 0))+
  ggplot2::scale_x_discrete(
    position =  "bottom",
    breaks = vis_df_hsc$gene[vis_df_hsc$is_gene])
```

#### ASSEMBLE

```{r heatmap_spec_assemble, fig.width = 4.2, fig.height = 2.17}
heatmap_spec_plot <- ggpubr::ggarrange(
  heatmap_spec_theme_plot
)
heatmap_spec_plot
```

#### EXPORT

```{r output_pdf_heatmap_spec_plot}
pdf(output_pdf_heatmap_spec_plot, width = 4.2, height = 2.17)
heatmap_spec_plot
dev.off()
```

```{r heatmap_spec_plot_remove}
rm(agg_hsc_ct)
rm(agg_hsc_df)
gc()
```

## Expression heatmap conditions

Scaled expression of signature genes [geneX], [geneY], and [geneZ] across
mouse species and ages (z-score per species and age).
For Figure 2f.

#### PREP

```{r heatmap_conds1}
metadata <- sce_hsc@colData@listData
sce_hsc_seurat <- SeuratObject::CreateSeuratObject(counts = sce_hsc@assays@data@listData[["logcounts"]],
                           meta.data = metadata)

# assign gene list!!
g <- c(
  "Nkx2-3",
  "Trib2", 
  "Blvrb"
)

metadata$expression <- SeuratObject::FetchData(sce_hsc_seurat, 
                        vars = g)
metadata$celltypes <- as.character(metadata$celltypes)
gene_expression <- metadata$expression %>%
  cbind(do.call(cbind, metadata[c("Species_ID", "Age_ID", "celltypes")]))
```

```{r heatmap_conds2}
plot_1 <- gene_expression[,c(g[1], "Species_ID", "Age_ID", "celltypes")] 
colnames(plot_1)[1] <- "gene"
plot_1 <- plot_1 %>%
  group_by(Species_ID, Age_ID, celltypes) %>%
  summarize(mean = mean(gene))
plot_1$signature <- lapply(plot_1$celltypes, function(x){
  g[1] %in% geneset_list_hsc[[x]][["conserved_signature"]]
}) %>% unlist()


plot_2 <- gene_expression[,c(g[2], "Species_ID", "Age_ID", "celltypes")] 
colnames(plot_2)[1] <- "gene"
plot_2 <- plot_2 %>%
  group_by(Species_ID, Age_ID, celltypes) %>%
  summarize(mean = mean(gene))
plot_2$signature <- lapply(plot_2$celltypes, function(x){
  g[2] %in% geneset_list_hsc[[x]][["conserved_signature"]]
}) %>% unlist()


plot_3 <- gene_expression[,c(g[3], "Species_ID", "Age_ID", "celltypes")] 
colnames(plot_3)[1] <- "gene"
plot_3 <- plot_3 %>%
  group_by(Species_ID, Age_ID, celltypes) %>%
  summarize(mean = mean(gene))
plot_3$signature <- lapply(plot_3$celltypes, function(x){
  g[3] %in% geneset_list_hsc[[x]][["conserved_signature"]]
}) %>% unlist()
```


```{r heatmap_conds3}
order_condition <- c(
 "mmus yng",
 "mmus old",
 "mcas yng",
 "mcas old",
 "mspr yng",
 "mspr old",
 "mcar yng",
 "mcar old"
)
plot_1$condition <- base::paste(plot_1$Species_ID, plot_1$Age_ID)
plot_1$condition_rev <- plot_1$condition
plot_1$condition <- factor(plot_1$condition, levels = order_condition)
plot_1$conditionrev <- factor(plot_1$condition, levels = rev(order_condition))

plot_2$condition <- base::paste(plot_2$Species_ID, plot_2$Age_ID)
plot_2$condition_rev <- plot_2$condition
plot_2$condition <- factor(plot_2$condition, levels = order_condition)
plot_2$conditionrev <- factor(plot_2$condition, levels = rev(order_condition))

plot_3$condition <- base::paste(plot_3$Species_ID, plot_3$Age_ID)
plot_3$condition_rev <- plot_3$condition
plot_3$condition <- factor(plot_3$condition, levels = order_condition)
plot_3$conditionrev <- factor(plot_3$condition, levels = rev(order_condition))
```

```{r heatmap_conds3_5}
# scale the expression levels now for better comparability/only one legend

plot_1$z_score <- (plot_1$mean - base::mean(plot_1$mean))/
    stats::sd(plot_1$mean)

plot_2$z_score <- (plot_2$mean - base::mean(plot_2$mean))/
    stats::sd(plot_2$mean)

plot_3$z_score <- (plot_3$mean - base::mean(plot_3$mean))/
    stats::sd(plot_3$mean)

max_zs <- base::max(plot_1$z_score, plot_2$z_score, plot_3$z_score)
min_zs <- base::min(plot_1$z_score, plot_2$z_score, plot_3$z_score)
midpoint <- (max_zs+min_zs)/2
```

#### PLOT

```{r heatmap_conds4}
marker_1_wide_zs <- ggplot(plot_1,
       aes(x = factor(celltypes, levels = names(geneset_list_hsc)),
           y = conditionrev,
           fill = z_score,
           color = signature,
           size = signature)) +
  geom_tile(width = 0.95, 
            height = 0.95) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, h = 1)) +
  scale_color_manual(guide = "none", values = c(alpha("black", 0), "grey50")) +
  scale_size_manual(guide = "none", values = c(0, 0.4)) +
  labs(x = NULL, y = NULL, title = g[1]) +
    ggplot2::scale_fill_gradientn(
    "z-score per gene", 
    colours = colors_z_score_cond,
    breaks = c(-5, -2.5, 0, 2.5, 5), 
    limits = c(-5,5))+
  theme(legend.position = "none")
```

```{r heatmap_conds5}
marker_2_wide_zs <- ggplot(plot_2,
       aes(x = factor(celltypes, levels = names(geneset_list_hsc)),
           y = conditionrev,
           fill = z_score,
           color = signature,
           size = signature)) +
  geom_tile(width = 0.95, 
            height = 0.95) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, h = 1),
        axis.text.y = element_blank()) +
  scale_color_manual(guide = "none", values = c(alpha("black", 0), "grey50")) +
  scale_size_manual(guide = "none", values = c(0, 0.4)) +
  labs(x = NULL, y = NULL, title = g[2]) +
    ggplot2::scale_fill_gradientn(
    "z-score per gene", 
    colours = colors_z_score_cond,
    breaks = c(-5, -2.5, 0, 2.5, 5), 
    limits = c(-5,5))+
  theme(legend.position = "none")
```

```{r heatmap_conds6}
marker_3_wide_zs <- ggplot(plot_3,
       aes(x = factor(celltypes, levels = names(geneset_list_hsc)),
           y = conditionrev,
           fill = z_score,
           color = signature,
           size = signature)) +
  geom_tile(width = 0.95, 
            height = 0.95) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, h = 1),
        axis.text.y = element_blank()) +
  scale_color_manual(guide = "none", values = c(alpha("black", 0), "grey50")) +
  scale_size_manual(guide = "none", values = c(0, 0.4)) +
  labs(x = NULL, y = NULL, title = g[3]) +
    ggplot2::scale_fill_gradientn(
    "z-score per gene", 
    colours = colors_z_score_cond,
    breaks = c(-5, -2.5, 0, 2.5, 5), 
    limits = c(-5,5))+
  theme(legend.position = "none")
```

```{r marker_1_theme}
marker_1_theme_stacked <- marker_1_wide_zs+
  theme_all+
  ggplot2::theme(
    axis.title.y = element_blank(),
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    axis.text.y = element_text(hjust = 1),
    plot.title = element_text(
      face = "italic",
      hjust = 0.5),
    legend.title = element_blank(),
    legend.position = "right")+
  ggplot2::scale_y_discrete(expand = c(0, 0))+
  ggplot2::scale_x_discrete(expand = c(0, 0))
```

```{r marker_2_theme}
marker_2_theme_stacked <- marker_2_wide_zs+
  theme_all+
  ggplot2::theme(
    axis.title.y = element_blank(),
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    axis.text.y = element_text(hjust = 1),
    plot.title = element_text(
      face = "italic",
      hjust = 0.5),
    legend.title = element_blank(),
    legend.position = "right")+
  ggplot2::scale_y_discrete(expand = c(0, 0))+
  ggplot2::scale_x_discrete(expand = c(0, 0))
```

```{r marker_3_theme}
marker_3_theme_stacked <- marker_3_wide_zs+
  theme_all+
  ggplot2::theme(
    axis.title.y = element_blank(),
    axis.title.x = element_blank(),
    axis.text.y = element_text(hjust = 1),
    axis.text.x = element_text(
      hjust = 1,
      vjust = 0.5,
      angle = 90),
    plot.title = element_text(
      face = "italic",
      hjust = 0.5),
    legend.title = element_blank(),
    legend.position = "right")+
  ggplot2::scale_y_discrete(expand = c(0, 0))+
  ggplot2::scale_x_discrete(expand = c(0, 0))
```

#### Assemble

```{r heatmap_conds_assemble_stacked, fig.width = 6, fig.height = 10}
heatmap_conds_stacked_plot <- ggpubr::ggarrange(
    marker_1_theme_stacked+theme(legend.position = "none"),
    marker_2_theme_stacked+theme(legend.position = "none"),
    marker_3_theme_stacked+theme(legend.position = "none"),
    ncol = 1,
    heights = c(1, 1, 1.7)
  )
heatmap_conds_stacked_plot
```

#### LEGEND

```{r heatmap_conds_assemble_legend, fig.width = 3, fig.height = 1}
legend_cond_plot <- marker_1_theme_stacked+
  ggplot2::theme(legend.position = "bottom")+
    ggplot2::scale_fill_gradientn(
    "z-score per gene", 
    colours = colors_z_score_cond,
    breaks = c(-5, 0,  5), 
    limits = c(-5,5))
legend_cond <- ggpubr::get_legend(legend_cond_plot)
legend_cond_theme <-  ggpubr::ggarrange(legend_cond)
legend_cond_theme
```

#### EXPORT

```{r output_pdf_heatmap_conds_stacked_plot}
pdf(output_pdf_heatmap_conds_stacked_plot, width = 6, height = 10)
heatmap_conds_stacked_plot
dev.off()
```

```{r output_pdf_heatmap_conds_stacked_legend}
pdf(output_pdf_heatmap_conds_stacked_legend, width = 3, height = 1)
legend_cond_theme
dev.off()
```

## Barplot Robustness to Aging

Number of marker genes per HSPC cell type identified when all, only old, 
or only young mouse samples are included.
For Figure 2g.

#### PREP

```{r aging1}
conserved_genes <- data.frame(celltype = rep(names(geneset_list_hsc)),
                              young = 0,
                              both = 0,
                              old = 0,
                              n_markers = 0,
                              young_in_markers = 0,
                              both_in_markers = 0,
                              old_in_markers = 0)
for(ct in names(geneset_list_hsc)){
  young <- markers_conservation_hsc[[grep(paste(ct, "young", sep = "_"),
                                      names(markers_conservation_hsc),
                                      value = T)]] %>% as.data.frame()
  young$gene <- rownames(young)
  young <- young[complete.cases(young),]

  old <- markers_conservation_hsc[[grep(paste(ct, "old", sep = "_"),
                                    names(markers_conservation_hsc),
                                    value = T)]] %>% as.data.frame()
  old$gene <- rownames(old)
  old <- old[complete.cases(old),]

  both <- young$gene[young$gene %in% old$gene]

  mark <- geneset_list_hsc[[grep(ct,
                                 names(geneset_list_hsc),
                                 value = T)]][["conserved_df"]]
  mark <- mark[complete.cases(mark),]

  conserved_genes[which(conserved_genes$celltype == ct),1] <- ct

  conserved_genes[which(conserved_genes$celltype == ct),2:8] <-
    c(nrow(young),
      length(both),
      nrow(old),
      nrow(mark),
      sum(young$gene %in% mark$gene),
      sum(both %in% mark$gene),
      sum(old$gene %in% mark$gene)) %>% as.numeric()
}

plot_hsc <- gather(conserved_genes[,c("celltype", 
                                      "n_markers",
                                      "young_in_markers", 
                                      "old_in_markers",
                                      "both_in_markers")], 
                   "group", "count", -celltype, -both_in_markers)
```

#### PLOT

```{r aging2, fig.width = 3, fig.height = 5}
print(
ggplot(plot_hsc,
       aes(y = factor(celltype, levels = rev(names(geneset_list_hsc))),
           fill = factor(group, levels = c("young_in_markers", 
                                           "n_markers", 
                                           "old_in_markers")))) +
  geom_col(aes(x = count),
           position = "dodge") +
  ggpattern::geom_col_pattern(aes(x = both_in_markers),
           position = "dodge",
           pattern = "stripe",
           pattern_color = "black",
           pattern_spacing = 0.07,
           pattern_density = 0.001,
           pattern_angle = 60) +
  theme_classic() +
  labs(y = NULL, 
       x = "# markers identified", 
       fill = NULL) +
  #theme(axis.text.y = element_text(angle = 90, h = 1)) +
  guides(fill=guide_legend(override.aes=list(pattern="none"))) +
  scale_fill_manual(labels = c("young", "all", "old"),
                      values = c("grey30", "grey60", "grey90"))
)
```

```{r aging_theme_plot, fig.height = 6}
# this is the same plot, with themes added and different format
# plus small adjustments: 
# -ggpattern (circles instead of hatches for now)
# -x axis on top and expanded all the way to 0
# -fill colors linked to variable names
# -variable factors reversed so that "young" comes before "old"
# -color and name for "all" changed to "all conserved markers"
# -legend generated separately so that "young" comes before "old"
# - y axis also expanded to 0

plot_hsc$group <- factor(
  plot_hsc$group, 
  levels = base::rev(c("young_in_markers", "n_markers", "old_in_markers")))

plot_hsc$celltype <- factor(
  plot_hsc$celltype, 
  levels = base::rev(names(geneset_list_hsc)))

plot_hsc$celltype <- factor(
  plot_hsc$celltype, 
  levels = names(geneset_list_hsc))

aging_theme_plot_wide <- ggplot2::ggplot(
  plot_hsc,
  aes(x = celltype,
      fill = group))+
  ggplot2::geom_col(
    aes(y = count),
    position = "dodge") +
  ggpattern::geom_col_pattern(
    data = plot_hsc,
    aes(y = both_in_markers),
    position = "dodge",
    pattern = "stripe",
    color = "black",
    linewidth = 0.5,
    pattern_angle = 55,
    pattern_fill = "black",
    pattern_colour = "black",
    pattern_density = pattern_density_val,
    pattern_spacing = pattern_spacing_val)+
  theme_all+
  ggplot2::theme(
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    legend.position = "none",
    plot.title = element_text(
      hjust = 0.5,
      color = col_cons_long["conserved markers"]))+
  ggplot2::labs(
    x = NULL, 
    y = "# conserved markers", 
    fill = NULL) +
  ggplot2::scale_fill_manual(
    values = c("young_in_markers" = "grey30",
               "n_markers" = col_cons_long[["conserved markers"]],
               "old_in_markers" = "grey80"))+
  ggplot2::scale_x_discrete(
    position = "top",
    expand = c(0, 0))
```

#### LEGEND

```{r aging_legend_plot}

# this is the same plot but adjusted for legend generation only
aging_legend_plot <- ggplot2::ggplot(
  plot_hsc,
  aes(y = celltype,
      fill = group))+
  ggplot2::geom_col(
    aes(x = count),
    position = "dodge")+
  theme_all+
  ggplot2::theme(
    legend.text = element_text(
      face = legend_text_face,
      size = legend_text_size,
      color = legend_text_color))+
  ggplot2::guides(fill=guide_legend(override.aes=list(pattern="none"))) +
  ggplot2::scale_fill_manual(
    "Conserved markers derived from",
    labels = c("young", "pooled", "old"),
    values = c("grey30", 
               col_cons_long[["conserved markers"]], 
               "grey80"))
aging_theme_legend <- ggpubr::get_legend(aging_legend_plot)
```

```{r aging_legend, fig.width = 8, fig.height = 2}
aging_legend <- ggpubr::ggarrange(aging_theme_legend)
aging_legend
```

```{r aging_plot_assemble_wide, fig.width = 4.57, fig.height = 2.7}
aging_theme_plot_wide
```

#### EXPORT

```{r output_pdf_aging_theme_plot}
pdf(output_pdf_aging_theme_plot, width = 4.57, height = 2.7)
aging_theme_plot_wide
dev.off()
```

```{r output_pdf_aging_theme_legend}
pdf(output_pdf_aging_theme_legend, width = 7, height = 2)
aging_legend
dev.off()
```

## Pseudotime-UMAP

Load pseudotime-related data now.
For Figure 2d.

#### PREP

```{r load_large}
# fully annotated SCE object (HSPCs) with diffusion pseudotime
sce_hsc_all_pseudotime <- base::readRDS(base::paste0(
 base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/05_cellrank/04_psce/sce_hsc_pseudotime"))


# transfer pseudotime info, then remove pseudotime SCE object
sce_hsc$pseudotime <- sce_hsc_all_pseudotime$dpt_pseudotime
rm(sce_hsc_all_pseudotime)
gc()

# fully annotated SCE object (HSPCs), lymphoid branch cells only, with 
# pseudotime and lymphoid differentiation probability
ery_pseudo <- base::readRDS(base::paste0(
  base_path, 
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/05_cellrank/05_bsce/sce_ery"))

```

```{r pt_umaps1}

# add ery info
# ery object is still needed later
sce_hsc$ery <- vector(length = ncol(sce_hsc))
sce_hsc$ery[base::match(colnames(ery_pseudo),
                        colnames(sce_hsc))] <- ery_pseudo$Erythroid

# make it so that pseudotime values are only kept for cells that are in the
# ery branch
ery_cut_off <- min(ery_pseudo$Erythroid)

sce_hsc$pt_ery <- sce_hsc$pseudotime
sce_hsc$pt_ery[sce_hsc$ery < ery_cut_off] <- NA

hsc_umap_comb <- base::as.data.frame(
  SingleCellExperiment::reducedDim(
    sce_hsc, type = "UMAP"))
hsc_umap_comb$color_by <- sce_hsc$pt_ery 
hsc_umap_comb <- hsc_umap_comb[base::sample(
  1:nrow(hsc_umap_comb),
  nrow(hsc_umap_comb),
  replace = FALSE),]
```

#### PLOT

```{r pt_umaps_base_plot_3}
pt_umaps_base_plot_3 <- ggplot2::ggplot(
  hsc_umap_comb, 
  aes(x = X1,
      y = X2,
      color = color_by))+
  ggplot2::geom_point(
    size = umap_point_size,
    alpha = umap_point_alpha)+
  ggplot2::xlim(c(base::min(hsc_umap_comb$X1)-1), 
                (base::max(hsc_umap_comb$X1)+1))+ 
  ggplot2::ylim(c(base::min(hsc_umap_comb$X2)-1), 
                (base::max(hsc_umap_comb$X2)+1))+
  ggplot2::scale_color_gradientn(
    limits = c(0, 1),
    breaks = c(0, 0.5, 1),
    na.value = "grey90",
    "Pseudotime within\nErythroid branch",
    colours = c("#061F42", "#8FBEFF"))
```

```{r pt_umaps_theme_plot_3, fig.width = 3, fig.height = 3}
pt_umaps_theme_plot_3 <- pt_umaps_base_plot_3+
  cowplot::theme_nothing()
pt_umaps_theme_plot_3
```

#### LEGEND

```{r pt_umaps_legend3, fig.width = 5, fig.height = 3}
pt_umaps_legend_3_plot <- pt_umaps_base_plot_3+
  theme_all+
  theme(legend.position = "bottom",
        legend.title.position = "top")

pt_umaps_legend3_theme <- ggpubr::get_legend(pt_umaps_legend_3_plot)
pt_umaps_legend3 <- ggpubr::ggarrange(pt_umaps_legend3_theme)
pt_umaps_legend3
```

#### EXPORT

```{r pt_umaps3_export}
pdf(output_pdf_pt_umap3_plot, width = 4, height = 3)
pt_umaps_theme_plot_3
dev.off()
```

```{r output_pdf_pt_umap_legend3}
pdf(output_pdf_pt_umap_legend3, width = 3, height = 2)
pt_umaps_legend3
dev.off()
```

```{r remove}
rm(
  heatmaps_conds_stacked_plot,
  heatmap_spec_plot,
  heatmap_spec_theme_plot,
  hsc_umap_comb,
  marker_1_theme_stacked,
  marker_1_theme_wide,
  marker_2_theme_stacked,
  marker_3_theme_stacked)
gc()
```

## Pseudotime-ranking and scaling 

Scaled expression of HSC (top), Mk/Erythroid prog. (middle) or Erythroid prog.
(bottom) conserved signature genes along diffusion pseudotime trajectory within 
the erythroid branch. Cells were partitioned into branches based on their 
lineage differentiation probability calculated by cellrank.
For Figure 2e.

#### PREP

```{r pt_ranking1, message = FALSE, warning= FALSE}
ery_pseudo_seurat <- CreateSeuratObject(counts = ery_pseudo@assays@data@listData[["logcounts"]],
                                        meta.data = ery_pseudo@colData@listData)

rm(ery_pseudo)
gc()

plot_data <- ery_pseudo_seurat@meta.data %>%
  select("pseudotime", "celltypes", "Mouse_ID", 
         "Species_ID", "Age_weeks", "Age_ID")
for(n in 1:length(geneset_list_hsc)){
  cell_sp_markers <- geneset_list_hsc[[n]][["conserved_signature"]]
  for(g in cell_sp_markers){
    plot_data[,g] <-unlist(SeuratObject::FetchData(ery_pseudo_seurat, g))
  }
}
plot_data[,7:ncol(plot_data)] <- scale(plot_data[,7:ncol(plot_data)])
```

```{r pt_ranking2}
plot_HSCs <- plot_data[,c(colnames(plot_data)[1:6],
                           geneset_list_hsc[["HSC"]][["conserved_signature"]])] %>%
  gather(gene, expression,
         -pseudotime, -celltypes, -Mouse_ID, 
         -Species_ID, -Age_weeks, -Age_ID)

n_col_hsc <- length(geneset_list_hsc[["HSC"]][["conserved_signature"]])
col_HSCs <- ggpubr::get_palette(palette = c("grey5", "grey90"), k = n_col_hsc)


plot_MkEry <- plot_data[,c(colnames(plot_data)[1:6],
                           geneset_list_hsc[["Mk/Ery prog"]][["conserved_signature"]])] %>%
  gather(gene, expression,
         -pseudotime, -celltypes, -Mouse_ID, 
         -Species_ID, -Age_weeks, -Age_ID)

n_col_mke <- length(geneset_list_hsc[["Mk/Ery prog"]][["conserved_signature"]])
col_MkEry <- ggpubr::get_palette(palette = c("grey5", "grey90"), k = n_col_mke)


plot_Ery <- plot_data[,c(colnames(plot_data)[1:6],
                           geneset_list_hsc[["Erythroid"]][["conserved_signature"]])] %>%
  gather(gene, expression,
         -pseudotime, -celltypes, -Mouse_ID, 
         -Species_ID, -Age_weeks, -Age_ID)

n_col_ery <- length(geneset_list_hsc[["Erythroid"]][["conserved_signature"]])
col_Ery <- ggpubr::get_palette(palette = c("grey5", "grey90"), k = n_col_ery)

```

#### PLOT

```{r pt_ranking3}

HSC_1 <- ggplot(plot_HSCs,
       aes(x = rank(pseudotime),
           y = expression,
           color = gene)) +
  geom_smooth(se = F) +
  guides(color = "none") +
  labs(x = "Pseudotime Erythroid", 
       y = "Scaled Expression") +
  scale_color_manual(values = col_HSCs) +
  theme_classic() +
  theme(legend.text = element_blank(),
        axis.text = element_blank())

MkEry_2 <- ggplot(plot_MkEry,
       aes(x = rank(pseudotime),
           y = expression,
           color = gene)) +
  geom_smooth(se = F) +
  guides(color = "none") +
  labs(x = "Pseudotime Erythroid", 
       y = NULL) +
  scale_color_manual(values = col_MkEry) +
  theme_classic() +
  theme(legend.text = element_blank(),
        axis.text = element_blank())

Ery_3 <- ggplot(plot_Ery,
       aes(x = rank(pseudotime),
           y = expression,
           color = gene)) +
  geom_smooth(se = F) +
  guides(color = "none") +
  labs(x = "Pseudotime Erythroid", 
       y = NULL) +
  scale_color_manual(values = col_Ery) +
  theme_classic() +
  theme(legend.text = element_blank(),
        axis.text = element_blank())
```

```{r hsc_pt_theme_plot}
# due to using ranks for a different number of data points due to different 
# numbers of genes, the actual x axis values are not identical, but the
# "meaning" of the x axis in pseudotime is identical 
# --> will remove x axis ticks to avoid confusion

# due to fitting the expression from the lowest to the largest point for each
# gene individually = min-max scaling, also no direct comparison is necessary 
# --> will remove y axis ticks to avoid confusion

HSC_theme <- HSC_1+ 
  theme_all+
  ggplot2::theme(   
    axis.title = element_blank(),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    plot.title = element_text(
      hjust = 0.5))+
  ggplot2::ggtitle("HSC")+
  ggplot2::ylab("Expression scaled along pseudo-time ranks")+
  ggplot2::xlab("Pseudo-time ranks of erythroid branch cells")
```

```{r mkery_pt_theme_plot}
MkEry_theme <- MkEry_2+ 
  theme_all+
  ggplot2::theme(   
    axis.title = element_blank(),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    plot.title = element_text(
      hjust = 0.5))+
  ggplot2::ggtitle("Mk/Ery prog")+
  ggplot2::ylab("Expression scaled along pseudo-time ranks")+
  ggplot2::xlab("Pseudo-time ranks of erythroid branch cells")
```

```{r ery_pt_theme_plot}
Ery_theme <- Ery_3+ 
  theme_all+
  ggplot2::theme(   
    axis.title = element_blank(),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    plot.title = element_text(
      hjust = 0.5))+
  ggplot2::ggtitle("Erythroid")+
  ggplot2::ylab("Expression scaled along pseudo-time ranks")+
  ggplot2::xlab("Pseudo-time ranks of erythroid branch cells")
```

```{r pt_ranking_remove1}
rm(
  plot_data,
  plot_HSCs,
  plot_MkEry,
  plot_Ery,
  HSC_1,
  MkEry_2,
  Ery_3,
  ery_pseudo_seurat
)
gc()
```

#### Assemble

```{r pt_ranking_assemble, fig.width = 3, fig.height = 7}
pt_ranking_theme_plot_long <- ggpubr::ggarrange(
    HSC_theme,
    MkEry_theme,
    Ery_theme,
    ncol = 1,
    align = "hv"
  )
pt_ranking_theme_plot_long
```

```{r pt_ranking_assemble2, fig.width = 3.9, fig.height = 9}
pt_ranking_plot_long <- ggpubr::annotate_figure(
  pt_ranking_theme_plot_long,
  top = ggpubr::text_grob(
    "Signature genes", 
    color = col_cons_long["conserved identity signature"], 
    face = plot_title_face,
    size = plot_title_size,),
  bottom = ggpubr::text_grob(
    "Pseudotime ranks\nof erythroid branch cells", 
    color = axis_title_color, 
    face = axis_title_face,
    size = axis_title_size),
  left = ggpubr::text_grob(
    rot = 90,
    "Scaled gene expression", 
    color = axis_title_color, 
    face = axis_title_face,
    size = axis_title_size),
    )
pt_ranking_plot_long
```

#### EXPORT

```{r output_pdf_pt_ranking_plot_long}
pdf(output_pdf_pt_scaled, width = 4, height = 8.1)
pt_ranking_plot_long
dev.off()
```

# UTILS

```{r session_info}
utils::sessionInfo()
```
