---
title: "Figure 6"
output: html_document
date: "2024-09-13"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Generate Figure 6, current title:
"Cross-species cell type marker identification method is robust at homologous cell type annotation across tissues and species."

#### Load objects

```{r seed, message = FALSE}
base::RNGkind(kind = "Mersenne-Twister")
set.seed(37)
```

```{r load_packages, message = FALSE}
library(tidyverse, quietly = TRUE)
library(Seurat, quietly = TRUE)
```

```{r load_manual}

# load data for manual use here
base_path_temp <- "/omics/odcf/analysis/OE0538_projects_temp/DO-0008/data_temp/"
brain_path <- "scRNAseq/main_analysis/sce_objects/08_sce_brain/"
  
colors_path <- base::paste0(
  base_path, 
  "/metadata/colors/colors.txt")

source("../source/colors.R")

```

```{r load_snakemake, include = FALSE, eval = FALSE}
# colors
colors_path <- snakemake@params[["colors_path"]]
colors_source <- snakemake@params[["colors"]]
source(colors_source)

#knitr::knit_exit()

```

```{r params}
# global params
source("determine_params.R")

col_cons_long <- c(
  "conserved identity signature" = "#db5b00ff",
  "conserved markers" = "#eb9523ff",
  "total BL6 markers" = "#a7c5a3ff",
  "ndges" = "#ea91adff",
  "random genes" = "grey80")
```

```{r export_paths}
output_pdf_b_plot <- base::paste0(
  base_path_temp, "manuscript1/figures/figure6_b.pdf")
output_pdf_b_legend <- base::paste0(
  base_path_temp, "manuscript1/figures/figure6_b_legend.pdf")
output_pdf_c_plot <- base::paste0(
  base_path_temp, "manuscript1/figures/figure6_c.pdf")
output_pdf_c_plot_poster <- base::paste0(
  base_path_temp, "manuscript1/figures/figure6_poster.pdf")
```

## A

A: Model of species used in available data (Bakken 2021 Nature: macaque, human, marmoset as ingroup, mouse as outgroup).

## B

Primate cell type identification using only primate-derived conserved identity signature or conserved marker genes.

#### Prep

```{r b_prep1}
# establish order of clusters for plotting based on 04_recluster.R
# sig = conserved identity signature genes
factorize <- list(human = c(9,6,8,13,5,0,10,12,3,14,7,2,11,1,4),
                  marmoset = c(0,3,11,8,1,10,5,14,9,13,4,6,2,7,12),
                  macaque = c(6,2,1,3,5,10,4,7,11,0,9,8))
```

```{r b_prep2}
# loop through all three primate species sig genes sets
# to create normalized matrices for plotting
plot_gg <- list()
for(sp in c("human", "marmoset", "macaque")){
  data <- readRDS(paste0(base_path,
                         brain_path,
                         "04_recluster/", sp, "_reclust_sig.rds"))

  plot <- data@meta.data[,c("subclass_label", "seurat_clusters")] %>%
    table() %>% as.matrix()

  plot_gg[[sp]] <- plot %>%
    as.data.frame() %>%
    group_by(seurat_clusters) %>%
    mutate(clust_prop = Freq/sum(Freq)) %>%
    ungroup %>%
    group_by(subclass_label) %>%
    mutate(subclass_prop = Freq/sum(Freq),
           order = match(seurat_clusters, factorize[[sp]]))
}

plot_gg_all <- bind_rows(plot_gg, .id = 'species')
```

#### Plot Components

```{r b_base_plot}
b_plot <- ggplot(plot_gg_all, 
       aes(x = factor(subclass_label,
                      levels = c("L2/3 IT", "L5 IT", "L6 IT", "L6 IT Car3", 
                                 "L5 ET", "L5/6 NP", "L6 CT", "L6b")), 
           y = order,
           fill = clust_prop)) + 
  facet_wrap(~species, 
             ncol = 1,
             strip.position = "left") +
  geom_tile()
```

```{r b_set_theme}
b_plot_theme <- b_plot + 
      ggplot2::theme_classic() +
      ggplot2::theme(
        strip.background = element_blank(),
        legend.position = "none",
        axis.title.x = element_blank(),
        axis.text = element_blank(),
        axis.text.x = element_text(
                          size = axis_text_size,
                          face = axis_text_face,
                          color = axis_text_color,
                          angle = 90,
                          hjust = 0,
                          vjust = 0.5),
        axis.ticks.y = element_blank(),
        axis.title.y = element_text(
          size = axis_title_size,
          face = axis_title_face,
          color = axis_title_color),
        plot.title = element_text(
          size = plot_title_size,
          face = plot_title_face,
          hjust = 0.5,
          color = col_cons_long["conserved identity signature"]))  +
       ggplot2::ggtitle("Conserved identity signature") +
       ggplot2::ylab("New clusters") +
       ggplot2::scale_fill_gradientn(
            colors = mycolors_to1,
            limits = c(0, 1),
            breaks = c(0, 0.5, 1))
```

#### Construct

```{r b_plot_export}
# pdf(output_pdf_b_plot, width = 3, height = 9)
# b_plot_theme
# dev.off()
```


## C

Mouse cell type identification using only primate marker genes.

#### Prep

```{r c_prep1}
# establish order of clusters for plotting based on 04_recluster.R
# core = conserved markers; sig = conserved identity signature genes
factorize_mouse <- list(core = c(10,2,5,7,6,0,4,8,9,1,12,3,11),
                        sig = c(0,11,6,10,2,7,8,4,5,9,3,1,12))
```

```{r c_prep2}
# loop through all three primate species and across core and sig genes sets
# to create normalized matrices for plotting
plot_gg_mouse <- list()
for(genes in c("core", "sig")){
  data <- readRDS(paste0(base_path, 
                         brain_path, 
                         "04_recluster/mouse_reclust_", genes, ".rds"))
  
  plot <- data@meta.data[,c("subclass_label", "seurat_clusters")] %>%
    table() %>% as.matrix()
  
  plot_gg_mouse[[genes]] <- plot %>%
    as.data.frame() %>%
    group_by(seurat_clusters) %>%
    mutate(clust_prop = Freq/sum(Freq)) %>%
    ungroup %>%
    group_by(subclass_label) %>%
    mutate(subclass_prop = Freq/sum(Freq),
           order = match(seurat_clusters, factorize_mouse[[genes]])) %>%
    filter(subclass_label != "L6 IT Car3")
}
```

#### Plot Components

```{r c_base_plot}
plotted_gg_mouse <- list()
for(genes in c("core", "sig")){
  plotted_gg_mouse[[genes]] <- ggplot(plot_gg_mouse[[genes]], 
         aes(x = factor(subclass_label, 
                              levels = c("L2/3 IT", "L5 IT", "L6 IT", 
                                         "L5 ET", "L5/6 NP", "L6 CT", "L6b")), 
             y = order,
             fill = clust_prop)) + 
    geom_tile()
}
```

```{r c_set_theme}
plotted_gg_mouse[["core"]] <- plotted_gg_mouse[["core"]] + 
    ggplot2::theme_classic() +
    ggplot2::theme(
      legend.position = "none",
      axis.title.x = element_blank(),
      axis.text = element_blank(),
      axis.text.x = element_text(
                    size = axis_text_size,
                    face = axis_text_face,
                    color = axis_text_color,
                    angle = 90,
                    hjust = 0,
                    vjust = 0.5),
      axis.ticks.y = element_blank(),
      axis.title.y = element_text(
        size = axis_title_size,
        face = axis_title_face,
        color = axis_title_color),
      plot.title = element_text(
        size = plot_title_size,
        face = plot_title_face,
        hjust = 0.5,
        color = col_cons_long["conserved markers"])) +
    ggplot2::ggtitle("Conserved markers") +
    ggplot2::scale_fill_gradientn(
      colors = mycolors_to1,
      limits = c(0, 1),
      breaks = c(0, 0.5, 1))+
    ggplot2::ylab("New clusters") +
    ggplot2::xlab("Original cell types")

plotted_gg_mouse[["sig"]] <- plotted_gg_mouse[["sig"]] + 
    ggplot2::theme_classic() +
    ggplot2::theme(
      legend.position = "none",
      axis.title.x = element_blank(),
      axis.text = element_blank(),
      axis.ticks.y = element_blank(),
      axis.title.y = element_text(
        size = axis_title_size,
        face = axis_title_face,
        color = axis_title_color),
      plot.title = element_text(
        size = plot_title_size,
        face = plot_title_face,
        hjust = 0.5,
        color = col_cons_long["conserved identity signature"]))  +
     ggplot2::ggtitle("Conserved identity signature") +
     ggplot2::scale_fill_gradientn(
          colors = mycolors_to1,
          limits = c(0, 1),
          breaks = c(0, 0.5, 1)) +
     ggplot2::ylab("New clusters")
```

#### Construct

```{r c_construct, fig.width = 3, fig.height = 4}
c_plot <- ggpubr::ggarrange(
  ncol = 1,
  plotted_gg_mouse[["sig"]] ,
  plotted_gg_mouse[["core"]],
  align = "v",
  heights = c(1, 1.5)
)
```

```{r eval = FALSE, include = FALSE, fig.width = 5, fig.height = 4}
c_plot_poster <- ggpubr::ggarrange(
  ncol = 1,
  plotted_gg_mouse[["sig"]]+theme_all+
    theme(axis.text.x = element_text(hjust = 0, angle = 90, vjust = 0.5),
          legend.position = "none",
          plot.title = element_text(
            hjust = 0.5,
            color = col_cons_long["conserved identity signature"]),
          axis.text.y = element_blank(),
          axis.ticks.y = element_blank())+
    xlab("Original cell types")
)
```


```{r c_plot_export}
# pdf(output_pdf_c_plot, width = 9, height = 4)
# c_plot
# dev.off()
```

```{r c_plot_export_poster, eval = FALSE, include = FALSE}
pdf(output_pdf_c_plot_poster, width = 5, height = 4)
c_plot_poster
dev.off()
```
