---
title: "Figure 6"
output: html_document
date: "2024-09-13"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Generate Figure 6, current title:
"Cross-species cell type marker identification method is robust at homologous cell type annotation across tissues and species."

#### Load objects

```{r seed, message = FALSE}
base::RNGkind(kind = "Mersenne-Twister")
set.seed(37)
```

```{r load_packages, message = FALSE}
library(tidyverse, quietly = TRUE)
library(Seurat, quietly = TRUE)
```

```{r load_manual, eval = FALSE, include = FALSE}

# load data for manual use here
base_path <- "/omics/odcf/analysis/OE0538_projects/DO-0008/data/"
brain_path <- "scRNAseq/main_analysis/sce_objects/08_sce_brain/"
  
colors_path <- base::paste0(
  base_path, 
  "/metadata/colors/colors.txt")

source("../source/colors.R")

```

```{r load_snakemake}
# colors
colors_path <- snakemake@params[["colors_path"]]
colors_source <- snakemake@params[["colors"]]
source(colors_source)

#knitr::knit_exit()

```

```{r params}
# global params
source("determine_params.R")

col_cons_long <- c(
  "conserved identity signature" = "#db5b00ff",
  "conserved markers" = "#eb9523ff",
  "total BL6 markers" = "#a7c5a3ff",
  "ndges" = "#ea91adff",
  "random genes" = "grey80")
```

```{r export_paths}
output_pdf_b_plot <- base::paste0(
  base_path, "manuscript1/figures/figure6_b.pdf")
output_pdf_b_legend <- base::paste0(
  base_path, "manuscript1/figures/figure6_b_legend.pdf")
```

## A

A: Model of species used in available data (Bakken 2021 Nature: macaque, human, marmoset as ingroup, mouse as outgroup).

## B

Primate cell type identification using only primate-derived conserved identity signature or conserved marker genes.

#### Prep

```{r b_prep1}
# establish order of clusters for plotting based on 04_recluster.R
# core = conserved markers; sig = conserved identity signature genes
factorize <- list(human = list(sig = c(9,6,8,13,5,0,10,12,3,14,7,2,11,1,4),
                               core = c(6,8,10,13,1,9,2,12,4,14,7,15,3,0,11,5)),
                  marmoset = list(sig = c(0,3,11,8,1,10,5,14,9,13,4,6,2,7,12),
                               core = c(0,2,16,5,7,4,11,16,14,10,15,8,13,3,1,9,12,6)),
                  macaque = list(sig = c(6,2,1,3,5,10,4,7,11,0,9,8),
                               core = c(7,9,8,3,5,10,2,12,1,13,6,4,0,11)))
```

```{r b_prep2}
# loop through all three primate species and across core and sig genes sets
# to create normalized matrices for plotting
plot_gg <- list()
for(sp in c("human", "marmoset", "macaque")){
  for(genes in c("core", "sig")){
    data <- readRDS(paste0(base_path, 
                           brain_path, 
                           "04_recluster/", sp, "_reclust_", genes, ".rds"))
    
    plot <- data@meta.data[,c("subclass_label", "seurat_clusters")] %>%
      table() %>% as.matrix()
    
    plot_gg[[sp]][[genes]] <- plot %>%
      as.data.frame() %>%
      group_by(seurat_clusters) %>%
      mutate(clust_prop = Freq/sum(Freq)) %>%
      ungroup %>%
      group_by(subclass_label) %>%
      mutate(subclass_prop = Freq/sum(Freq))
  }
}
```

#### Plot Components

```{r d_sign_base_plot}
plotted_gg <- list()
for(sp in c("human", "marmoset", "macaque")){
  for(genes in c("core", "sig")){
    plotted_gg[[sp]][[genes]] <- ggplot(plot_gg[[sp]][[genes]], 
           aes(x = factor(subclass_label, 
                                levels = c("L2/3 IT", "L5 IT", "L6 IT", "L6 IT Car3", 
                                           "L5 ET", "L5/6 NP", "L6 CT", "L6b")), 
               y = factor(seurat_clusters, 
                          levels = unique(factorize[[sp]][[genes]])),
               fill= clust_prop)) + 
      geom_tile()
  }
}
```

```{r b_set_theme}
for(sp in c("human", "marmoset", "macaque")){
  for(genes in c("core", "sig")){
    if(genes == "core"){ # bottom of plot
      title <- "conserved markers"
      plotted_gg[[sp]][[genes]] <- plotted_gg[[sp]][[genes]] + 
          ggplot2::theme_classic()+
          ggplot2::theme(
            legend.position = "none",
            axis.title.x = element_blank(),
            axis.text = element_blank(),
            axis.text.x = element_text(
                          size = axis_text_size,
                          face = axis_text_face,
                          color = axis_text_color,
                          angle = 90,
                          hjust = 0,
                          vjust = 0.5),
            axis.ticks.y = element_blank(),
            axis.title.y = element_text(
              size = axis_title_size,
              face = axis_title_face,
              color = axis_title_color),
            plot.title = element_text(
              size = plot_title_size,
              face = plot_title_face,
              hjust = 0.5,
              color = col_cons_long[title])) +
          ggplot2::ggtitle(paste(sp, title)) +
          ggplot2::scale_fill_gradientn(
            colors = mycolors_to1,
            limits = c(0, 1),
            breaks = c(0, 0.5, 1))+
          ggplot2::ylab("New clusters") +
          ggplot2::xlab("Original cell types")
    } else {
      title <- "conserved identity signature"
      plotted_gg[[sp]][[genes]] <- plotted_gg[[sp]][[genes]] + 
          ggplot2::theme_classic() +
          ggplot2::theme(
            legend.position = "none",
            axis.title.x = element_blank(),
            axis.text = element_blank(),
            axis.ticks.y = element_blank(),
            axis.title.y = element_text(
              size = axis_title_size,
              face = axis_title_face,
              color = axis_title_color),
            plot.title = element_text(
              size = plot_title_size,
              face = plot_title_face,
              hjust = 0.5,
              color = col_cons_long[title]))  +
           ggplot2::ggtitle(paste(sp, title)) +
           ggplot2::scale_fill_gradientn(
                colors = mycolors_to1,
                limits = c(0, 1),
                breaks = c(0, 0.5, 1)) +
           ggplot2::ylab("New clusters")
    }
    # print(plotted_gg[[sp]][[genes]])
  }
}
```

#### Construct

```{r b_construct, fig.width = 9, fig.height = 4}

b_plot <- ggpubr::ggarrange(
  nrow = 2,
  ncol = 3,
  plotted_gg[["human"]][["sig"]],
  plotted_gg[["marmoset"]][["sig"]],
  plotted_gg[["macaque"]][["sig"]],
  plotted_gg[["human"]][["core"]],
  plotted_gg[["marmoset"]][["core"]],
  plotted_gg[["macaque"]][["core"]],
  align = "v",
  heights = c(1, 1.5, 1, 1.5, 1, 1.5)
)

b_plot
```

```{r b_plot_export}
pdf(output_pdf_b_plot, width = 9, height = 4)
b_plot
dev.off()
```



