---
title: "Figure 4"
date: '2024-10-03'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Generate Figure 4, current title:
"Mouse conserved marker genes alone are effective at identifying homologous cell types in evolutionarily distant species."

Also contains supplementary figure parts because it's easier.

#### Load objects

```{r seed, message = FALSE}
base::RNGkind(kind = "Mersenne-Twister")
set.seed(37)
```

```{r load_packages, message = FALSE}
library(Seurat, quietly = TRUE)
library(SingleCellExperiment, quietly = TRUE)
library(ggpubr, quietly = TRUE)
library(tidyverse, quietly = TRUE)
library(gghalves, quietly = TRUE) # not in conda env
library(ggpattern, quietly = TRUE) # not in conda env
```

```{r base_path}
base_path_input <- "/omics/odcf/analysis/OE0538_projects/DO-0008/data/"
base_path_output_temp <- "/omics/odcf/analysis/OE0538_projects_temp/DO-0008/data_temp/"
```

```{r base_path_snakemake, eval = FALSE, include = FALSE}
base_path_input <- snakemake@input[["base_path_input"]]
base_path_output_temp <- snakemake@params[["base_path_output_temp"]]
```

```{r load_manual_remaining}
#-------------------------------------------------------------------------------
#### REMAINING

# list of gene sets for HSPCs (conserved signature, conserved markers, etc.)
geneset_list_str <-base::readRDS(base::paste0(
  base_path_input,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/01_reclustering_own/01_gens/geneset_list_str"))

# data frame containing the correct resolutions for other datasets as visually determined
resolution_df_path <- base::paste0(
  base_path_input,
  "/metadata/scRNAseq/03_sce_analysis/reclustering_bm/reclustering_other_resolution.txt")

resolution_df <- utils::read.csv(file = resolution_df_path, 
                               header = TRUE, 
                               sep = ";", 
                               check.names=FALSE, 
                               stringsAsFactors=FALSE, 
                               as.is=TRUE, 
                               colClasses = "character")

corr_pval_df <-  readRDS(base::paste0(
  base_path_output_temp,
  "/04_signatures/02_reclustering_other/09_crpv/all_corrected_pval"))

colors_path <- base::paste0(
  base_path_input,
  "/metadata/colors/colors.txt")

ensembl_sign <- readRDS(base::paste0(
  base_path_output_temp,
  "/04_signatures/01_reclustering_own/02_endf/ensembl_sign_hsc"))

source("../source/colors.R")
source("../source/plotting.R")
source("../source/sce_functions_reclustering.R")
```

```{r load_manual_other1}
#-------------------------------------------------------------------------------
#### OTHER MOUSE DATASET 1 (HSPC)
#### TABULA SAPIENS BONE MARROW

dataset_o1 <- "ts_bone_marrow"

seu_orig_o1 <- seu_str_recl_list_o1 <- base::readRDS(base::paste0(
  base_path_input,
  "/metadata/scRNAseq/03_sce_analysis/reclustering_bm/prepared/ts_bone_marrow"))

# list of seurat objects with reclustering vectors for each gene set and resolution
seu_str_recl_list_o1 <- base::readRDS(base::paste0(
  base_path_input,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/03_recl/reclustered_ts_bone_marrow_list"))

# list of dataframes of re-clustering scores for each gene set and resolution
score_df_str_list_o1 <- base::readRDS(base::paste0(
  base_path_input,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/04_rcls/score_df_ts_bone_marrow_list"))

# dataframes of re-clustering scores after permutation comparing gene sets
psdf_str_mark_signrand_o1 <- base::readRDS(base::paste0(
  base_path_input,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/05_perg/perm_score_df_mark_ts_bone_marrow"))
psdf_str_mmms_signrand_o1 <- base::readRDS(base::paste0(
  base_path_input,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/05_perg/perm_score_df_mmms_ts_bone_marrow"))
psdf_str_mmms_markrand_o1 <- base::readRDS(base::paste0(
  base_path_input,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/05_perg/perm_score_df_mmms_mark_ts_bone_marrow"))

# perm score DF from background permutation (SIGN)
psdf_str_sign_rand_o1 <- base::readRDS(base::paste0(
  base_path_input,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/06_psig/perm_score_df_ts_bone_marrow"))

# perm score DF from background permutation (MARK)
# psdf_str_mark_rand_o1 <- base::readRDS(base::paste0(
#   base_path,
#   "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/07_pmrk/perm_score_df_li_all_stromal"))
```

```{r selected_seu_o1}

resolution_df_o1 <- resolution_df[resolution_df$dataset == dataset_o1,]

selected_seu_list_o1 <- lapply(seu_str_recl_list_o1, function(seu_list){
  
  conslev_curr <- seu_list[[1]]@misc$used_genes
  
  resdf_temp <- resolution_df_o1[resolution_df_o1$conservation_level == conslev_curr,]
  res_curr <- resdf_temp$resolution
  
  seu_return <- seu_list[[res_curr]]
  return(seu_return)
  
})

rm(seu_str_recl_list_o1)
gc()
```

```{r load_manual_other2}
#-------------------------------------------------------------------------------
#### OTHER MOUSE DATASET 2 (HSPC)
#### Tabula sapiens HSCs and Progs (fetal)

dataset_o2 <- "ts_hscs_progenitors"

# original seurat dataset
seu_orig_o2 <- base::readRDS(base::paste0(
  base_path_input,
  "/metadata/scRNAseq/03_sce_analysis/reclustering_bm/prepared/ts_hscs_progenitors"))

# list of seurat objects with reclustering vectors for each gene set and resolution
seu_str_recl_list_o2 <- base::readRDS(base::paste0(
  base_path_input,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/03_recl/reclustered_ts_hscs_progenitors_list"))

# list of dataframes of re-clustering scores for each gene set and resolution
score_df_str_list_o2 <- base::readRDS(base::paste0(
  base_path_input,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/04_rcls/score_df_ts_hscs_progenitors_list"))

# dataframes of re-clustering scores after permutation comparing gene sets
psdf_str_mark_signrand_o2 <- base::readRDS(base::paste0(
  base_path_input,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/05_perg/perm_score_df_mark_ts_hscs_progenitors"))
psdf_str_mmms_signrand_o2 <- base::readRDS(base::paste0(
  base_path_input,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/05_perg/perm_score_df_mmms_ts_hscs_progenitors"))
psdf_str_mmms_markrand_o2 <- base::readRDS(base::paste0(
  base_path_input,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/05_perg/perm_score_df_mmms_mark_ts_hscs_progenitors"))

# perm score DF from background permutation (SIGN)
psdf_str_sign_rand_o2 <- base::readRDS(base::paste0(
  base_path_input,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/06_psig/perm_score_df_ts_hscs_progenitors"))
```

```{r selected_seu_o2}

resolution_df_o2 <- resolution_df[resolution_df$dataset == dataset_o2,]

selected_seu_list_o2 <- lapply(seu_str_recl_list_o2, function(seu_list){
  
  conslev_curr <- seu_list[[1]]@misc$used_genes
  
  resdf_temp <- resolution_df_o2[resolution_df_o2$conservation_level == conslev_curr,]
  res_curr <- resdf_temp$resolution
  
  seu_return <- seu_list[[res_curr]]
  return(seu_return)
  
})

rm(seu_str_recl_list_o2)
gc()
```

```{r params}
# global params
source("determine_params.R")
```

```{r export_paths}
output_pdf_ab_plot <- base::paste0(
  base_path_output_temp, "/manuscript1/figures/figure4_ab.pdf")
output_pdf_c_plot <- base::paste0(
  base_path_output_temp, "/manuscript1/figures/figure4_c.pdf")
output_pdf_c_legend <- base::paste0(
  base_path_output_temp, "/manuscript1/figures/figure4_c_legend.pdf")
# output_pdf_e_plot <- base::paste0(
#   base_path_output_temp, "/manuscript1/figures/figure5_e.pdf")
# output_pdf_e_legend <- base::paste0(
#   base_path_output_temp, "/manuscript1/figures/figure5_e_legend.pdf")
```

## Reclustering heatmaps 

### Re-annotate reclustered populations based on signature genes only

```{r prepare_annotate_o1}

seu_sign_o1 <- selected_seu_list_o1$seu_sign
reclustered_o1 <- base::unique(seu_sign_o1$seurat_clusters)

clust_list_o1 <- as.list(reclustered_o1)

ct_markers_o1 <- lapply(clust_list_o1, function(x){
  markers <- Seurat::FindMarkers(seu_sign_o1,
                                 test.use = "wilcox",
                                 ident.1 = x)
  markers <- markers[base::order(base::abs(markers$avg_log2FC), 
                                 decreasing=TRUE),]
  markers$which_cluster <- base::rep(x, nrow(markers))
  markers <- rownames_to_column(markers, var = "genes")
  
  # need to add only ensembl gene symbols as I will not be using others
  markers$symbol <- markers$genes
  markers$symbol <- ensembl_sign$HSAPIENS_SYMBOL[base::match(markers$genes, ensembl_sign$ENSG_ID)]
  return(markers)
})
names(ct_markers_o1) <- reclustered_o1

ct_markers_df_o1 <- bind_rows(ct_markers_o1)
head(ct_markers_df_o1)

```

```{r prepare_annotate_o2}

seu_sign_o2 <- selected_seu_list_o2$seu_sign
reclustered_o2 <- base::unique(seu_sign_o2$seurat_clusters)

clust_list_o2 <- as.list(reclustered_o2)

ct_markers_o2 <- lapply(clust_list_o2, function(x){
  markers <- Seurat::FindMarkers(seu_sign_o2,
                                 test.use = "wilcox",
                                 ident.1 = x)
  markers <- markers[base::order(base::abs(markers$avg_log2FC), 
                                 decreasing=TRUE),]
  markers$which_cluster <- base::rep(x, nrow(markers))
  markers <- rownames_to_column(markers, var = "genes")
  
  # need to add only ensembl gene symbols as I will not be using others
  markers$symbol <- markers$genes
  markers$symbol <- ensembl_sign$HSAPIENS_SYMBOL[base::match(markers$genes, ensembl_sign$ENSG_ID)]
  return(markers)
})
names(ct_markers_o2) <- reclustered_o2

ct_markers_df_o2 <- bind_rows(ct_markers_o2)
head(ct_markers_df_o2)

```

```{r prepare_annotate_o2_hsc_only}

seu_sign_o2_hsc <- seu_sign_o2[,seu_sign_o2$seurat_clusters %in% c("4", "10", "1")]
reclustered_o2_hsc <- base::unique(seu_sign_o2_hsc$seurat_clusters)

clust_list_o2_hsc <- as.list(unfactor(reclustered_o2_hsc))

ct_markers_o2_hsc <- lapply(clust_list_o2_hsc, function(x){
  markers <- Seurat::FindMarkers(seu_sign_o2_hsc,
                                 test.use = "wilcox",
                                 ident.1 = x)
  markers <- markers[base::order(base::abs(markers$avg_log2FC), 
                                 decreasing=TRUE),]
  markers$which_cluster <- base::rep(x, nrow(markers))
  markers <- rownames_to_column(markers, var = "genes")
  
  # need to add only ensembl gene symbols as I will not be using others
  markers$symbol <- markers$genes
  markers$symbol <- ensembl_sign$HSAPIENS_SYMBOL[base::match(markers$genes, ensembl_sign$ENSG_ID)]
  return(markers)
})
names(ct_markers_o2_hsc) <- reclustered_o2_hsc

ct_markers_df_o2_hsc <- bind_rows(ct_markers_o2_hsc)
head(ct_markers_df_o2_hsc)

```

```{r prepare_annotate_o2_mono_only}

seu_sign_o2_mono <- seu_sign_o2[,seu_sign_o2$seurat_clusters %in% c("9", "0")]
reclustered_o2_mono <- base::unique(seu_sign_o2_mono$seurat_clusters)

clust_list_o2_mono <- as.list(unfactor(reclustered_o2_mono))

ct_markers_o2_mono <- lapply(clust_list_o2_mono, function(x){
  markers <- Seurat::FindMarkers(seu_sign_o2_mono,
                                 test.use = "wilcox",
                                 ident.1 = x)
  markers <- markers[base::order(base::abs(markers$avg_log2FC), 
                                 decreasing=TRUE),]
  markers$which_cluster <- base::rep(x, nrow(markers))
  markers <- rownames_to_column(markers, var = "genes")
  
  # need to add only ensembl gene symbols as I will not be using others
  markers$symbol <- markers$genes
  markers$symbol <- ensembl_sign$HSAPIENS_SYMBOL[base::match(markers$genes, ensembl_sign$ENSG_ID)]
  return(markers)
})
names(ct_markers_o2_mono) <- reclustered_o2_mono

ct_markers_df_o2_mono <- bind_rows(ct_markers_o2_mono)
head(ct_markers_df_o2_mono)

# monos 9: TMEM176A TMEM176B LTB MEIS1 H2AX ANXA2 HMBS FABP5 CYC1 MDH2 BOLA3 TRIB2 SELL TCF4 CLEC12A LY86 COPE TMEM147

# - TMEM176A TMEM176B (immature) dendritic like 
# - LTB: early dendritic cells/precursors
# - MEIS1: progenitor state
# - H2AX: dna repair, differentiation
# - ANXA2: general inflammation in mono macro and DC
# - HMBS: erythrocytes?
# - FABP5: expressed in macrophages and dendritic cells
# - TCF4: DC critical


# monos 0: MS4A2 CKAP2L H1-3 KNL1 MSI2 CDCA8 RACGAP1 MYB CENPA TPX2 KIF20B H1-4 H1-2 ANXA3 MKI67 CENPF CENPE H1-10 PNRC1

# - MS4A2: mast cell IgE related
# - CKAP2L: cell cycle/spindle related 
# - H1 histone linkers: in earlier monos, because they are downregulated during diff https://pmc.ncbi.nlm.nih.gov/articles/PMC4288251/
# - KNL1: related to proliferation
# - MSI2 progenitor/cancer
# - CDC8: cell division
# - TPX2 also related to cell cycle

```

#### Others - 1

```{r umap_others_1}
DimPlot(seu_sign_o1, dims = c(1,2), group.by = "seurat_clusters")
seu_sign_o1$annotated_reclustered <- vector(length = ncol(seu_sign_o1))
```

```{r 4_others_1}
nr <- "4"
markers <- ct_markers_o1[[nr]]$genes[
  !is.na(ct_markers_o1[[nr]]$symbol) &
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o1[[nr]]$genes) 

ct_markers_o1[[nr]][
  !is.na(ct_markers_o1[[nr]]$symbol) & 
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o1, markers[1:6])

seu_sign_o1$annotated_reclustered[seu_sign_o1$seurat_clusters == nr] <- "HSCs"
```

```{r 5_others_1}
nr <- "5"
markers <- ct_markers_o1[[nr]]$genes[
  !is.na(ct_markers_o1[[nr]]$symbol) &
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o1[[nr]]$genes) 

ct_markers_o1[[nr]][
  !is.na(ct_markers_o1[[nr]]$symbol) & 
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o1, markers[1:6])

seu_sign_o1$annotated_reclustered[seu_sign_o1$seurat_clusters == nr] <- "Mk/Ery p."
```

```{r 6_others_1}
nr <- "6"
markers <- ct_markers_o1[[nr]]$genes[
  !is.na(ct_markers_o1[[nr]]$symbol) &
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o1[[nr]]$genes) 

ct_markers_o1[[nr]][
  !is.na(ct_markers_o1[[nr]]$symbol) & 
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o1, markers[1:6])

seu_sign_o1$annotated_reclustered[seu_sign_o1$seurat_clusters == nr] <- "Mk p."
```

```{r 8_others_1}
nr <- "8"
markers <- ct_markers_o1[[nr]]$genes[
  !is.na(ct_markers_o1[[nr]]$symbol) &
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o1[[nr]]$genes) 

ct_markers_o1[[nr]][
  !is.na(ct_markers_o1[[nr]]$symbol) & 
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o1, markers[1:6])

seu_sign_o1$annotated_reclustered[seu_sign_o1$seurat_clusters == nr] <- "Late Ery p."
```

```{r 1_others_1}
nr <- "1"
markers <- ct_markers_o1[[nr]]$genes[
  !is.na(ct_markers_o1[[nr]]$symbol) &
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o1[[nr]]$genes) 

ct_markers_o1[[nr]][
  !is.na(ct_markers_o1[[nr]]$symbol) & 
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o1, markers[1:6])

seu_sign_o1$annotated_reclustered[seu_sign_o1$seurat_clusters == nr] <- "Ery p."
```

```{r 7_others_1}
nr <- "7"
markers <- ct_markers_o1[[nr]]$genes[
  !is.na(ct_markers_o1[[nr]]$symbol) &
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o1[[nr]]$genes) 

ct_markers_o1[[nr]][
  !is.na(ct_markers_o1[[nr]]$symbol) & 
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o1, markers[1:6])

seu_sign_o1$annotated_reclustered[seu_sign_o1$seurat_clusters == nr] <- "Early myeloid p."
```

```{r 3_others_1}
nr <- "3"
markers <- ct_markers_o1[[nr]]$genes[
  !is.na(ct_markers_o1[[nr]]$symbol) &
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o1[[nr]]$genes) 

ct_markers_o1[[nr]][
  !is.na(ct_markers_o1[[nr]]$symbol) & 
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o1, markers[1:6])

seu_sign_o1$annotated_reclustered[seu_sign_o1$seurat_clusters == nr] <- "Neutro p."
```

```{r 2_others_o1}
nr <- "2"
markers <- ct_markers_o1[[nr]]$genes[
  !is.na(ct_markers_o1[[nr]]$symbol) &
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o1[[nr]]$genes) 

ct_markers_o1[[nr]][
  !is.na(ct_markers_o1[[nr]]$symbol) & 
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o1, markers[1:6])

seu_sign_o1$annotated_reclustered[seu_sign_o1$seurat_clusters == nr] <- "Cycling myeloid p."
```

```{r 0_others_1}
nr <- "0"
markers <- ct_markers_o1[[nr]]$genes[
  !is.na(ct_markers_o1[[nr]]$symbol) &
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o1[[nr]]$genes) 

ct_markers_o1[[nr]][
  !is.na(ct_markers_o1[[nr]]$symbol) & 
    ct_markers_o1[[nr]]$avg_log2FC > 0 &
    ct_markers_o1[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o1, markers[1:6])

seu_sign_o1$annotated_reclustered[seu_sign_o1$seurat_clusters == nr] <- "batch - Smart-seq2"
```

```{r}
seu_sign_o1$annotated_reclustered <- factor(seu_sign_o1$annotated_reclustered,
                                            levels = c("HSCs",
                                                       "Mk/Ery p.",
                                                       "Mk p.",
                                                       "Ery p.",
                                                       "Late Ery p.",
                                                       "Early myeloid p.",
                                                       "Neutro p.",
                                                       "Cycling myeloid p.",
                                                       "batch - Smart-seq2"))
plot_suppl_a1 <- DimPlot(seu_sign_o1, dims = c(1,2), group.by = "annotated_reclustered")
plot_suppl_a2 <- DimPlot(seu_sign_o1, dims = c(1,2), group.by = "cell_type")
plot_suppl_a3 <- DimPlot(seu_sign_o1, dims = c(1,2), group.by = "assay")
```

```{r fig.width = 16, fig.height = 3.5}
suppl_plot_a <- ggpubr::ggarrange(
  plot_suppl_a1,
  plot_suppl_a2,
  plot_suppl_a3,
  nrow = 1,
  align = "hv"
)
```


#### Others - 2

```{r umap_others_2}
DimPlot(seu_sign_o2, dims = c(1,2), group.by = "seurat_clusters")
DimPlot(seu_sign_o2, dims = c(1,2), group.by = "assay")
DimPlot(seu_sign_o2, dims = c(1,2), group.by = "donor_id")
DimPlot(seu_sign_o2, dims = c(1,2), group.by = "sex")
DimPlot(seu_sign_o2, dims = c(1,2), group.by = "development_stage")
DimPlot(seu_sign_o2, dims = c(1,2), group.by = "tissue")

seu_sign_o2$annotated_reclustered <- vector(length = ncol(seu_sign_o2))
```

```{r 1_others_2}
nr <- "1"
markers <- ct_markers_o2[[nr]]$gene[
  !is.na(ct_markers_o2[[nr]]$symbol) &
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
  !is.na(ct_markers_o2[[nr]]$symbol) & 
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o2, markers[1:6])

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "HSCs - batch 10X 3'V2"
```

```{r 4_others_2}
nr <- "4"
markers <- ct_markers_o2[[nr]]$genes[
  !is.na(ct_markers_o2[[nr]]$symbol) &
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
  !is.na(ct_markers_o2[[nr]]$symbol) & 
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o2, markers[1:6])

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "HSCs - batch 10X 5'V3"

```

```{r 10_others_2}
nr <- "10"
markers <- ct_markers_o2[[nr]]$genes[
  !is.na(ct_markers_o2[[nr]]$symbol) &
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
  !is.na(ct_markers_o2[[nr]]$symbol) & 
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])
FeaturePlot(seu_sign_o2, markers[13:18])

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "HSCs - mixed"
```

```{r 3_others_2}
nr <- "3"
markers <- ct_markers_o2[[nr]]$genes[
  !is.na(ct_markers_o2[[nr]]$symbol) &
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
  !is.na(ct_markers_o2[[nr]]$symbol) & 
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o2, markers[1:6])

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Early Cycling"

```

```{r 6_others_2}
nr <- "6"
markers <- ct_markers_o2[[nr]]$genes[
  !is.na(ct_markers_o2[[nr]]$symbol) &
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
  !is.na(ct_markers_o2[[nr]]$symbol) & 
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o2, markers[1:6])

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Neutro"

```

```{r 2_others_2}
nr <- "2"
markers <- ct_markers_o2[[nr]]$genes[
  !is.na(ct_markers_o2[[nr]]$symbol) &
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
  !is.na(ct_markers_o2[[nr]]$symbol) & 
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o2, markers[1:6])

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Myeloid p."

```

```{r 9_others_2}
nr <- "9"
markers <- ct_markers_o2[[nr]]$genes[
  !is.na(ct_markers_o2[[nr]]$symbol) &
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
  !is.na(ct_markers_o2[[nr]]$symbol) & 
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05,]

# comparing both mono subsets directly
ct_markers_o2_mono[[nr]][
  !is.na(ct_markers_o2_mono[[nr]]$symbol) & 
    ct_markers_o2_mono[[nr]]$avg_log2FC > 0 &
    ct_markers_o2_mono[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Mono (Dendritic) p."
```

```{r 0_others_2}
nr <- "0"
markers <- ct_markers_o2[[nr]]$genes[
  !is.na(ct_markers_o2[[nr]]$symbol) &
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
  !is.na(ct_markers_o2[[nr]]$symbol) & 
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05,]

# comparing both mono subsets directly
ct_markers_o2_mono[[nr]][
  !is.na(ct_markers_o2_mono[[nr]]$symbol) & 
    ct_markers_o2_mono[[nr]]$avg_log2FC > 0 &
    ct_markers_o2_mono[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Mono (Cycling) p."
```

```{r 5_others_2}
nr <- "5"
markers <- ct_markers_o2[[nr]]$genes[
  !is.na(ct_markers_o2[[nr]]$symbol) &
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
  !is.na(ct_markers_o2[[nr]]$symbol) & 
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Lymphoid"
```

```{r 7_others_2}
nr <- "7"
markers <- ct_markers_o2[[nr]]$gene[
  !is.na(ct_markers_o2[[nr]]$symbol) &
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
  !is.na(ct_markers_o2[[nr]]$symbol) & 
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o2, markers[1:6])

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Mk/Ery p. - batch 10X 3'V2"
```


```{r 8_others_2}
nr <- "8"
markers <- ct_markers_o2[[nr]]$gene[
  !is.na(ct_markers_o2[[nr]]$symbol) &
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
  !is.na(ct_markers_o2[[nr]]$symbol) & 
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o2, markers[1:6])

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Mk/Ery p. - batch 10X 5'V3"
stop()
```

```{r}
seu_sign_o2$annotated_reclustered <- factor(seu_sign_o2$annotated_reclustered,
                                            levels = c(  
    "HSCs - batch 10X 5'V3",
    "HSCs - batch 10X 3'V2",
    "HSCs - mixed",
    "Early Cycling",
    "Neutro",
    "Myeloid p.",
    "Mono (Dendritic) p.",
    "Mono (Cycling) p.",
    "Lymphoid",
    "Mk/Ery p. - batch 10X 5'V3",
    "Mk/Ery p. - batch 10X 3'V2"))

seu_sign_o2$cell_type <- unfactor(seu_sign_o2$cell_type)
seu_sign_o2$cell_type[seu_sign_o2$cell_type == "hematopoietic stem cell"] <- "HSC"
seu_sign_o2$cell_type[seu_sign_o2$cell_type == "hematopoietic multipotent progenitor cell"] <- "hematopoietic MPP"
seu_sign_o2$cell_type[seu_sign_o2$cell_type == "common myeloid progenitor"] <- "common myeloid p."
seu_sign_o2$cell_type[seu_sign_o2$cell_type == "granulocyte monocyte progenitor cell"] <- "granulocyte monocyte p."
seu_sign_o2$cell_type[seu_sign_o2$cell_type == "early lymphoid progenitor"] <- "early lymphoid p."
seu_sign_o2$cell_type[seu_sign_o2$cell_type == "megakaryocyte-erythroid progenitor cell"] <- "megakaryocyte-erythroid p."

seu_sign_o2$cell_type <- factor(
  seu_sign_o2$cell_type,
  levels = base::c(
    "HSC",
    "hematopoietic MPP",
    "common myeloid p.",
    "granulocyte monocyte p.",
    "promonocyte",
    "promyelocyte",
    "early lymphoid p.",
    "megakaryocyte-erythroid p."
  ))

plot_suppl_b1 <- DimPlot(seu_sign_o2, dims = c(1,2), group.by = "annotated_reclustered")
plot_suppl_b2 <- DimPlot(seu_sign_o2, dims = c(1,2), group.by = "cell_type")
plot_suppl_b3 <- DimPlot(seu_sign_o2, dims = c(1,2), group.by = "assay")
```

```{r fig.width = 16, fig.height = 3.5}
suppl_plot_b <- ggpubr::ggarrange(
  plot_suppl_b1,
  plot_suppl_b2,
  plot_suppl_b3,
  nrow = 1,
  align = "hv"
)
```
### Make plots


#### Other 1

```{r mat_sign_o1}
mat_sign_o1_anno <- base::table(seu_sign_o1$cell_type, 
                                seu_sign_o1$annotated_reclustered)

for(i in 1:ncol(mat_sign_o1_anno)){
  mat_sign_o1_anno[,i] <- mat_sign_o1_anno[,i]/base::sum(mat_sign_o1_anno[,i])
}
sign_df_o1_anno <- base::as.data.frame(mat_sign_o1_anno)

sign_df_o1_anno$Var2 <- factor(
  sign_df_o1_anno$Var2,
  levels = base::rev(c("HSCs", 
                       "Mk/Ery p.",
                       "Mk p.", 
                       "Ery p.",
                       "Late Ery p.",
                       "Early myeloid p.", 
                       "Neutro p.",
                       "Cycling myeloid p.",
                       "batch - Smart-seq2" )))

levels(sign_df_o1_anno$Var1)

sign_df_o1_anno$Var1 <- unfactor(sign_df_o1_anno$Var1)
sign_df_o1_anno$Var1[sign_df_o1_anno$Var1 == "hematopoietic stem cell"] <- "HSC"
sign_df_o1_anno$Var1[sign_df_o1_anno$Var1 == "common myeloid progenitor"] <- "common myeloid p."
sign_df_o1_anno$Var1[sign_df_o1_anno$Var1 == "erythroid progenitor cell"] <- "erythroid p."

sign_df_o1_anno$Var1 <- factor(
  sign_df_o1_anno$Var1,
  levels = base::c("HSC", 
                   "common myeloid p.",
                   "erythroid p.",
                   "granulocyte"
                   ))
```

```{r sign_base_plot_o1}
sign_base_plot_o1 <- ggplot2::ggplot(
  sign_df_o1_anno, 
  aes(x = Var1, 
      y = Var2,
      fill = Freq))+
  ggplot2::geom_tile()
```

```{r sign_theme_plot_o1}
sign_theme_plot_o1 <- sign_base_plot_o1+
  theme_all+
  ggplot2::theme(
    legend.position = "none",
    axis.title.x = element_blank(),
    axis.ticks.y = element_blank(),
    axis.text.y  = element_text(
      vjust = 0.5,
      hjust = 0),
    axis.text.x = element_text(
      angle = 90,
      vjust = 0.5,
      hjust = 1))+
  ggplot2::scale_fill_gradientn(
    colors = mycolors_to1,
    limits = c(0, 1),
    breaks = c(0, 0.5, 1))+
  ggplot2::ylab("Reclustered + re-annotated")
```

#### Others 2

```{r mat_sign_o2}
mat_sign_o2 <- base::table(seu_sign_o2$cell_type, 
                            seu_sign_o2$annotated_reclustered)

for(i in 1:ncol(mat_sign_o2)){
  mat_sign_o2[,i] <- mat_sign_o2[,i]/base::sum(mat_sign_o2[,i])
}
sign_df_o2 <- base::as.data.frame(mat_sign_o2)

sign_df_o2$Var2 <- factor(
  sign_df_o2$Var2,
  levels = base::rev(c(
    "HSCs - batch 10X 5' V3",
    "HSCs - batch 10X 3' V2",
    "HSCs - mixed",
    "Early Cycling",
    "Neutro",
    "Myeloid p.",
    "Mono (Dendritic) p.",
    "Mono (Cycling) p.",
    "Lymphoid",
      "Mk/Ery p. - batch 10X 5' V3",
    "Mk/Ery p. - batch 10X 3' V2"
  )))


sign_df_o2$Var1 <- unfactor(sign_df_o2$Var1)
sign_df_o2$Var1[sign_df_o2$Var1 == "hematopoietic stem cell"] <- "HSC"
sign_df_o2$Var1[sign_df_o2$Var1 == "hematopoietic multipotent progenitor cell"] <- "hematopoietic MPP"
sign_df_o2$Var1[sign_df_o2$Var1 == "common myeloid progenitor"] <- "common myeloid p."
sign_df_o2$Var1[sign_df_o2$Var1 == "granulocyte monocyte progenitor cell"] <- "granulocyte monocyte p."
sign_df_o2$Var1[sign_df_o2$Var1 == "early lymphoid progenitor"] <- "early lymphoid p."
sign_df_o2$Var1[sign_df_o2$Var1 == "megakaryocyte-erythroid progenitor cell"] <- "megakaryocyte-erythroid p."

sign_df_o2$Var1 <- factor(
  sign_df_o2$Var1,
  levels = base::c(
    "HSC",
    "hematopoietic MPP",
    "common myeloid p.",
    "granulocyte monocyte p.",
    "promonocyte",
    "promyelocyte",
    "early lymphoid p.",
    "megakaryocyte-erythroid p."
  ))
```

```{r sign_base_plot_o2}
sign_base_plot_o2 <- ggplot2::ggplot(
  sign_df_o2, 
  aes(x = Var1, 
      y = Var2,
      fill = Freq))+
  ggplot2::geom_tile()
```

```{r sign_theme_plot_o2, fig.height = 6, fig.width = 6}
sign_theme_plot_o2 <- sign_base_plot_o2+
  theme_all+
  ggplot2::theme(
    legend.position = "none",
    #axis.title.x = element_blank(),
    #axis.text.y = element_blank(),
    axis.text.x = element_text(
      angle = 90,
      hjust = 1, 
      vjust = 0.5),
    axis.text.y = element_text(
      hjust = 0, 
      vjust = 0.5),
    axis.ticks.y = element_blank())+
  ggplot2::scale_fill_gradientn(
    colors = mycolors_to1,
    limits = c(0, 1),
    breaks = c(0, 0.5, 1))+
  ggplot2::ylab("Reclustered + re-annotated")+
  ggplot2::xlab("Original cell types")
```


### Assemble

```{r ab_plotparts, fig.width = 5, fig.height = 9}

sign_plots_ab <- ggpubr::ggarrange(
  ncol = 1,
  align = "v",
  sign_theme_plot_o1,
  sign_theme_plot_o2,
  heights = c(2, 2.6))
```

```{r ab_plot_export}
pdf(output_pdf_ab_plot, width = 5, height = 9)
sign_plots_ab
dev.off()
```

## D Pval Plot

```{r d_prep1}

# add more columns for nicer visualisation
corr_pval_df$color_id <- vector(length = nrow(corr_pval_df))
corr_pval_df$color_id[corr_pval_df$pval_corrected < 0.05] <- "significant"
corr_pval_df$color_id[corr_pval_df$pval_corrected == 0] <- "is_null"
corr_pval_df$color_id[corr_pval_df$pval_corrected >= 0.05] <- "non_significant"

corr_pval_df$cons_level_long <- vector(length = nrow(corr_pval_df))
corr_pval_df$cons_level_long[
  corr_pval_df$conservation_level == "conserved_signature"] <- "conserved identity signature"
corr_pval_df$cons_level_long[
  corr_pval_df$conservation_level == "conserved_markers"] <- "conserved markers"
corr_pval_df$cons_level_long[
  corr_pval_df$conservation_level == "mmusall_markers"] <- "all BL6 markers"

corr_pval_df$score_long <- vector(length = nrow(corr_pval_df))
corr_pval_df$score_long[
  corr_pval_df$type == "adjusted_rand_index"] <- "Adjusted Rand Index"
corr_pval_df$score_long[
  corr_pval_df$type == "variation_information"] <- "Variation of Information"
corr_pval_df$score_long[
  corr_pval_df$type == "mean_prop_cells_cluster"] <- "Mean prop."

corr_pval_df$cons_level_long <- factor(
  corr_pval_df$cons_level_long, levels = c(
    "conserved identity signature",
    "conserved markers",
    "all BL6 markers"))

corr_pval_df$score_long <- factor(
  corr_pval_df$score_long, levels = c(
    "Adjusted Rand Index",
    "Variation of Information",
    "Mean prop."))

corr_pval_df$color_id <- factor(
  corr_pval_df$color_id, levels = rev(c(
    "non_significant",
    "significant",
    "is_null")))

color_significance <- c("non_significant" = "grey90",
                        "significant" = "grey40", 
                        "is_null" = "blue")
```

```{r d_prep2_subset}

corr_pval_df_sub <- corr_pval_df[corr_pval_df$condition %in% c(
  dataset_o1,
  dataset_o2,
  "mus_weinreb_hspc",
  "mus_tm_bonemarrow"),]

corr_pval_df_sub <- corr_pval_df_sub[corr_pval_df_sub$conservation_level != "conserved_markers",]

corr_pval_df_sub$condition_long <- corr_pval_df_sub$condition
corr_pval_df_sub$condition_long[
  corr_pval_df_sub$condition == dataset_o1] <- "tabula sapiens BM"
corr_pval_df_sub$condition_long[
  corr_pval_df_sub$condition == dataset_o2] <- "tabula sapiens fetal"
corr_pval_df_sub$condition_long[
  corr_pval_df_sub$condition == "mus_tm_bonemarrow"] <- "tabula muris"
corr_pval_df_sub$condition_long[
  corr_pval_df_sub$condition == "mus_weinreb_hspc"] <- "Weinreb et al."

corr_pval_df_sub$condition_long <- factor(
  corr_pval_df_sub$condition_long, levels = rev(c(
    "tabula muris",
    "Weinreb et al.",
    "tabula sapiens BM",
    "tabula sapiens fetal")))

corr_pval_df_sub$comparison <- factor(
  corr_pval_df_sub$comparison, levels = c(
    "sign-vs-rand",
    "mmms-vs-signrand"))

corr_pval_df_sub <- corr_pval_df_sub[
  corr_pval_df_sub$comparison %in% c("mmms-vs-signrand", "sign-vs-rand"),]

corr_pval_df_sub$x_axis <- rep("1", nrow(corr_pval_df_sub))
```

```{r d_base_plot, fig.width = 10, fig.height= 3} 
d_base_plot <- ggplot2::ggplot(
  corr_pval_df_sub,
  aes(y = condition_long, 
      x = x_axis,
      color = color_id, 
      size = 10^-pval_corrected))+
  ggplot2::geom_point(alpha = 1)+
  ggplot2::facet_grid(cols = vars(corr_pval_df_sub$score_long),
                      switch = "y",
                      rows = vars(corr_pval_df_sub$comparison))+
  ggplot2::theme_classic()+
  ggplot2::scale_color_manual(values = color_significance)
```

```{r d_theme_plot, fig.width = 8, fig.height = 3} 
d_theme_plot <- d_base_plot+
  theme_all+
  # ggplot2::scale_x_continuous(
  #   limits = c(0, base::max(corr_pval_df_sub$nr_genes_used)))+
  ggplot2::scale_color_manual(values = color_significance)+
  ggplot2::theme(
    legend.position = "none",
    strip.text.x = element_text(
      size = axis_title_size,
      color = axis_title_color,
      face = axis_title_face,
      hjust = 1,
      angle = 0),
    axis.text.y = element_text(
      hjust = 0),
    axis.text.x = element_blank(),   
    axis.ticks.x = element_blank(),   
    strip.text.y.left = element_text(
      size = axis_text_size,
      color = axis_text_color,
      face = axis_text_face,
      hjust = 0,
      angle = 0),
    strip.background = element_blank(),
    strip.placement = "outside",
    axis.title.y = element_blank(),
    axis.title = element_blank())
d_theme_plot
```

```{r}
d_legend_base_plot <- ggplot2::ggplot(
  corr_pval_df,
  aes(y = condition, 
      x = nr_genes_used,
      color = color_id, 
      size = 10^-pval_corrected))+
  ggplot2::geom_point(alpha = 1)+
  ggplot2::facet_grid(cols = vars(corr_pval_df$score_long),
                      switch = "y",
                      rows = vars(corr_pval_df$cons_level_long))+
  ggplot2::theme_classic()+
  ggplot2::scale_x_continuous(limits = c(0, max(corr_pval_df$nr_genes_used)))+
  ggplot2::scale_color_manual(values = color_significance)
```

```{r d_legend_theme_plot, fig.width = 3, fig.height= 4}
d_legend_theme_plot <- d_legend_base_plot+
   ggplot2::theme_classic()+
  ggplot2::scale_x_continuous(
    limits = c(0, base::max(corr_pval_df_sub$nr_genes_used)))+
  ggplot2::scale_color_manual("", values = color_significance)+
  ggplot2::scale_size_continuous("10 adjusted p val")+
  ggplot2::theme(
    legend.title = element_text(
      size = legend_title_size,
      color = legend_title_color,
      face = legend_title_face),
    legend.text = element_text(
      size = legend_text_size,
      color = legend_text_color,
      face = legend_text_face))

d_legend_theme <- ggpubr::get_legend(d_legend_theme_plot)

d_legend <- ggpubr::ggarrange(d_legend_theme)
d_legend
```

```{r d_plot_export}
pdf(output_pdf_c_plot, width = 7, height = 3)
d_theme_plot
dev.off()
```

```{r d_legend_export}
pdf(output_pdf_c_legend, width = 3, height = 3)
d_legend
dev.off()
```

## Other2 expression heatmap

```{r e_prep1, include = FALSE, eval = FALSE}

# get all used conserved signature genes for zebrafish
# these are all genes remaining in the reclustered object
sign_ids <- rownames(selected_seu_list_o2$seu_sign)

# aggregate per cell type
# convert to SCE
sce_o2 <- Seurat::as.SingleCellExperiment(dataset_o2_orig)
agg_o2 <- scuttle::aggregateAcrossCells(
    sce_o2, 
    id=colData(sce_o2)[,c("cell_type")], 
    statistics = "mean",
    use.assay.type = "logcounts") 

# make a visualisation DF
vis_df <- base::as.data.frame(
  t(SummarizedExperiment::assays(agg_o2)$logcounts))
  
# subset to all signature genes of this tissue
vis_df <- vis_df[,colnames(vis_df) %in% sign_ids]
  
# change df format 
vis_df_mat <- vis_df
vis_df <- tibble::rownames_to_column(vis_df, "cell_type")
vis_df <- tidyr::pivot_longer(vis_df, 
                              cols = c(2:ncol(vis_df)),
                              values_to = "expression",
                              names_to = "gene")
 
# calculate z_score
vis_df$zscore_gene <- vector(length = nrow(vis_df), mode = "numeric")
for(g in base::unique(vis_df$gene)){
  
  expr_vector_g <- vis_df[vis_df$gene == g,]$expression

  # calculate z score per gene
  z_score_g <- (expr_vector_g - base::mean(expr_vector_g))/
    stats::sd(expr_vector_g)
  vis_df$zscore_gene[vis_df$gene == g] <- z_score_g
} 
```

```{r seurat_plot, eval = FALSE, include = FALSE}
dataset_o2_orig <- Seurat::ScaleData(dataset_o2_orig, features = sign_ids)

Seurat::DoHeatmap(
  dataset_o3_orig, 
  features = sign_ids, 
  assay = "RNA", 
  group.by = "cell_type",)
```

```{r sce_plot, fig.height = 8, fig.width = 7, eval = FALSE, include = FALSE}

scater::plotHeatmap(show_rownames = FALSE,show_legend = FALSE,
  agg_o3,
  features=sign_ids, 
  scale = TRUE,
  exprs_values = "logcounts",
  symmetric=TRUE,
  #colour_columns_by = "cell_type",
  center = TRUE,
  #colour = colors_z_score,
  order_columns_by = "cell_type", 
  show_colnames = TRUE)+
  theme(legend.position = "none")
```

```{r let_pheatmap_do_clustering, include = FALSE, eval = FALSE}

# use pheatmap to get a clustering and then extract it
vis_df_prep <- vis_df[,c(1,2,4)]
vis_df_prep <- tidyr::pivot_wider(
  vis_df_prep,
  names_from = "gene", 
  values_from = "zscore_gene")

vis_df_prep <- tibble::column_to_rownames(vis_df_prep, var = "cell_type")

# extract restuls
res <- pheatmap::pheatmap(t(vis_df_prep))
clusters <- cutree(res$tree_row, k = 10)
genes <- names(clusters)

# use order to factorize genesin original vis_df
vis_df_gene <- vis_df[res$tree_row$order,]
vis_df$gene <- factor(vis_df$gene, levels = base::unique(vis_df_gene$gene))
```


```{r e_base_plot, fig.height = 4, fig.width = 5, include = FALSE, eval = FALSE}
e_base_plot <- ggplot2::ggplot(
  vis_df,
  aes(x = cell_type, y = gene, fill = zscore_gene))+
  ggplot2::geom_tile()+
  scale_fill_gradientn(
    "z-score per gene", 
    colours = colors_z_score)
```

```{r e_theme_plot, fig.height = 7, fig.width = 4.5, include = FALSE, eval = FALSE}
e_theme_plot <- e_base_plot+
  ggplot2::theme_classic()+
  theme_all+
  ggplot2::theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.text.x = element_text(angle = 90)
  )+
  ggplot2::xlab("Cell types")+
  ggplot2::ylab("Conserved identity signature gene orthologue")+
  theme(legend.position = "none")

e_theme_plot
```

```{r e_legend_theme_plot, include = FALSE, eval = FALSE}
e_legend_theme_plot <- e_base_plot+
  ggplot2::theme_classic()+
  theme_all+
  ggplot2::theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.text.x = element_text(angle = 90)
  )+
  ggplot2::xlab("Cell types")+
  ggplot2::ylab("Conserved identity signature gene orthologue")
e_theme_legend <- ggpubr::get_legend(e_legend_theme_plot)

e_legend <- ggpubr::ggarrange(e_theme_legend)
e_legend
```

```{r e_plot_export, include = FALSE, eval = FALSE}
pdf(output_pdf_e_plot, width = 4.5, height = 7)
e_theme_plot
dev.off()
```

```{r e_legend_export, include = FALSE, eval = FALSE}
pdf(output_pdf_e_legend, width = 3, height = 3)
e_legend
dev.off()
```

# Utils

```{r sessioninfo}
utils::sessionInfo()
```

