---
title: "Figure 5"
date: '2024-09-23'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Generate Figure 5

#### Load objects

```{r seed, message = FALSE}
base::RNGkind(kind = "Mersenne-Twister")
set.seed(37)
```

```{r load_packages, message = FALSE}
library(SingleCellExperiment, quietly = TRUE)
library(ggpubr, quietly = TRUE)
library(tidyverse, quietly = TRUE)
library(Seurat, quietly = TRUE)
```

```{r define_basepath}
base_path <- "/omics/odcf/analysis/OE0538_projects_temp/DO-0008/data_temp/micromamba_test_rerun/data/"
```

```{r load_short}
#-------------------------------------------------------------------------------
#### REMAINING

# list of gene sets for HSPCs (conserved signature, conserved markers, etc.)
geneset_list_str <-base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/01_reclustering_own/01_gens/geneset_list_str"))

# data frame containing the correct resolutions for other datasets as visually determined
resolution_df_path <- base::paste0(
  base_path,
  "/metadata/scRNAseq/03_sce_analysis/reclustering_bm/reclustering_other_resolution.txt")

resolution_df <- utils::read.csv(
  file = resolution_df_path, 
  header = TRUE, 
  sep = ";", 
  check.names=FALSE, 
  stringsAsFactors=FALSE, 
  as.is=TRUE, 
  colClasses = "character")

# dataframe containing corrected pvalues for all tests
# corr_pval_df <-  readRDS(base::paste0(
#   base_path,
#   "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/09_crpv/all_corrected_pval"))

ensembl_sign <- readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/01_reclustering_own/02_endf/ensembl_sign_str"))

colors_path <- base::paste0(
  base_path,
  "/metadata/colors/colors.txt")

source("../source/colors.R")
source("../source/plotting.R")
```

```{r load_own}
#-------------------------------------------------------------------------------
#### OWN DATASET (Niche)

# annotated SCE
sce_str <- base::readRDS(base::paste0(
  base_path, 
  "/scRNAseq/main_analysis/sce_objects/02_sce_anno/10_anns/sce_str-10"))

# SCE with reclustering vectors for each gene set
sce_str_recl <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/01_reclustering_own/03_recl/sce_str"))

# dataframe of re-clustering scores for each gene set
score_df_str <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/01_reclustering_own/04_rcls/score_df_str"))

# dataframes of re-clustering scores after permutation comparing gene sets
psdf_str_mark_signrand <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/01_reclustering_own/06_perg/perm_score_df_mark_str"))
psdf_str_mmms_signrand <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/01_reclustering_own/06_perg/perm_score_df_mmms_str"))

# dataframes of re-clustering scores after background permutation
psdf_str_sign_rand <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/01_reclustering_own/05_psig/perm_score_df_str"))
```

```{r load_other2}
#-------------------------------------------------------------------------------
#### OTHER MOUSE DATASET 2 (Niche)
#### MUS_TIK_STROMAL

dataset_o2 <- "mus_tik_stromal"

# list of seurat objects with reclustering vectors for each gene set and resolution
seu_str_recl_list_o2 <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/03_recl/reclustered_mus_tik_stromal_list"))

# list of dataframes of re-clustering scores for each gene set and resolution
score_df_str_list_o2 <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/04_rcls/score_df_mus_tik_stromal_list"))

# dataframes of re-clustering scores after permutation comparing gene sets
psdf_str_mark_signrand_o2 <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/07_perg/perm_score_df_mark_mus_tik_stromal"))
psdf_str_mmms_signrand_o2 <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/07_perg/perm_score_df_mmms_mus_tik_stromal"))
psdf_str_mmms_markrand_o2 <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/07_perg/perm_score_df_mmms_mark_mus_tik_stromal"))

# perm score DF from background permutation (SIGN)
psdf_str_sign_rand_o2 <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/05_psig/perm_score_df_mus_tik_stromal"))
```

```{r selected_seu_o2}

resolution_df_o2 <- resolution_df[resolution_df$dataset == dataset_o2,]

selected_seu_list_o2 <- lapply(seu_str_recl_list_o2, function(seu_list){
  
  conslev_curr <- seu_list[[1]]@misc$used_genes
  
  resdf_temp <- resolution_df_o2[
    resolution_df_o2$conservation_level == conslev_curr,]
  res_curr <- resdf_temp$resolution
  
  seu_return <- seu_list[[res_curr]]
  return(seu_return)
  
})

rm(seu_str_recl_list_o2)
gc()
```

```{r load_other3}
#-------------------------------------------------------------------------------
#### OTHER MOUSE DATASET 3 (Niche)
#### li_all_stromal

dataset_o3 <- "li_all_stromal"

# list of seurat objects with reclustering vectors for each gene set and resolution
seu_str_recl_list_o3 <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/03_recl/reclustered_li_all_stromal_list"))

# list of dataframes of re-clustering scores for each gene set and resolution
score_df_str_list_o3 <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/04_rcls/score_df_li_all_stromal_list"))

# dataframes of re-clustering scores after permutation comparing gene sets
psdf_str_mark_signrand_o3 <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/07_perg/perm_score_df_mark_li_all_stromal"))
psdf_str_mmms_signrand_o3 <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/07_perg/perm_score_df_mmms_li_all_stromal"))
psdf_str_mmms_markrand_o3 <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/07_perg/perm_score_df_mmms_mark_li_all_stromal"))

# perm score DF from background permutation (SIGN)
psdf_str_sign_rand_o3 <- base::readRDS(base::paste0(
  base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/05_psig/perm_score_df_li_all_stromal"))

# perm score DF from background permutation (MARK)
# psdf_str_mark_rand_o1 <- base::readRDS(base::paste0(
#   base_path,
#   "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/02_reclustering_other/07_pmrk/perm_score_df_li_all_stromal"))
```

```{r selected_seu_o3}

resolution_df_o3 <- resolution_df[resolution_df$dataset == dataset_o3,]

selected_seu_list_o3 <- lapply(seu_str_recl_list_o3, function(seu_list){
  
  conslev_curr <- seu_list[[1]]@misc$used_genes
  
  resdf_temp <- resolution_df_o3[
    resolution_df_o3$conservation_level == conslev_curr,]
  res_curr <- resdf_temp$resolution
  
  seu_return <- seu_list[[res_curr]]
  return(seu_return)
  
})

rm(seu_str_recl_list_o3)
gc()
```

```{r params}
# global params
source("determine_params.R")

cts_exclude <- c("Lymphoid ECs", "Chondrocytes")
print(cts_exclude)
```

```{r export_paths}
output_pdf_a_plot <- base::paste0(
  base_path, "/manuscript1/figure5/figure5_a.pdf")
output_pdf_a_legend <- base::paste0(
  base_path, "/manuscript1/figure5/figure5_a_legend.pdf")
output_pdf_b_plot <- base::paste0(
  base_path, "/manuscript1/figure5/figure5_b.pdf")
output_pdf_cd_plot <- base::paste0(
  base_path, "/manuscript1/figure5/figure5_cd.pdf")
output_pdf_e_plot <- base::paste0(
  base_path, "/manuscript1/figure5/figure5_e.pdf")
output_pdf_e_legend <- base::paste0(
  base_path, "/manuscript1/figure5/figure5_e_legend.pdf")
```

## Obtain data

```{r obtain_data}
#print(sort(geneset_list_str$Fibroblasts$conserved_signature))
#print(sort(geneset_list_str$`Sinusoidal ECs`$conserved_signature))
#print(sort(geneset_list_str$`Sinusoidal/Venular ECs`$conserved_signature))
#print(sort(geneset_list_str$`Arteriolar/Capillary ECs`$conserved_signature))
```

## A Expression heat map

Expression of conserved signature genes in all mouse species across HSPC
cell types (scaled by z-score per gene).

### Prep

```{r a_prep1}
# get all unique signature genes
all_sign_genes <- vector()
for(i in 1:length(geneset_list_str)){
  sign_genes <- geneset_list_str[[i]]$conserved_signature
  all_sign_genes <- c(all_sign_genes, sign_genes)
}
all_sign_genes_str <- base::unique(all_sign_genes)
```

```{r a_prep2}
# aggregate per cell type and species
cts_str <- as.list(names(geneset_list_str))

sce_str$species_pub <- vector(length = ncol(sce_str))
sce_str$species_pub[sce_str$Species_ID == "mmus"] <- "BL6"
sce_str$species_pub[sce_str$Species_ID == "mcas"] <- "CAST"
sce_str$species_pub[sce_str$Species_ID == "mspr"] <- "SPRET"
sce_str$species_pub[sce_str$Species_ID == "mcar"] <- "CAROLI"
sce_str$species_pub <- factor(sce_str$species_pub, 
                              levels = names(col_spc_pub))
# remove cell types to be removed
cts_str <- cts_str[!cts_str %in% cts_exclude]
cts_list <- as.list(cts_str)

sce_str <- sce_str[,!sce_str$celltypes %in% cts_exclude]
sce_str$celltypes <- factor(sce_str$celltypes, levels(sce_str$celltypes)[
  !levels(sce_str$celltypes) %in% cts_exclude])

# separate SCE objects by cell type, then aggregate the counts to obtain
# the average counts per species for each cell type separately
# use the multi-batchnorm-corrected logcounts 

agg_str_list <- lapply(cts_str, function(ct){
  
  sce_ct <- sce_str[,sce_str$celltypes == ct]
  agg_str <- scuttle::aggregateAcrossCells(
    sce_ct, 
    id=colData(sce_ct)[,c("species_pub")], 
    statistics = "mean",
    use.assay.type = "logcounts") #multibatchnorm
  
  return(agg_str)
})
names(agg_str_list) <- names(geneset_list_str)
```

```{r a_prep3}
expr_df_list_str <- lapply(agg_str_list, function(agg_str){
  
  # get current cell type and relevant signature genes
  ct_curr <- agg_str$celltypes[1]
  print(ct_curr)
  
  sign_genes_ct <- geneset_list_str[[
    which(names(geneset_list_str) == ct_curr)]]$conserved_signature
  print(head(sign_genes_ct))
  
  # get a df from average logcounts per ct and species
  vis_df <- base::as.data.frame(
    t(SummarizedExperiment::assays(agg_str)$logcounts))
  
  # subset to all signature genes of this tissue
  vis_df <- vis_df[,colnames(vis_df) %in% all_sign_genes_str]
  
  # change df format 
  vis_df <- tibble::rownames_to_column(vis_df, "species")
  vis_df <- tidyr::pivot_longer(vis_df, 
                                cols = c(2:ncol(vis_df)),
                                values_to = "expression",
                                names_to = "gene")
  
  # add ct info
  vis_df$cell_type <- base::rep(ct_curr)

  # add info on genes that are signature genes of this ct
  vis_df$is_sign_gene <- vector(length = nrow(vis_df))
  vis_df$is_sign_gene[vis_df$gene %in% sign_genes_ct] <- TRUE

  # scale expression of all genes per cell type 
  # for ordering of genes ONLY
  vis_df$scaled_expression_per_ct <- base::scale(vis_df$expression)
  vis_df <- vis_df[base::order(vis_df$scaled_expression_per_ct, 
                               decreasing = TRUE),]

  return(vis_df)
})

# bind dfs 
agg_str_df <- dplyr::bind_rows(expr_df_list_str)
```

```{r a_prep4}
# scale expression per gene using z-scores
agg_str_df$zscore_gene <- vector(length = nrow(agg_str_df), mode = "numeric")
for(g in base::unique(agg_str_df$gene)){
  
  expr_vector_g <- agg_str_df[agg_str_df$gene == g,]$expression

  # calculate z score per gene
  z_score_g <- (expr_vector_g - base::mean(expr_vector_g))/
    stats::sd(expr_vector_g)
  agg_str_df$zscore_gene[agg_str_df$gene == g] <- z_score_g
}
```

```{r a_prep5}
# get a well ordered list of unique genes for factorization (after order above)
all_sign_genes_for_factor <- base::unique(agg_str_df$gene[
  agg_str_df$is_sign_gene == TRUE])

# factorize
agg_str_df$gene <- factor(agg_str_df$gene,
                          levels = all_sign_genes_for_factor)
agg_str_df$species <- factor(agg_str_df$species,
                             levels = base::rev(names(col_spc_pub)))

#print(levels(agg_str_df$gene))
```

```{r a_prep6}
interest_genes <- c("Cxcl12",
                    "Marcks",
                    "Kitl",
                    "Cd34",
                    "Efnb2",
                    "Mmp13",
                    "Tie1",
                    "Actn4",
                    "Lrg1")

agg_str_df$is_gene <- vector(length = nrow(agg_str_df))
agg_str_df$is_gene[agg_str_df$gene %in% interest_genes] <- "TRUE"
agg_str_df$is_gene <- as.logical(agg_str_df$is_gene)

print(levels(agg_str_df$gene)[levels(agg_str_df$gene) %in% interest_genes])
```

#### Plot

```{r a_base_plot}
a_base_plot <- ggplot2::ggplot(
  agg_str_df,
  aes(y = species, x = gene, fill = zscore_gene))+
  ggplot2::geom_tile()+
  ggplot2::facet_grid(rows = vars(agg_str_df$cell_type))
```

```{r a_theme_plot}
a_theme_plot <- a_base_plot+
  theme_all+
  ggplot2::theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    plot.title = element_text(
      vjust = 0.5,
      hjust = 0.5,
      color = col_cons_long["conserved identity signature"]),
    axis.title.y = element_blank(),
    axis.line.x = element_blank(),
    axis.text.y = element_text(
      angle = 0,
      vjust = 0.5,
      hjust = 1),
    strip.text.y = element_text(
      face = axis_text_face,
      size = axis_text_size,
      color = axis_text_color,
      angle = 0,
      vjust = 0.5,
      hjust = 0),
    strip.background.y = element_blank())+
  ggplot2::ggtitle("Conserved identity signature genes")+
  ggplot2::scale_fill_gradientn(
    "z-score per gene", 
    colours = colors_z_score,
    breaks = c(-5, -2.5,  0, 2.5, 5), 
    limits = c(-5,5))+
  ggplot2::scale_x_discrete(
    position =  "bottom",
    breaks = agg_str_df$gene[agg_str_df$is_gene])

a_legend <- ggpubr::get_legend(a_theme_plot)
a_legend <- ggpubr::ggarrange(a_legend)
```

#### Assemble

```{r a_assemble, fig.width = 5, fig.height = 5}
a_plot <- ggpubr::ggarrange(
  a_theme_plot + theme(legend.position = "none",
                       plot.title = element_blank())
)
a_plot
```

```{r a_plot_export}
pdf(output_pdf_a_plot, width = 6.5, height = 4)
a_plot
dev.off()
```

```{r a_legend_export}
pdf(output_pdf_a_legend, width = 3, height = 3)
a_legend
dev.off()
```

## BCD Re-clusterung Heatmaps 

Heatmaps of own re-clustered HSPC mouse data using signature genes/
conserved markers/BL6 markers/nDGEs.

For other datasets, this requires BLINDED RE-ANNOTATION BASED ON
CLUSTER MARKER GENES

### Prepare

```{r prepare_annotate_o2}

# mus tik stromal

seu_sign_o2 <- selected_seu_list_o2$seu_sign
reclustered_o2 <- base::unique(seu_sign_o2$seurat_clusters)

clust_list_o2 <- as.list(reclustered_o2)

ct_markers_o2 <- lapply(clust_list_o2, function(x){
  markers <- Seurat::FindMarkers(seu_sign_o2,
                                 test.use = "wilcox",
                                 ident.1 = x)
  markers <- markers[base::order(base::abs(markers$avg_log2FC), 
                                 decreasing=TRUE),]
  markers$which_cluster <- base::rep(x, nrow(markers))
  markers <- rownames_to_column(markers, var = "genes")
  
  return(markers)
})
names(ct_markers_o2) <- reclustered_o2

ct_markers_df_o2 <- bind_rows(ct_markers_o2)
head(ct_markers_df_o2)

```

```{r prepare_annotate_o2_ec}

# all ECs
seu_sign_o2_ec <- seu_sign_o2[,seu_sign_o2$seurat_clusters %in%
                                c("6", "1", "2", "3")]

reclustered_o2_ec <- base::unique(seu_sign_o2_ec$seurat_clusters)

clust_list_o2_ec <- as.list(reclustered_o2_ec)

ct_markers_o2_ec <- lapply(clust_list_o2_ec, function(x){
  markers <- Seurat::FindMarkers(seu_sign_o2_ec,
                                 test.use = "wilcox",
                                 ident.1 = x)
  markers <- markers[base::order(base::abs(markers$avg_log2FC), 
                                 decreasing=TRUE),]
  markers$which_cluster <- base::rep(x, nrow(markers))
  markers <- rownames_to_column(markers, var = "genes")
  
  return(markers)
})
names(ct_markers_o2_ec) <- reclustered_o2_ec

ct_markers_df_o2_ec <- bind_rows(ct_markers_o2_ec)
head(ct_markers_df_o2_ec)

```

```{r prepare_annotate_o3}

seu_sign_o3 <- selected_seu_list_o3$seu_sign
reclustered_o3 <- base::unique(seu_sign_o3$seurat_clusters)

clust_list_o3 <- as.list(reclustered_o3)

ct_markers_o3 <- lapply(clust_list_o3, function(x){
  markers <- Seurat::FindMarkers(seu_sign_o3,
                                 test.use = "wilcox",
                                 ident.1 = x)
  markers <- markers[base::order(base::abs(markers$avg_log2FC), 
                                 decreasing=TRUE),]
  markers$which_cluster <- base::rep(x, nrow(markers))
  markers <- rownames_to_column(markers, var = "genes")
  
  # need to add only ensembl gene symbols as I will not be using others
  markers$symbol <- markers$genes
  markers$symbol <- ensembl_sign$HSAPIENS_SYMBOL[base::match(markers$genes, ensembl_sign$ENSG_ID)]
  return(markers)
})
names(ct_markers_o3) <- reclustered_o3

ct_markers_df_o3 <- bind_rows(ct_markers_o3)
head(ct_markers_df_o3)

```

#### Others -3

```{r umap_others_3}
DimPlot(seu_sign_o3, dims = c(1,2), group.by = "seurat_clusters")
DimPlot(seu_sign_o3, dims = c(1,2), group.by = "orig.ident")

seu_sign_o3$annotated_reclustered <- vector(length = ncol(seu_sign_o3))
```

order: 6, 1, 2, 0, 7, 4, 5, 3

```{r others3_cluster6}
nr <- "6"
markers <- ct_markers_o3[[nr]]$gene[
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05]

ct_markers_o3[[nr]][
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o3, markers[1:6])
FeaturePlot(seu_sign_o3, markers[7:12])
FeaturePlot(seu_sign_o3, markers[13:18])

FeaturePlot(seu_sign_o3, "ENSG00000145824") #CXCL14
FeaturePlot(seu_sign_o3, "ENSG00000164176") #EDIL3
FeaturePlot(seu_sign_o3, "ENSG00000154175") #ABI3BP
FeaturePlot(seu_sign_o3, "ENSG00000109610") #SOD3
FeaturePlot(seu_sign_o3, "ENSG00000163359") #COL6A3
FeaturePlot(seu_sign_o3, "ENSG00000141338") #ABCA8

# CXCL14  human niche stromal cells https://ashpublications.org/blood/article/142/1/73/495206/Characterization-of-the-bone-marrow-niche-in
#         dermal fibro, anti CXCL12 https://journal-inflammation.biomedcentral.com/articles/10.1186/s12950-015-0109-9
#         fibro growth factor https://doi.org/10.1073/pnas.0813144106
#         human umbilical MSCs https://www.nature.com/articles/jhg201463

# ABCA8   cancer fibro https://www.sciencedirect.com/science/article/pii/S1535947624001336
#         Ccl11+ fibroblasts https://www.sciencedirect.com/science/article/pii/S2352345X24000286
#         high density lipo protein efflux https://www.ahajournals.org/doi/pdf/10.1161/atvbaha.117.309574
#         fibro subset in intestine https://pmc.ncbi.nlm.nih.gov/articles/PMC8604730/

# EDIL3   osteo diff https://pmc.ncbi.nlm.nih.gov/articles/PMC5705136/
#         EMT regulator https://www.nature.com/articles/s41420-020-00322-x
#         CAR and osteo cells, niche https://pmc.ncbi.nlm.nih.gov/articles/PMC5617665/
#         osteogenic activity https://www.cell.com/iscience/fulltext/S2589-0042(24)00019-1

# ABI3BP  MSC, quiescence, osteogenic and adipogenic differentiation. https://pmc.ncbi.nlm.nih.gov/articles/PMC3775980/
#         may be non-proliferative https://www.frontiersin.org/journals/cardiovascular-medicine/articles/10.3389/fcvm.2019.00023/full
#         MSc and osteoporosis, might switch BMSCs between proliferative and differentiating state https://pubmed.ncbi.nlm.nih.gov/38164361/
#         MSC https://www.sciencedirect.com/science/article/pii/S2352304219300169
#         articular cartilage https://pmc.ncbi.nlm.nih.gov/articles/PMC4089504/
#         articular cartilage https://pmc.ncbi.nlm.nih.gov/articles/PMC7402331/
#         articular cartilage origin paper https://www.sciencedirect.com/science/article/pii/S1063458410003328

# SOD3    chondrogenesis https://pubmed.ncbi.nlm.nih.gov/30654942/
#         osteogenesis https://pubmed.ncbi.nlm.nih.gov/38141889/
#         MSC maintenance and differentiation review, chondrogenesis https://cellandbioscience.biomedcentral.com/articles/10.1186/s13578-020-00386-3
#         cartilage https://www.sciencedirect.com/science/article/pii/S1568163720303846

# COL6A3  osteoadipo diff https://pubmed.ncbi.nlm.nih.gov/38270688/
#         related to osteoprosos https://pubmed.ncbi.nlm.nih.gov/32999266/
#         osteoblast matrix https://stemcellres.biomedcentral.com/articles/10.1186/s13287-015-0223-x
#         in chondros https://www.biorxiv.org/content/10.1101/2022.12.19.520461v1.full

seu_sign_o3$annotated_reclustered[seu_sign_o3$seurat_clusters == nr] <- "Osteo-Chondro-like CAR"
```

```{r others3_cluster1}
nr <- "1"
markers <- ct_markers_o3[[nr]]$gene[
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05]

ct_markers_o3[[nr]][
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o3, markers[1:6])
FeaturePlot(seu_sign_o3, markers[7:12])
FeaturePlot(seu_sign_o3, markers[13:18])

FeaturePlot(seu_sign_o3, "ENSG00000163430") #FSTL1
FeaturePlot(seu_sign_o3, "ENSG00000196549") #MME
FeaturePlot(seu_sign_o3, "ENSG00000257017") #HP
FeaturePlot(seu_sign_o3, "ENSG00000104067") #TJP1
FeaturePlot(seu_sign_o3, "ENSG00000152583") #SPARCL1
FeaturePlot(seu_sign_o3, "ENSG00000166147") #FBN1
FeaturePlot(seu_sign_o3, "ENSG00000146374") #RSPO3
FeaturePlot(seu_sign_o3, "ENSG00000140092") #FBLN5

# FSTL1, MME, HP, TJP1, SPARCL1, FBN1, RSPO3

# FSTL1   suppressed osteo during infl https://pubmed.ncbi.nlm.nih.gov/35026647/
#         MSC protection https://pmc.ncbi.nlm.nih.gov/articles/PMC6330478/
#         regulator of proliferation and differentiation in MSCs https://stemcellres.biomedcentral.com/articles/10.1186/s13287-022-03042-4
#         enhance MSC differentiation to chondrocytes https://www.sciencedirect.com/science/article/pii/S156757692030059X
#         adipogenesis https://pmc.ncbi.nlm.nih.gov/articles/PMC2849861/

# HP      haptoglobin, adipogenesis, https://academic.oup.com/edrv/article/37/4/403/2567100

# MME     osteo and adipo genesis https://pmc.ncbi.nlm.nih.gov/articles/PMC9526692/ # CD92
#         osteo and adipo genesis https://pubmed.ncbi.nlm.nih.gov/24239963/
#         enriched in osteoblasts https://pmc.ncbi.nlm.nih.gov/articles/PMC11301355/
#         preadipocytes https://www.nature.com/articles/s42003-023-04504-y

# TJP1    tight junction protein, also called ZO1
#         MSC migration https://www.nature.com/articles/s41420-023-01793-4
#         can be expressed in MSC https://www.cell.com/cms/10.1016/j.stemcr.2015.02.002/attachment/f3ad90ab-28e4-4980-990b-e6f35611c440/mmc3.pdf
#         can be expressed in MSC https://pmc.ncbi.nlm.nih.gov/articles/PMC8275434/

# SPARCL1 MSCs/Progs also related to DCN https://pmc.ncbi.nlm.nih.gov/articles/PMC10192814/
#         skeletal stem cell https://www.biorxiv.org/content/10.1101/2020.06.17.156836v1.full i think this is just biorxiv of the first one
#         also expressed in ECs, especially sinusoidal https://www.nature.com/articles/s41467-021-27161-3
#         negative regulator of adipo https://pmc.ncbi.nlm.nih.gov/articles/PMC8654481/
#         mouse muscle cell diff https://www.nature.com/articles/s41419-019-2049-4
#         chondros in arthritis https://www.sciencedirect.com/science/article/pii/S2214031X24000287
#         liver cancer fibros https://www.sciencedirect.com/science/article/pii/S1936523320304733

# FBN1    related to TGFÎ², MSC diff, https://pubmed.ncbi.nlm.nih.gov/26610678/
#         MSC niche/decisions https://www.frontiersin.org/journals/cell-and-developmental-biology/articles/10.3389/fcell.2023.1302285/full
#         MSC activitiy regulation https://onlinelibrary.wiley.com/doi/full/10.1002/jbmr.2598

# RSPO3   osteoblast diffhttps://www.nature.com/articles/s41467-021-25124-2
#         niche https://pmc.ncbi.nlm.nih.gov/articles/PMC9979769/
#         bone density https://pmc.ncbi.nlm.nih.gov/articles/PMC9681208/

seu_sign_o3$annotated_reclustered[seu_sign_o3$seurat_clusters == nr] <- "Osteo-like CAR"
```

```{r others3_cluster2}
nr <- "2"
markers <- ct_markers_o3[[nr]]$gene[
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05]

ct_markers_o3[[nr]][
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o3, markers[1:6])
FeaturePlot(seu_sign_o3, markers[7:12])
FeaturePlot(seu_sign_o3, markers[13:18])

FeaturePlot(seu_sign_o3, "ENSG00000107562") #Cxcl12
FeaturePlot(seu_sign_o3, "ENSG00000049130") #Kitl

FeaturePlot(seu_sign_o3, "ENSG00000113721") #PDGFRB
FeaturePlot(seu_sign_o3, "ENSG00000166147") #FBN1
FeaturePlot(seu_sign_o3, "ENSG00000168542") #COL3A1
FeaturePlot(seu_sign_o3, "ENSG00000182326") #C1S
FeaturePlot(seu_sign_o3, "ENSG00000152583") #SPARCL1
FeaturePlot(seu_sign_o3, "ENSG00000141338") #ABCA8

# PDGFRB, FBN1, COL3A1, C1S, CYP1B1, SPARCL1,  ABCA8

# Pdgfrb  fibroblast activation https://pubmed.ncbi.nlm.nih.gov/23748876/
#         migration of MSC, but osteo inhibition https://pubmed.ncbi.nlm.nih.gov/18410236/
#         fibrotic BM https://haematologica.org/article/view/9982
#         

# FBN1    related to TGFÎ², MSC diff, https://pubmed.ncbi.nlm.nih.gov/26610678/
#         MSC niche/decisions https://www.frontiersin.org/journals/cell-and-developmental-biology/articles/10.3389/fcell.2023.1302285/full
#         MSC activitiy regulation https://onlinelibrary.wiley.com/doi/full/10.1002/jbmr.2598

# COL3A1  osteoblast diff, bone related https://pubmed.ncbi.nlm.nih.gov/24626604/
#         fetal MSCs, co-expressed with Dcn, and Clec3b, Col1a1, Pdgfra https://www.nature.com/articles/s41467-022-28775-x
#         human cancer fibros https://www.frontiersin.org/journals/immunology/articles/10.3389/fimmu.2023.1313536/full
#         human fibrosis https://www.sciencedirect.com/science/article/pii/S0378111919304652
#         mouse pdpn-expressing cells in fibrosis https://academic.oup.com/discovimmunology/article/1/1/kyac008/6835131#google_vignette

# C1S     chondrocytes, ossification https://pubmed.ncbi.nlm.nih.gov/8082118/
#         cartilage remodelling https://pubmed.ncbi.nlm.nih.gov/8766156/

# SPARCL1 MSCs/Progs also related to DCN https://pmc.ncbi.nlm.nih.gov/articles/PMC10192814/
#         skeletal stem cell https://www.biorxiv.org/content/10.1101/2020.06.17.156836v1.full i think this is just biorxiv of the first one
#         also expressed in ECs, especially sinusoidal https://www.nature.com/articles/s41467-021-27161-3
#         negative regulator of adipo https://pmc.ncbi.nlm.nih.gov/articles/PMC8654481/
#         mouse muscle cell diff https://www.nature.com/articles/s41419-019-2049-4
#         chondros in arthritis https://www.sciencedirect.com/science/article/pii/S2214031X24000287
#         liver cancer fibros https://www.sciencedirect.com/science/article/pii/S1936523320304733

# ABCA8   cancer fibro https://www.sciencedirect.com/science/article/pii/S1535947624001336
#         Ccl11+ fibroblasts https://www.sciencedirect.com/science/article/pii/S2352345X24000286
#         high density lipo protein efflux https://www.ahajournals.org/doi/pdf/10.1161/atvbaha.117.309574
#         fibro subset in intestine https://pmc.ncbi.nlm.nih.gov/articles/PMC8604730/

seu_sign_o3$annotated_reclustered[seu_sign_o3$seurat_clusters == nr] <- "Fibro-like CAR"
```

```{r others3_cluster0}
nr <- "0"
markers <- ct_markers_o3[[nr]]$gene[
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05]

ct_markers_o3[[nr]][
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o3, markers[1:6])
FeaturePlot(seu_sign_o3, markers[7:12])
FeaturePlot(seu_sign_o3, markers[13:18])

# EMP1, ID1, DEPP1, LMNA, FABP4, OSR2, PPP1R15A, CCN2, OLFML3, RARRES2

# EMP1  cancer fibros https://www.frontiersin.org/journals/immunology/articles/10.3389/fimmu.2023.1313536/full
#       fibro https://onlinelibrary.wiley.com/doi/10.1002/jbm4.10795
#       general cell adhesion and growth, cancer gene

# ID1   required for HSC niche https://pubmed.ncbi.nlm.nih.gov/19478045/
#       osteogenic https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-8-70

# DEPP1 committed pre-adipocytes https://www.nature.com/articles/s41413-020-0097-0
#       cancer fibroblasts https://pmc.ncbi.nlm.nih.gov/articles/PMC10140349/
#       autoophagy https://faseb.onlinelibrary.wiley.com/doi/pdf/10.1096/fj.202100896R

# LMNA  osteoporosis https://pmc.ncbi.nlm.nih.gov/articles/PMC5978267/
#       migration and osteogenesis https://pmc.ncbi.nlm.nih.gov/articles/PMC7348862/
#       anit-chondro diff https://www.sciencedirect.com/science/article/pii/S187350611300086X

# FABP4 adipogenic https://pmc.ncbi.nlm.nih.gov/articles/PMC8910760/
#       senscence in MSC, fatty acid metabolic activity https://pmc.ncbi.nlm.nih.gov/articles/PMC8891618/
#       marrow adipocytes https://www.sciencedirect.com/science/article/pii/S0006497118500550
#       adipogenesis status https://biosignaling.biomedcentral.com/articles/10.1186/s12964-023-01231-z
#       leukemia bone marrow MSC https://www.nature.com/articles/s41375-019-0568-8
#       adipocyte control https://pmc.ncbi.nlm.nih.gov/articles/PMC9867004/

# OSRs  uncommited adipose progenitors https://pmc.ncbi.nlm.nih.gov/articles/PMC10604679/

# PPP1R15A  related to weight gain https://www.nature.com/articles/s41598-019-39562-y 
#           adipocytic signalling https://pmc.ncbi.nlm.nih.gov/articles/PMC9067510/
#           also called GADD34
#           related to autphagy and stress

# RARRES2 osteosarcoma stemness, interaction with macrophages https://www.nature.com/articles/s41598-024-52738-5
#         interaction of SPP1 cells with macrophages https://www.nature.com/articles/s41467-022-29366-6
#         global adipokine https://pubmed.ncbi.nlm.nih.gov/19303973/
#         related to obesity and adipocytes https://www.sciencedirect.com/science/article/pii/S2589004223005722
#         expression in adipocytes https://www.nature.com/articles/s41598-020-70625-7
#         

seu_sign_o3$annotated_reclustered[seu_sign_o3$seurat_clusters == nr] <- "Adipo-like CAR"
```

```{r others3_cluster7}
DimPlot(seu_sign_o3, dims = c(1,2), group.by = "seurat_clusters")

nr <- "7"
markers <- ct_markers_o3[[nr]]$gene[
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05]

ct_markers_o3[[nr]][
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o3, markers[1:6])
FeaturePlot(seu_sign_o3, markers[7:12])
FeaturePlot(seu_sign_o3, markers[13:18])

# CPE, SP7, PTPRD, RUNX2, PALS2, OMD, MMP13, PTH1R, CD9, TGM2

seu_sign_o3$annotated_reclustered[seu_sign_o3$seurat_clusters == nr] <- "Osteo-Chondro"
```

```{r others3_cluster5}
nr <- "5"
markers <- ct_markers_o3[[nr]]$gene[
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05]

ct_markers_o3[[nr]][
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o3, markers[1:6])
FeaturePlot(seu_sign_o3, markers[7:12])
FeaturePlot(seu_sign_o3, markers[13:18])

FeaturePlot(seu_sign_o3, "ENSG00000149564") #ESAM
FeaturePlot(seu_sign_o3, "ENSG00000174059") #CD34
FeaturePlot(seu_sign_o3, "ENSG00000102760") #RGCC
FeaturePlot(seu_sign_o3, "ENSG00000066056") #TIE1
FeaturePlot(seu_sign_o3, "ENSG00000130300") #PLVAP

FeaturePlot(seu_sign_o3, "ENSG00000149564") #CAVIN2
FeaturePlot(seu_sign_o3, "ENSG00000135218") #CD36
FeaturePlot(seu_sign_o3, "ENSG00000148400") #Notch1

# some classical EC markers
# other specufuc genes: TSPAN13, CAPG, MGST1, CD36, GPX1, PPP1R14B, CAVIN2, TXN

# Cd36    also related to fatty acid, FA uptake, angiogenesis, response to injury

# TSPAN13 dysfunctional Endothelium https://pubmed.ncbi.nlm.nih.gov/36140731/

# CAPG    motility and shear stress in ECs https://pubmed.ncbi.nlm.nih.gov/12754261/
#         shear stress exposure https://journals.physiology.org/doi/full/10.1152/ajpcell.00218.2005

# MGST1   antioxidant https://pmc.ncbi.nlm.nih.gov/articles/PMC10586476/
#         oxidative stress https://www.ahajournals.org/doi/10.1161/HYPERTENSIONAHA.119.13472
    
# GPX1    oxidative stress https://pubmed.ncbi.nlm.nih.gov/12829378/ in ECs
#         endothelial dysfunction, oxidative stress
#         oxidant stress

# PPP1R14B  endo migration https://www.frontiersin.org/journals/genetics/articles/10.3389/fgene.2021.763561/full

# CAVIN2   EC function, stability, caveolae

# TXN     antioxidant (thioredoxin)
#         redox sensor (https://pubmed.ncbi.nlm.nih.gov/24130874/)

# RGCC capillary
# CD34 a little bit

# TIE1, PLVAP, ESAM, CAVIN2, = classical EC markers

seu_sign_o3$annotated_reclustered[seu_sign_o3$seurat_clusters == nr] <- "EC1"
```

```{r others3_cluster3}
nr <- "3"
markers <- ct_markers_o3[[nr]]$gene[
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05]

ct_markers_o3[[nr]][
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o3, markers[1:6])
FeaturePlot(seu_sign_o3, markers[7:12])
FeaturePlot(seu_sign_o3, markers[13:18])

# PECAM1, FLI1, ICAM2, CRIP1, FCER1G, SH2D3C, GNGT2, LYN, IFI30, ARPC3, PLEKHO1, CRIP1

FeaturePlot(seu_sign_o3, "ENSG00000261371") #PECAM1 # way more in 3
FeaturePlot(seu_sign_o3, "ENSG00000151702") #FLI1 # more in 3
FeaturePlot(seu_sign_o3, "ENSG00000179776") #CDH5
FeaturePlot(seu_sign_o3, "ENSG00000184113") #CLDN5
FeaturePlot(seu_sign_o3, "ENSG00000125266") #EFNB2
FeaturePlot(seu_sign_o3, "ENSG00000102755") #FLT1

# PCAM1 FLI1 ICAM2 typical EC
# Fcer1g, IFI30 inflammation

# CRIP1 prolif

# SH2D3C https://www.ahajournals.org/doi/10.1161/CIRCULATIONAHA.120.052318 capillary signalling
# GNGT2: signalling

# ARPC3 cytoskeleton

# Lyn   EC barrier function https://pmc.ncbi.nlm.nih.gov/articles/PMC3862279/
#       healthy bone marrow EC https://www.nature.com/articles/s41467-023-36824-2 anti-inflammatory

# IFI30   sprouting angiogenesis zebrafish https://pmc.ncbi.nlm.nih.gov/articles/PMC9325957/
#         niche interaction https://pmc.ncbi.nlm.nih.gov/articles/PMC8302694/
#         upreglated by inflammation https://www.sciencedirect.com/science/article/pii/S0161589015004320


seu_sign_o3$annotated_reclustered[seu_sign_o3$seurat_clusters == nr] <- "EC3"
```

```{r others3_cluster4}
nr <- "4"
markers <- ct_markers_o3[[nr]]$gene[
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05]

ct_markers_o3[[nr]][
  !is.na(ct_markers_o3[[nr]]$symbol) & 
    ct_markers_o3[[nr]]$avg_log2FC > 0 &
    ct_markers_o3[[nr]]$p_val_adj < 0.05,]

FeaturePlot(seu_sign_o3, markers[1:6])
FeaturePlot(seu_sign_o3, markers[7:12])
FeaturePlot(seu_sign_o3, markers[13:18])

# NFKBIA, SRGN, PECAM1, SULF2

seu_sign_o3$annotated_reclustered[seu_sign_o3$seurat_clusters == nr] <- "EC2"
```

#### Others 2

```{r umap_others_2}
DimPlot(seu_sign_o2, dims = c(1,2), group.by = "seurat_clusters")
DimPlot(seu_sign_o2, dims = c(1,2), group.by = "batch")

seu_sign_o2$annotated_reclustered <- vector(length = ncol(seu_sign_o2))
```

Order: 0, 5, 4, 7, 10, 9, 11, 8, 12, 13, 6, 1, 2, 3

```{r others2_cluster0}
nr <- "0"
markers <- ct_markers_o2[[nr]]$genes[
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

ct_markers_o2[[nr]][
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])
FeaturePlot(seu_sign_o2, markers[13:18])

FeaturePlot(seu_sign_o2, "Cxcl12")

# Adipo/CAR

# Mgp, Hp, Notch3, Ebf3, Dpep1, Cxcl12, Cxcl14, Cyp1b1, Esm1

# Mgp - adipogenesis
# Hp - adipocytes
# Notch3 - osteoblastic - lymphoid niche
# Dpep1 - adipogenic

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Adipo/CAR1, pbs"
# CHECK BATCH
```

```{r others2_cluster5}
nr <- "5"
markers <- ct_markers_o2[[nr]]$genes[
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])
FeaturePlot(seu_sign_o2, markers[13:18])

FeaturePlot(seu_sign_o2, "Cxcl12")

# Osteo/CAR

# Alpl, Limch1, Lrp4, Dlx5, Olfml3, Sp7, Pth1r, Satb2

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Osteo/CAR"
```

```{r others2_cluster4}
nr <- "4"
markers <- ct_markers_o2[[nr]]$genes[
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

ct_markers_o2[[nr]][
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])
FeaturePlot(seu_sign_o2, markers[13:18])

FeaturePlot(seu_sign_o2, "Cxcl12")

# Adipo-CAR

# Angptl1, Kitl, Esm1, Igfbp5, Mme, C1s1, Dpep1, Vegfc

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Adipo/CAR3, non-pbs"
# CHECK BATCH
```

```{r others2_cluster7}
nr <- "7"
markers <- ct_markers_o2[[nr]]$genes[
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])
FeaturePlot(seu_sign_o2, markers[13:18])

# Mmp13, Slit2, Lrp4, Ptgis, Gpc1, Mrc2, s100a6, Satb2, Pth1r, Cfh

# Sp7, Runx2, Alpl, Satb2, Ptgis, S100a6, are osteogenic
# Mmp13 Pth1r, Tpm1 are also chondro

# Tpm1 :  mouse chondro https://academic.oup.com/rheumatology/advance-article/doi/10.1093/rheumatology/keae587/7829155
#         human chondro https://www.nature.com/articles/s41419-021-03832-3
#         human chondro/OA https://pmc.ncbi.nlm.nih.gov/articles/PMC9713432/
#         mouse tendon development https://www.cell.com/developmental-cell/fulltext/S1534-5807(23)00072-2?dgcid=raven_jbs_aip_email
# 

# Mmmp13: human chondros https://pmc.ncbi.nlm.nih.gov/articles/PMC5146981/
#         human chondros https://www.sciencedirect.com/science/article/pii/S0021925820366953
#         mouse cartilage ans arthritis https://pubmed.ncbi.nlm.nih.gov/19950295/
#         mouse chondrocytes https://www.nature.com/articles/s41598-019-52125-5

# Pth1r   mouse cartilage but from osteos https://pubmed.ncbi.nlm.nih.gov/25196529/
#         mouse hypertrophic chondros https://pubmed.ncbi.nlm.nih.gov/11181543/
#         chondro diff https://elifesciences.org/articles/82142
#         osteoblasts https://www.nature.com/articles/s41413-023-00255-y
#         osteo diff https://elifesciences.org/articles/83345

# Ptgis   osteo https://www.sciencedirect.com/science/article/pii/S0006497119386240
#         endosteal niche https://stemcellsjournals.onlinelibrary.wiley.com/doi/10.1002/stem.3438

# S100a6  chondro apoptosis https://pubmed.ncbi.nlm.nih.gov/33942537/
#         MSC https://biomarkerres.biomedcentral.com/articles/10.1186/s40364-023-00515-3

# Satb2   mostly osteo, some chondro https://pmc.ncbi.nlm.nih.gov/articles/PMC3118613/
#         osteo marker https://pubmed.ncbi.nlm.nih.gov/36442415/
#         osteo diff https://www.cell.com/fulltext/S0092-8674(06)00578-2

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Osteo-Chondro"
```

```{r others2_cluster10}
nr <- "10"
markers <- ct_markers_o2[[nr]]$genes[
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])
FeaturePlot(seu_sign_o2, markers[13:18])

# aspn, edil3, fmod, lum, mmp2, Prrx2, Mmp2, Sema3c

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Chondro1"
```

```{r others2_cluster9}
nr <- "9"
markers <- ct_markers_o2[[nr]]$genes[
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])
FeaturePlot(seu_sign_o2, markers[13:18])

# thbs4, clec3b, Crispld2, Mfap5, Tnxb, Itgbl1, Sema3c, Prrx2, Ssc5d

# thbs4 articular cartilage, articular stem and progenitor cells, 
# clec3b osteoblasts, mineralization
# Crispld2 osteogenic potential, cartilage, osteparthritis, cartilage
# tnxb articular cartilage
# Itgbl1, developing chondrocytes acc to one study (Song)
# Sema3c related to Itgbl1, cartilage
# Prrx2 osteogenic

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Chondro2"
```

```{r others2_cluster11}
nr <- "11"
markers <- ct_markers_o2[[nr]]$genes[
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

FeaturePlot(seu_sign_o2, markers[1:4])
FeaturePlot(seu_sign_o2, markers[5:8])
FeaturePlot(seu_sign_o2, markers[9:12])
FeaturePlot(seu_sign_o2, markers[13:16])

# Adipo/CAR, highly Adipo, batch pbs

# scara5, Pdgfrb, Snrk, Ebf3, Ltbp3, Mgp, Hp, Foxc1, Mgp

# scara5 adipogenesis
# Pdgfrb fibro
# ltbp3 early MSC diff, downregulated during osteo, required for adipo
# foxc1, highly adipo
# snrk, also adipi
# Notch3, adipo

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Adipo/CAR2, pbs"
```

```{r others2_cluster8}
nr <- "8"
markers <- ct_markers_o2[[nr]]$genes[
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])
FeaturePlot(seu_sign_o2, markers[13:18])

# Serpinf1, Sgms2, Col5a2, Cpe, Dcn, Satb2, Sp7, Serpinh1, Lum, Ptgis

# Serpins related to osteogenesis
# Sgms2 osteogenesis and osteo blasts
# col5a2 osteo progenitors, 
# Satb2 osteogenesis
# sp7 osteo
# Ptgis higher in osteo diff

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Osteo1"
```

```{r others2_cluster12}
nr <- "12"
markers <- ct_markers_o2[[nr]]$genes[
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])
FeaturePlot(seu_sign_o2, markers[13:18])

# Ackr3, Jam2, Mustn1, Cryab, Sgms2, Ptgis
# Ackr3 osteogenic differentiation
# mustn1 pan osteo muscle, muscle
# cryab increased during osteogebic diff in human, osteogenesis (human)
# Sgms2 osteogenesis and osteo blasts

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Osteo2"
```

```{r others2_cluster13}

nr <- "13"
markers <- ct_markers_o2[[nr]]$genes[
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

length(markers)/length(ct_markers_o2[[nr]]$genes) 

ct_markers_o2[[nr]][
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])
FeaturePlot(seu_sign_o2, markers[13:18])

# highest expression of almost all osteo markers

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Osteo3"
```

```{r compare}

# compare ct of origin for signature genes
# basically, check which cell type a signature was derived from which
# may help assignments

geneset_list_str[["Sinusoidal ECs"]]$conserved_signature[which(
  !geneset_list_str[["Sinusoidal ECs"]]$conserved_signature %in% c(
    geneset_list_str[["Arteriolar/Capillary ECs"]]$conserved_signature,
    geneset_list_str[["Sinusoidal/Venular ECs"]]$conserved_signature))]

sinusoidal <- c(
  "Sox17", 
  "Fcer1g",
  "Tgm2",
  "Npr1",
  "Kdr", 
  "Cdh5",
  "Clec14a",
  "Mef2c", 
  "Fcer1g",
  "Lmo2",
  "Clec2d",
  "Cldn5", 
  "Lrg1")

geneset_list_str[["Arteriolar/Capillary ECs"]]$conserved_signature[which(
  !geneset_list_str[["Arteriolar/Capillary ECs"]]$conserved_signature %in% c(
    geneset_list_str[["Sinusoidal ECs"]]$conserved_signature,
    geneset_list_str[["Sinusoidal/Venular ECs"]]$conserved_signature))]

arteriolar <- c(
  "Cavin2",
  "Tagln2",
  "Itga6", 
  "Dll4",
  "Fabp5",
  "Cdh5",
  "Mecom", 
  "Txnip", 
  "Anxa3", 
  "Sparcl1",
  "Cav2",
  "Robo4",
  "Rgcc",
  "Bcam",
  "Efnb2",
  "Mcf2l",
  "Nrp1", 
  "Tmcc3", 
  "Carin1", 
  "Lims2")

geneset_list_str[["Sinusoidal/Venular ECs"]]$conserved_signature[which(
  !geneset_list_str[["Sinusoidal/Venular ECs"]]$conserved_signature %in% c(
    geneset_list_str[["SSinusoidal ECs"]]$conserved_signature,
    geneset_list_str[["Arteriolar/Capillary ECs"]]$conserved_signature))]

sin_ven <- c(
  "Cd34",
  "Ndufa8",
  "Ehd4",
  "Klf4",
  "Ephb4",
  "Ier3",
  "Nfkbia",
  "Foxp1",
  "Plvap"
)
```


```{r others2_cluster6}

nr <- "6"
markers <- ct_markers_o2[[nr]]$genes[
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

markers_ec <- ct_markers_o2[[nr]]$genes[
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

ct_markers_o2[[nr]][
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

ct_markers_o2_ec[[nr]][
    ct_markers_o2_ec[[nr]]$avg_log2FC > 0 &
    ct_markers_o2_ec[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2_ec[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])
FeaturePlot(seu_sign_o2, markers[13:18])

FeaturePlot(seu_sign_o2, markers_ec[1:6])
FeaturePlot(seu_sign_o2, markers_ec[7:12])
FeaturePlot(seu_sign_o2, markers_ec[13:18])

# marker genes (cap)
FeaturePlot(seu_sign_o2, "Mcf2l")
FeaturePlot(seu_sign_o2, "Cd34")
FeaturePlot(seu_sign_o2, "Lims2")
FeaturePlot(seu_sign_o2, "Mecom")
FeaturePlot(seu_sign_o2, "Cav1")
FeaturePlot(seu_sign_o2, "Sparcl1")

# arteriolar genes
FeaturePlot(seu_sign_o2, "Sox18")
FeaturePlot(seu_sign_o2, "Efnb2")

# all cell types:
# Mcf2l, Cd34, Lims2, Cav1,  Sparcl1, Mecom, Fbln2, Podxl, Bcam, Ltbp4

# between ECs:  
# Mcf2l, Fbln2, Meox2, Sparcl1, Mecom, Cav1, Cd34, Thbs4

# Meox2: EC activation, endothelial cell proliferation
# Sparcl1 _ QUIESCENCE

# Mcf2l mouse pancreas EC, could be related to artery malformations in human, human atherosclerotic coronary arterial tissue 
#       pan-pancreatic https://www.ahajournals.org/doi/10.1161/ATVBAHA.124.321781
#       related to a lung capillary disease? https://onlinelibrary.wiley.com/doi/full/10.1002/ajmg.a.63141
#       capillary signature: https://journals.physiology.org/doi/full/10.1152/physrev.00006.2023
#       extracted from Art-EC

# Cd34  human capillaries and some venular, but not large vessels (https://www.sciencedirect.com/science/article/pii/S0006497120831425)
#       renal mouse capillaries (https://bmcnephrol.biomedcentral.com/articles/10.1186/s12882-017-0694-3)
#       related to EC proliferation (https://www.nature.com/articles/s41598-023-40622-7)
#       human brain capillaries (https://link.springer.com/article/10.1007/BF00227737)
#       from sin_ven

# Lims2 human shear stress (https://pmc.ncbi.nlm.nih.gov/articles/PMC7802379/)
#       capillary and post capillary venule (https://www.sciencedirect.com/science/article/pii/S0022202X22015421)
#       from arteriolar

# Cav1  no specific to one of our EC subtypes
#       related to EC differentiation (before angio), downregulated during proliferation, capillary process (https://pubmed.ncbi.nlm.nih.gov/11748236/) human
#       mouse retinal arterioles (https://molmed.biomedcentral.com/articles/10.1186/s10020-023-00749-9)
#       Cav1 molecules, and caveolae comprise 30% of the total surface area of capillary endothelial cells (https://www.sciencedirect.com/science/article/pii/S2405844023058619)
#       

# Sparcl1 from art-cap ECs
#         quiescence marker, stability, capillary morphogenesis (https://pubmed.ncbi.nlm.nih.gov/33393634/)
#         pulmonary capillary during inflammation (https://www.nature.com/articles/s41467-024-48589-3)

# Mecom from Art cap ECs
#       human EC differentiation master regulator (https://www.nature.com/articles/s41467-023-38002-w?fromPaywallRec=false)
#       related to human vena porta (https://pmc.ncbi.nlm.nih.gov/articles/PMC6379353/)
#       one of the most differentially expressed arterial TFs in mammals, arterial identity (https://pmc.ncbi.nlm.nih.gov/articles/PMC10844100/)
#       artery (Kalucka)

# Meox2 human heart capillary (https://pubmed.ncbi.nlm.nih.gov/25561514/)
#       human brain capillary density (https://pubmed.ncbi.nlm.nih.gov/16116430/)
#     

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "Art/Cap-EC"
knitr::knit_exit()
```

```{r others2_cluster1}

nr <- "1"

markers <- ct_markers_o2[[nr]]$genes[
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

markers_ec <- ct_markers_o2_ec[[nr]]$genes[
    ct_markers_o2_ec[[nr]]$avg_log2FC > 0 &
    ct_markers_o2_ec[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2_ec[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

ct_markers_o2[[nr]][
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

ct_markers_o2_ec[[nr]][
    #ct_markers_o2_ec[[nr]]$avg_log2FC > 0 &
    ct_markers_o2_ec[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2_ec[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])
FeaturePlot(seu_sign_o2, markers[13:18])

FeaturePlot(seu_sign_o2, markers_ec[1:6])
FeaturePlot(seu_sign_o2, markers_ec[7:12])
FeaturePlot(seu_sign_o2, markers_ec[13:18])


FeaturePlot(seu_sign_o2, "Axl")
FeaturePlot(seu_sign_o2, "Lpar6")
FeaturePlot(seu_sign_o2, "Aebp1")
FeaturePlot(seu_sign_o2, "Tgfbi")

# AEC transition
FeaturePlot(seu_sign_o2, "Sox18") # not only cluster 1, also cluster 6 --> ArtcCap
FeaturePlot(seu_sign_o2, "Clec14a") # not only cluster 1, also cluster 6 --> ArtCap

# VEC transition
FeaturePlot(seu_sign_o2, "Nr2f2") # not only cluster 1, also clusters 2 and 3
FeaturePlot(seu_sign_o2, "Ephb4") # not only cluster 1, also clusters 2 and 3

# non capillary
FeaturePlot(seu_sign_o2, "Rgcc") 
FeaturePlot(seu_sign_o2, "Cd34") 


# this cluster is mostly marked by NEGATIVE markers
# this could be a populationin transition

# all cell types:
# Sox18, Lpar6, Clec14a, Nr2f2, Prex2, Lrg1, Kdr1, Il6st

# ECs only:
# Axl, Lpar6, Clec14a, Aebp1, Tgfbi, Ephb4, Cd302

# Axl:  not specific from any EC type
#       carotid artery, related to inflammation, apoptosis, and flow-dependent modelling (https://pubmed.ncbi.nlm.nih.gov/16627783/)
#       vein graft SMC remoddeling (smooth muscle) (https://pubmed.ncbi.nlm.nih.gov/26276821/)
#       vascular remodelling rat after injury (https://pubmed.ncbi.nlm.nih.gov/17923589/)
#       carotid remoddelin (https://www.sciencedirect.com/science/article/pii/S000294401200171X)

# Lpar6 also not specific   
        #there is also a metabolite called Lpar6
#       EC barrier, human pulmonary arterial (HPAEC) and microvascular (HLMVEC) ECs (https://www.ncbi.nlm.nih.gov/books/NBK597453/)
#       related to angiogenesis from mouse KO , exrpessed in tumor vessels (https://www.mdpi.com/1422-0067/24/12/9812) 
#       related to BBB in rat? (https://www.sciencedirect.com/science/article/pii/S0006291X18311690)
#       human barrier EC function (https://pubmed.ncbi.nlm.nih.gov/23084965/)

# Aebp1 also not specific
#       tumor ECs, proliferation, migration, and in vitro tubem angiogenesis, microvessel formation (https://pubmed.ncbi.nlm.nih.gov/32086986/)
      
# Tgfbi not specific
#       inhibitor of angiogenesis in ECs mozse (https://www.nature.com/articles/s41598-021-88959-1)
#       related to tumor extravsation, human in several EC types (https://pmc.ncbi.nlm.nih.gov/articles/PMC2216691/)
#       tgf beta pathway member related to EC barrier (https://pmc.ncbi.nlm.nih.gov/articles/PMC7443870/)

# Sox18 not specific, also expressed in cluster 1
#       human EC barrier and maturation TF (https://pmc.ncbi.nlm.nih.gov/articles/PMC7054428/)
#       barrier via Cldn5 (https://grantome.com/grant/NIH/P01-HL134610-04-5237)
#       barrier via Cldn5 (https://journals.physiology.org/doi/full/10.1152/ajpheart.01248.2007)
#       barrier and NFkB (https://www.frontiersin.org/journals/physiology/articles/10.3389/fphys.2022.947537/full)
#       mouse vascular development (https://www.sciencedirect.com/science/article/pii/S1050173801001311)
#       related to arterial fate (https://www.nature.com/articles/s41375-024-02453-x)

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "EC1, non-pbs"
# CHECK BATCH
```

```{r others2_cluster2}

nr <- "2"
markers <- ct_markers_o2[[nr]]$genes[
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

markers_ec <- ct_markers_o2_ec[[nr]]$genes[
    ct_markers_o2_ec[[nr]]$avg_log2FC > 0 &
    ct_markers_o2_ec[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2_ec[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

ct_markers_o2[[nr]][
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

ct_markers_o2_ec[[nr]][
    #ct_markers_o2_ec[[nr]]$avg_log2FC > 0 &
    ct_markers_o2_ec[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2_ec[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])
FeaturePlot(seu_sign_o2, markers[13:18])

FeaturePlot(seu_sign_o2, markers_ec[1:6])
FeaturePlot(seu_sign_o2, markers_ec[7:12])
FeaturePlot(seu_sign_o2, markers_ec[13:18])

FeaturePlot(seu_sign_o2, "Tmcc3")
FeaturePlot(seu_sign_o2, "Spry2")
FeaturePlot(seu_sign_o2, "Sdc4")
FeaturePlot(seu_sign_o2, "Nfkbia")

FeaturePlot(seu_sign_o2, "Dll4") # arteriole, angiogenesis
FeaturePlot(seu_sign_o2, "Klf4") # angiogenesis
FeaturePlot(seu_sign_o2, "Rgcc") # capillary
 
# also mostly negative markers when comparing to other ECs

# markers all
# Tmcc3, Dll4, Klf4, Itga6

# markers ECs
# Spry2, Tmcc3, Klf4, Sdc4, Nfkbia, Foxc1

# Tmcc3   in arteriolar
#         cannot find anything useful relating to ECs

# Spry2   not specific
#         embryonic development (https://pubmed.ncbi.nlm.nih.gov/35297995/), related to Dll4, 
#         negatively impacts negatively regulate angiogenesis (https://bmcmedgenomics.biomedcentral.com/articles/10.1186/1755-8794-8-S1-S8)
#         RTK for growth and proliferation

# Klf4    from sinusoidal-venous   
#         related to arterial hypertension, anti-inflammatory and anticoagulant states (https://pubmed.ncbi.nlm.nih.gov/24156273/)
#         anti-inflammatory program that promotes vascular resilience (https://www.nature.com/articles/s44161-024-00522-z)
#         vascular integrity, Key EC TF https://pubmed.ncbi.nlm.nih.gov/28239661/
#         pro inflammation and proliferation, expression is induced by proinflammatory stimuli and shear stress (https://pmc.ncbi.nlm.nih.gov/articles/PMC5391259/)
#         homeostatic EC genes (homeostatic endothelial gene)
#         angiogenesis (https://pmc.ncbi.nlm.nih.gov/articles/PMC4467843/)

# Dll4    from arteriolar/Cap ECs
#         promotes arterial differentiation and restricts vessel branching, angiogenesis in the context of stress (https://pubmed.ncbi.nlm.nih.gov/23533173/)
#         Dll4âNotch signalling couples sprouting angiogenesis and artery formation (https://www.nature.com/articles/ncb3555)
#         highly expressed in capillary endothelial tip cells, angiogenesis (https://www.nature.com/articles/srep04031)
#         capillarization of liver sinusoids (https://www.sciencedirect.com/science/article/pii/S016748891930103X)

# Itga6   arteriolar/Cap ECs
#         human angiogenesis (https://aacrjournals.org/cancerres/article/70/14/5759/559331/Increased-Expression-of-6-Integrin-in-Endothelial)
#         tumor ECs, https://pmc.ncbi.nlm.nih.gov/articles/PMC10453423/
#         related to sprouting (https://pubmed.ncbi.nlm.nih.gov/32409567/)
#         endothelial sprouting (https://pmc.ncbi.nlm.nih.gov/articles/PMC8909174/#sec4-cells-11-00816)

# Sdc4    not specific
#         important cell-surface receptor in wound healing and angiogenesis https://pubmed.ncbi.nlm.nih.gov/11160142/
#         EC dysfunction in patients (https://pmc.ncbi.nlm.nih.gov/articles/PMC7564403/)
#         EC inflammation in patients

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "EC2, pbs"
```

```{r others2_cluster3}

nr <- "3"

markers <- ct_markers_o2[[nr]]$genes[
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

markers_ec <- ct_markers_o2_ec[[nr]]$genes[
    ct_markers_o2_ec[[nr]]$avg_log2FC > 0 &
    ct_markers_o2_ec[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2_ec[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL]

ct_markers_o2[[nr]][
    ct_markers_o2[[nr]]$avg_log2FC > 0 &
    ct_markers_o2[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]

ct_markers_o2_ec[[nr]][
    #ct_markers_o2_ec[[nr]]$avg_log2FC > 0 &
    ct_markers_o2_ec[[nr]]$p_val_adj < 0.05 &
    ct_markers_o2_ec[[nr]]$genes %in% ensembl_sign$MMUS_SYMBOL,]


FeaturePlot(seu_sign_o2, markers[1:6])
FeaturePlot(seu_sign_o2, markers[7:12])
FeaturePlot(seu_sign_o2, markers[13:18])


FeaturePlot(seu_sign_o2, markers_ec[1:6])
FeaturePlot(seu_sign_o2, markers_ec[7:12])
FeaturePlot(seu_sign_o2, markers_ec[13:18])

FeaturePlot(seu_sign_o2, "Vcam1") # activation
FeaturePlot(seu_sign_o2, "Kdr") # activation
FeaturePlot(seu_sign_o2, "Cd36") # activation all kind of unspecific

FeaturePlot(seu_sign_o2, "Ier3") 
FeaturePlot(seu_sign_o2, "Fabp4")
FeaturePlot(seu_sign_o2, "Fam167b") 

FeaturePlot(seu_sign_o2, "Il6st")
FeaturePlot(seu_sign_o2, "Fabp4")
FeaturePlot(seu_sign_o2, "Vcam1")
FeaturePlot(seu_sign_o2, "Lrg1")

# mostly negative marker genes, too

# markers all
# Fabp4, Ier3, Cd36, Srgn, Lrg1, Fam167b

# markers EC
# only mesenchymal markers, probably contamination (weak)

# Fabp4 in ECs related to VEGFA and DLL4, ECs important source of FABP4, pro-angiogenic
# Ier3: immediate early response
# Cd36: also related to fatty acid, FA uptake, angiogenesis, response to injury
# Lrg1: angiogenesis, mitogenic, neovascularization, 
# Srgn: shear-stress response
# cldn5 - sinusoidal in human, tight junction protein in ECs
# Fam167b, aorta-specific EC, pancreas ECs, kidney capillary cells

seu_sign_o2$annotated_reclustered[seu_sign_o2$seurat_clusters == nr] <- "EC3, non-pbs"
```

### Plots

#### Own

```{r mat_sign}
mat_sign <- base::table(
  sce_str_recl$celltypes, 
  sce_str_recl$cluster_signt)

# normalise per column = new clusters and turn into DF
for(i in 1:ncol(mat_sign)){
  mat_sign[,i] <- mat_sign[,i]/base::sum(mat_sign[,i])
}
sign_df <- base::as.data.frame(mat_sign)

#manually factorise clusters for nice order
sign_df$Var2 <- factor(
  sign_df$Var2,
  levels = base::rev(c(4, 8, 1, 3, 5, 2, 6, 7)))
```

```{r sign_base_plot}
sign_base_plot <- ggplot2::ggplot(
  sign_df, 
  aes(x = Var1, 
      y = Var2,
      fill = Freq))+
  ggplot2::geom_tile()
```

```{r sign_theme_plot}
sign_theme_plot <- sign_base_plot+
  theme_all+
  ggplot2::theme(
    legend.position = "none",
    axis.text.y = element_blank(),
    axis.title.x = element_blank(),
    axis.text.x = element_text(
      angle = 90,
      hjust = 0,
      vjust = 0.5),
    axis.ticks.y = element_blank())+
  ggplot2::scale_fill_gradientn(
    colors = mycolors_to1,
    limits = c(0, 1),
    breaks = c(0, 0.5, 1))+
  ggplot2::ylab("Reclustered")
```

### Other 2

```{r mat_sign_o2}
mat_sign_o2 <- base::table(seu_sign_o2$cell_type, 
                           seu_sign_o2$annotated_reclustered)

# normalise per column = new clusters and turn into DF
for(i in 1:ncol(mat_sign_o2)){
  mat_sign_o2[,i] <- mat_sign_o2[,i]/base::sum(mat_sign_o2[,i])
}
sign_df_o2 <- base::as.data.frame(mat_sign_o2)

sign_df_o2$Var2 <- unfactor(sign_df_o2$Var2)

sign_df_o2$Var2 <- factor(
  sign_df_o2$Var2,
  levels = base::rev(c(
    "Adipo/CAR1, pbs",
    "Adipo/CAR2, pbs",
    "Adipo/CAR3, non-pbs",
    "Osteo/CAR",
    "Osteo-Chondro",
    "Chondro1",
    "Chondro2",
    "Osteo1",
    "Osteo2",
    "Osteo3",
    "Art/Cap-EC",
    "EC1, non-pbs",
    "EC2, pbs",
    "EC3, non-pbs"
)))

sign_df_o2$Var1 <- unfactor(sign_df_o2$Var1)
sign_df_o2$Var1[sign_df_o2$Var1 == "P1"] <- "P1 adipogenesis1"
sign_df_o2$Var1[sign_df_o2$Var1 == "P2"] <- "P2 adipogenesis2"
sign_df_o2$Var1[sign_df_o2$Var1 == "P3"] <- "P3 osteo early"
sign_df_o2$Var1[sign_df_o2$Var1 == "P4"] <- "P4 osteo later"

sign_df_o2$Var1[sign_df_o2$Var1 == "O1"] <- "O1 osteo, Col16a"
sign_df_o2$Var1[sign_df_o2$Var1 == "O2"] <- "O2 osteo-chondro"
sign_df_o2$Var1[sign_df_o2$Var1 == "O3"] <- "O3 osteoblasts"

sign_df_o2$Var1[sign_df_o2$Var1 == "V1"] <- "V1 arterial"
sign_df_o2$Var1[sign_df_o2$Var1 == "V2"] <- "V2 sinusoidal" 

sign_df_o2$Var1 <- factor(
  sign_df_o2$Var1,
  levels = c(
    "P1 adipogenesis1",
    "P2 adipogenesis2",
    "P3 osteo early",
    "P4 osteo later",
    "O1 osteo, Col16a",
    "O2 osteo-chondro",
    "O3 osteoblasts",
    "V1 arterial",
    "V2 sinusoidal" 
))

```

```{r sign_base_plot_o2}
sign_base_plot_o2 <- ggplot2::ggplot(
  sign_df_o2, 
  aes(x = Var1, 
      y = Var2,
      fill = Freq))+
  ggplot2::geom_tile()
```

```{r sign_theme_plot_o2, fig.height = 5.5, fig.width = 5}
sign_theme_plot_o2 <- sign_base_plot_o2+
  theme_all+
  ggplot2::theme(
    legend.position = "none",
    axis.text.y = element_text(
      hjust = 1),
    axis.text.x = element_text(
      angle = 90,
      hjust = 0,
      vjust = 0.5),
    axis.ticks.y = element_blank())+
  ggplot2::scale_fill_gradientn(
    colors = mycolors_to1,
    limits = c(0, 1),
    breaks = c(0, 0.5, 1))+
  ggplot2::xlab("Original cell types")+
  ggplot2::ylab("Reclustered & Re-annotated")
```

### Other 3

```{r mat_sign_o3}
mat_sign_o3 <- base::table(seu_sign_o3$cell_type, 
                           seu_sign_o3$annotated_reclustered)

# normalise per column = new clusters and turn into DF
for(i in 1:ncol(mat_sign_o3)){
  mat_sign_o3[,i] <- mat_sign_o3[,i]/base::sum(mat_sign_o3[,i])
}
sign_df_o3 <- base::as.data.frame(mat_sign_o3)

sign_df_o3$Var2 <- factor(
  sign_df_o3$Var2,
  levels = base::rev(c(
    "Fibro-like CAR",
    "Osteo-Chondro-like CAR",
    "Osteo-like CAR",
    "Adipo-like CAR",
    "Osteo-Chondro",
    "EC1",
    "EC2",
    "EC3"
    )))

sign_df_o3$Var1 <- unfactor(sign_df_o3$Var1)
sign_df_o3$Var1[sign_df_o3$Var1 == "multipotent stromal stem cells"] <- "stromal stem cell"
sign_df_o3$Var1[sign_df_o3$Var1 == "balanced prog."] <- "balanced prog"
sign_df_o3$Var1[sign_df_o3$Var1 == "highly adipocytic gene-expressing progenitors"] <- "adipo-prog"
sign_df_o3$Var1[sign_df_o3$Var1 == "osteochondrogenic progenitors 1"] <- "osteo-chondro prog 1"
sign_df_o3$Var1[sign_df_o3$Var1 == "osteochondrogenic progenitors 2"] <- "osteo-chondro prog 2"

sign_df_o3$Var1 <- factor(
  sign_df_o3$Var1,
  levels = c(
    "stromal stem cell",
    "balanced prog",
    "adipo-prog",
    "pre-osteoblast",
    "osteo-chondro prog 1",
    "osteo-chondro prog 2",
    "pre-fibroblast 1",
    "pre-fibroblast 2",
    "pre-fibroblast 3",
    "endothelial"
    ))
```

```{r sign_base_plot_o3}
sign_base_plot_o3 <- ggplot2::ggplot(
  sign_df_o3, 
  aes(x = Var1, 
      y = Var2,
      fill = Freq))+
  ggplot2::geom_tile()
```

```{r sign_theme_plot_o3}
sign_theme_plot_o3 <- sign_base_plot_o3+
  theme_all+
  ggplot2::theme(
    legend.position = "none",
    axis.title = element_blank(),
    axis.text.y = element_text(
      hjust = 1),
    axis.text.x = element_text(
      angle = 90,
      hjust = 0,
      vjust = 0.5),
    axis.ticks.y = element_blank())+
   ggplot2::scale_fill_gradientn(
    "Proportion of cells per cluster",
    colors = mycolors_to1,
    limits = c(0, 1),
    breaks = c(0, 0.5, 1))
```

### Assemble

```{r assemble_cd, fig.width = 15, fig.height =5.5}

cd_plots <- ggpubr::ggarrange(
  ncol = 2,
  align = "h",
  sign_theme_plot_o2,
  sign_theme_plot_o3,
  widths = c(1.6, 1.4))

```

```{r b_plot_export}
pdf(output_pdf_b_plot, width = 5, height = 5.5)
sign_theme_plot
dev.off()
```

```{r cd_plot_export}
pdf(output_pdf_cd_plot, width = 15, height = 7)
cd_plots
dev.off()
```

```{r session_info}
utils::sessionInfo()
```
