---
title: "Figure 2"
date: '2024-09-04'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Generate Figure 2, current title:
"Cell type conserved signature genes are robust across mouse species and age."

#### Load objects

```{r seed, message = FALSE}
base::RNGkind(kind = "Mersenne-Twister")
set.seed(37)
```

```{r load_packages, message = FALSE}
library(tidyverse, quietly = TRUE)
library(SingleCellExperiment, quietly = TRUE)
library(ggpubr, quietly = TRUE)
library(scuttle, quietly = TRUE)
library(Seurat, quietly = TRUE)
library(SeuratObject, quietly = TRUE) 
library(ggpattern, quietly = TRUE) # this is not currently in conda envt
library(grDevices, quietly = TRUE) # this is not currently in conda envt
```

```{r base_path}
base_path_input <- "/omics/odcf/analysis/OE0538_projects/DO-0008/data/"
base_path_output_temp <- "/omics/odcf/analysis/OE0538_projects_temp/DO-0008/data_temp/"
```

```{r base_path_snakemake, eval = FALSE, include = FALSE}
base_path_input <- snakemake@input[["base_path_input"]]
base_path_output_temp <- snakemake@params[["base_path_output_temp"]]
```

```{r load_manual}
# fully annotated SCE object (HSPCs)
sce_hsc <- base::readRDS(base::paste0(
  base_path, 
  "/scRNAseq/main_analysis/sce_objects/02_sce_anno/10_anns/sce_hsc-10"))

# fully annotated SCE object (HSPCs) with diffusion pseudotime
sce_hsc_all_pseudotime <- base::readRDS(base::paste0(
 base_path,
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/05_cellrank/04_psce/sce_hsc_pseudotime"))

# fully annotated SCE object (HSPCs), erythroid branch cells only, with 
# pseudotime and erythroid differentiation probability
ery_pseudo <- base::readRDS(base::paste0(
  base_path_input, 
  "/scRNAseq/main_analysis/sce_objects/03_sce_analysis/05_cellrank/05_bsce/sce_ery"))

# list of gene sets for HSPCs (conserved signature, conserved markers, etc.)
geneset_list_hsc <- base::readRDS(base::paste0(
  base_path_input, 
  "scRNAseq/main_analysis/sce_objects/03_sce_analysis/04_signatures/01_reclustering_own/01_gens/geneset_list_hsc"))

# conserved markers per cell type per age (markers_conservation_hsc)
base::load(base::paste0(
base_path_input,
"scRNAseq/main_analysis/sce_objects/03_sce_analysis/03_marker_conservation/02_age/cons_markers_hsc.RData"))
  
colors_path <- base::paste0(
  base_path_input, 
  "/metadata/colors/colors.txt")

source("../source/colors.R")
```

```{r params}
# global params
source("determine_params.R")

# specific params

pattern_density_val <- 0.08
pattern_spacing_val <- 0.01

celltype <- "HSCs"
```

```{r export_paths}
output_pdf_bcd_plot <- base::paste0(
  base_path_output_temp, "/manuscript1/figures/figure2_bcd.pdf")
output_pdf_b_legend <- base::paste0(
  base_path_output_temp, "/manuscript1/figures/figure2_b_legend.pdf")
output_pdf_c_legend <- base::paste0(
  base_path_output_temp, "/manuscript1/figures/figure2_c_legend.pdf")
output_pdf_ef_plot <- base::paste0(
  base_path_output_temp, "/manuscript1/figures/figure2_ef.pdf")
```

## A

Model of methods used to extract conserved signature genes for each cell type. 
Conserved signature genes are non-differentially expressed conserved markers
within the same cell type across all species.

@Lea

## B

Expression of conserved signature genes in all mouse species across HSPC
cell types (scaled by z-score per gene).

@Lea

#### Plot Components

```{r b_prep1}
# get all unique signature genes
all_sign_genes <- vector()
for(i in 1:length(geneset_list_hsc)){
  sign_genes <- geneset_list_hsc[[i]]$conserved_signature
  all_sign_genes <- c(all_sign_genes, sign_genes)
}
all_sign_genes_hsc <- base::unique(all_sign_genes)
```

```{r b_prep2}
# aggregate per cell type and species
cts_hsc <- as.list(names(geneset_list_hsc))

# separate SCE objects by cell type, then aggregate the counts to obtain
# the average counts per species for each cell type separately
# use the multi-batchnorm-corrected logcounts 

agg_hsc_list <- lapply(cts_hsc, function(ct){
  
  sce_ct <- sce_hsc[,sce_hsc$celltypes == ct]
  agg_hsc <- scuttle::aggregateAcrossCells(
    sce_ct, 
    id=colData(sce_ct)[,c("Species_ID")], 
    statistics = "mean",
    use.assay.type = "logcounts") #multibatchnorm
  
  return(agg_hsc)
})
names(agg_hsc_list) <- names(geneset_list_hsc)
```

```{r b_prep3}
expr_df_list_hsc <- lapply(agg_hsc_list, function(agg_hsc){
  
  # get current cell type and relevant signature genes
  ct_curr <- agg_hsc$celltypes[1]
  sign_genes_ct <- geneset_list_hsc[[ct_curr]]$conserved_signature
  
  # get a df from average logcounts per ct and species
  vis_df <- base::as.data.frame(
    t(SummarizedExperiment::assays(agg_hsc)$logcounts))
  
  # subset to all signature genes of this tissue
  vis_df <- vis_df[,colnames(vis_df) %in% all_sign_genes_hsc]
  
  # change df format 
  vis_df <- tibble::rownames_to_column(vis_df, "species")
  vis_df <- tidyr::pivot_longer(vis_df, 
                                cols = c(2:ncol(vis_df)),
                                values_to = "expression",
                                names_to = "gene")
  
  # add ct info
  vis_df$cell_type <- base::rep(ct_curr)

  # add info on genes that are signature genes of this ct
  vis_df$is_sign_gene <- vector(length = nrow(vis_df))
  vis_df$is_sign_gene[vis_df$gene %in% sign_genes_ct] <- TRUE

  # scale expression of all genes per cell type 
  # for ordering of genes ONLY
  vis_df$scaled_expression_per_ct <- base::scale(vis_df$expression)
  vis_df <- vis_df[base::order(vis_df$scaled_expression_per_ct, 
                               decreasing = TRUE),]

  return(vis_df)
})

# bind dfs 
agg_hsc_df <- dplyr::bind_rows(expr_df_list_hsc)
```

```{r b_prep4}
# scale expression per gene using z-scores
agg_hsc_df$zscore_gene <- vector(length = nrow(agg_hsc_df), mode = "numeric")
for(g in base::unique(agg_hsc_df$gene)){
  
  expr_vector_g <- agg_hsc_df[agg_hsc_df$gene == g,]$expression

  # calculate z score per gene
  z_score_g <- (expr_vector_g - base::mean(expr_vector_g))/
    stats::sd(expr_vector_g)
  agg_hsc_df$zscore_gene[agg_hsc_df$gene == g] <- z_score_g
}
```

```{r b_prep5}
# get a well ordered list of unique genes for factorization (after order above)
all_sign_genes_for_factor <- base::unique(agg_hsc_df$gene[
  agg_hsc_df$is_sign_gene == TRUE])

# factorize
agg_hsc_df$gene <- factor(agg_hsc_df$gene,
                          levels = all_sign_genes_for_factor)
agg_hsc_df$species <- factor(agg_hsc_df$species,
                             levels = base::rev(names(col_spc)))
```

```{r b_prep6}
interest_genes <- c("Msi2",
                    "Angpt1",
                    "Cd34",
                    "Mpo",
                    "Tcf4", 
                    "Itga2b",
                    "Klf1",
                    "Ccnb2")

agg_hsc_df$is_gene <- vector(length = nrow(agg_hsc_df))
agg_hsc_df$is_gene[agg_hsc_df$gene %in% interest_genes] <- "TRUE"
agg_hsc_df$is_gene <- as.logical(agg_hsc_df$is_gene)
```

```{r b_base_plot}
b_base_plot <- ggplot2::ggplot(
  agg_hsc_df,
  aes(y = species, x = gene, fill = zscore_gene))+
  ggplot2::geom_tile()+
  ggplot2::facet_grid(rows = vars(agg_hsc_df$cell_type))
```

```{r b_theme_plot}
b_theme_plot <- b_base_plot+
  theme_all+
  ggplot2::theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    plot.title = element_text(
      vjust = 0.5,
      hjust = 0.5,
      color = col_cons_long["conserved identity signature"]),
    axis.title.y = element_blank(),
    axis.line.x = element_blank(),
    axis.text.y = element_blank(),
    strip.text.y = element_text(
      size = axis_text_size,
      face = axis_text_face,
      color = axis_text_color,
      angle = 0,
      vjust = 0.5,
      hjust = 0),
    strip.background.y = element_blank())+
  ggplot2::ggtitle("Conserved identity signature genes")+
  ggplot2::scale_fill_gradientn(
    "z-score per gene", 
    colours = colors_z_score,
    breaks = c(-5, -2.5,  0, 2.5, 5), 
    limits = c(-5,5))+
  ggplot2::scale_x_discrete(
    position =  "bottom",
    breaks = agg_hsc_df$gene[agg_hsc_df$is_gene])

b_legend_theme <- ggpubr::get_legend(b_theme_plot)
```

### Assemble

```{r b_construct, fig.width = 6, fig.height = 6}
b_plot <- ggpubr::ggarrange(
  b_theme_plot + ggplot2::theme(legend.position = "none")
)
b_plot
```

```{r b_construct_legend, fig.width = 4, fig.height = 4}
b_legend <- ggpubr::ggarrange(b_legend_theme )
```

```{r b_construct_legend_long, fig.width = 4, fig.height = 4}
b_legend_long <- ggpubr::get_legend(
  b_theme_plot+ 
    theme(legend.position = "bottom",
    legend.title = element_text(
      margin = margin(r = 10),
      vjust = 0.85))+
  ggplot2::scale_fill_gradientn(
    "z-score per gene", 
    colours = colors_z_score,
    breaks = c(-5, 0,  5), 
    limits = c(-5,5)))

b_legend_theme_long <- ggpubr::ggarrange(
  b_legend_long 
)
```

## C Species-specific marke rgenes

Scaled expression of species-specific marker genes in all mouse species within
HSCs (scaled by z-score per gene).

@Lea

#### Prep

```{r c_prep}
# get ct specific data
hsc_cons_df_ct <- geneset_list_hsc$HSCs$conserved_df

hsc_mmus_specific <- hsc_cons_df_ct$gene[which(
  !is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]

hsc_mcas_specific <- hsc_cons_df_ct$gene[which(
  is.na(hsc_cons_df_ct$mmus) &
    !is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]

hsc_mspr_specific <- hsc_cons_df_ct$gene[which(
  is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    !is.na(hsc_cons_df_ct$mspr) &
    is.na(hsc_cons_df_ct$mcar))]

hsc_mcar_specific <- hsc_cons_df_ct$gene[which(
  is.na(hsc_cons_df_ct$mmus) &
    is.na(hsc_cons_df_ct$mcas) &
    is.na(hsc_cons_df_ct$mspr) &
    !is.na(hsc_cons_df_ct$mcar))]
  
hsc_genes <- c(hsc_mmus_specific,
               hsc_mcas_specific,
               hsc_mspr_specific,
               hsc_mcar_specific)

stopifnot(!duplicated(hsc_genes))
```

```{r c_prep2}
agg_hsc_ct <- agg_hsc_list[[celltype]]

# make visualisation DF
vis_df_hsc <- as.data.frame(
  t(SummarizedExperiment::assays(agg_hsc_ct)$logcounts)) # batchnorm, aggregated
vis_df_hsc <- vis_df_hsc[,colnames(vis_df_hsc) %in% hsc_genes]
vis_df_hsc <- tibble::rownames_to_column(vis_df_hsc, "species")
vis_df_hsc <- tidyr::pivot_longer(
  vis_df_hsc, 
  cols = c(2:ncol(vis_df_hsc)),
  values_to = "expression",
  names_to = "gene")

vis_df_hsc$species <- factor(vis_df_hsc$species, levels = base::rev(names(col_spc)))
vis_df_hsc$gene <- factor(vis_df_hsc$gene, levels = hsc_genes)

vis_df_hsc$zscore_gene <- vector(length = nrow(vis_df_hsc), mode = "numeric")

for(g in base::unique(vis_df_hsc$gene)){
  
  expr_vector_g <- vis_df_hsc[vis_df_hsc$gene == g,]$expression

  # calculate z score per gene
  z_score_g <- (expr_vector_g - base::mean(expr_vector_g))/stats::sd(expr_vector_g)
  vis_df_hsc$zscore_gene[vis_df_hsc$gene == g] <- z_score_g
  
}
levels(vis_df_hsc$gene)
```

```{r c_prep3}
interest_genes2 <- c(
  "Procr",
  "Ly6a",
  "Cited2",
  "Cd74",
  "Krt18",
  "Socs2",
  "Cd34",
  "Cd63",
  "Pbx1",
  "Plek"
)

vis_df_hsc$is_gene <- vector(length = nrow(vis_df_hsc))
vis_df_hsc$is_gene[vis_df_hsc$gene %in% interest_genes2] <- "TRUE"
vis_df_hsc$is_gene <- as.logical(vis_df_hsc$is_gene)
print(levels(vis_df_hsc$gene))
```

#### Plot parts

```{r c_base_plot}
c_base_plot <- ggplot2::ggplot(
  vis_df_hsc, 
  aes(y = species, 
      x = gene, 
      fill = zscore_gene))+
  ggplot2::geom_tile()
```

```{r c_theme_plot}
c_theme_plot <- c_base_plot +
  theme_all+
  ggplot2::theme(
    axis.title = element_blank(),
    plot.title = element_text(
      hjust = 0.5),
    axis.text.x = element_blank(),
    axis.line.x = element_blank(),
    axis.text.y = element_text(
      angle = 0,
      vjust = 0.5,
      hjust = 0))+
  ggplot2::ggtitle("Species-specific HSC marker genes")+
  ggplot2::scale_fill_gradientn(
    "z-score per gene", 
    colours = colors_z_score,
    breaks = c(-5, -2.5, 0, 2.5, 5), 
    limits = c(-5,5))+
  ggplot2::scale_y_discrete(expand = c(0, 0))+
  ggplot2::scale_x_discrete(
    position =  "bottom",
    breaks = vis_df_hsc$gene[vis_df_hsc$is_gene])

c_theme_legend <- ggpubr::get_legend(c_theme_plot)
```

### Assemble

```{r c_construct, fig.width = 6, fig.height = 2}
c_plot <- ggpubr::ggarrange(
  c_theme_plot+theme(legend.position = "none")
)
c_plot
```

## BCD PLOT ASSEMBLY

```{r b_c_d_plotX, fig.width = 11, fig.height = 12.5, eval = FALSE, include = FALSE}

#TODO: Decide how to assemble after decision on how drastically the figure should be revised,
b_c_d_plot <- ggpubr::ggarrange(
  b_theme_plot + ggplot2::theme(
    plot.margin = margin("r" = 0, unit = "cm"),
    legend.position = "none",
    plot.title = element_text(margin = margin("b" = 1.38, unit = "cm"))),
  c_theme_plot +  ggplot2::theme(
    axis.text.y = element_blank(),
    plot.margin = margin("l" = 0, unit = "cm")),
  d_theme_plot +  ggplot2::theme(
    legend.position = "none",
    plot.margin = margin("r" = 4.2, t = "2", unit = "cm"),
    axis.text.y = element_text(margin = margin(
      "l" = 0, "r" = 0.11, unit = "cm"))),
  widths = c(1, 0.6),
  heights = c(10, 2.5)
)

b_c_d_plot
```

```{r bcd_plot_exportX, eval = FALSE, include = FALSE}
pdf(output_pdf_bcd_plot, width = 11, height = 12.5)
b_c_d_plot
dev.off()
```

```{r b_legend_exportX, eval = FALSE, include = FALSE}
pdf(output_pdf_b_legend, width = 4, height = 1)
b_legend_theme_long
dev.off()
```

```{r d_legend_exportX, eval = FALSE, include = FALSE}
pdf(output_pdf_c_legend, width = 3, height = 1.4)
c_theme_legend
dev.off()
```

```{r remove_big_objects}
rm(agg_hsc_list)
rm(agg_hsc_ct)
rm(agg_hsc_df)

rm(b_base_plot)
rm(b_c_d_plot)
rm(b_theme_plot)

gc()
```

## D Umap Pseudotime (and possibly erythroid branch)

TODO: Think about naming of erythroid branch and other branches

```{r d_prep1}

sce_hsc_all_pseudotime

sce_hsc$pseudotime <- sce_hsc_all_pseudotime$dpt_pseudotime
sce_hsc$ery <- ery_pseudo$

hsc_umap_pt <- base::as.data.frame(
  SingleCellExperiment::reducedDim(
    sce_hsc, type = "UMAP"))
hsc_umap_pt$color_by <- sce_hsc$pseudotime 
hsc_umap_pt <- hsc_umap_pt[base::sample(
  1:nrow(hsc_umap_pt),
  nrow(hsc_umap_pt),
  replace = FALSE),]
```

```{r d_base_plot_1}
d_base_plot_1 <- ggplot2::ggplot(
  hsc_umap_pt, 
  aes(x = X1,
      y = X2,
      color = color_by))+
  ggplot2::geom_point(
    size = umap_point_size,
    alpha = umap_point_alpha)+
  ggplot2::xlim(c(base::min(hsc_umap_pt$X1)-1, base::max(hsc_umap_pt$X1))+1)+
  ggplot2::ylim(c(base::min(hsc_umap_pt$X2)-1, base::max(hsc_umap_pt$X2))+1)+
  ggplot2::scale_color_gradientn(colours = mycolors_to1_grey)
```

```{r d_theme_plot_1}
d_theme_plot_1 <- d_base_plot_1+
  cowplot::theme_nothing()
```

## E

Scaled expression of HSC (top), Mk/Erythroid prog. (middle) or Erythroid prog.
(bottom) conserved signature genes along diffusion pseudotime trajectory within 
the erythroid branch. Cells were partitioned into branches based on their 
lineage differentiation probability calculated by cellrank.

@Veronica

```{r e_prep_1, message = FALSE}
ery_pseudo_seurat <- CreateSeuratObject(counts = ery_pseudo@assays@data@listData[["logcounts"]],
                                        meta.data = ery_pseudo@colData@listData)

rm(ery_pseudo)
gc()

plot_data <- ery_pseudo_seurat@meta.data %>%
  select("pseudotime", "celltypes", "Mouse_ID", 
         "Species_ID", "Age_weeks", "Age_ID")
for(n in 1:length(geneset_list_hsc)){
  cell_sp_markers <- geneset_list_hsc[[n]][["conserved_signature"]]
  for(g in cell_sp_markers){
    plot_data[,g] <-unlist(SeuratObject::FetchData(ery_pseudo_seurat, g))
  }
}
plot_data[,7:ncol(plot_data)] <- scale(plot_data[,7:ncol(plot_data)])
```

```{r e_prep_2}
plot_HSCs <- plot_data[,c(colnames(plot_data)[1:6],
                           geneset_list_hsc[["HSCs"]][["conserved_signature"]])] %>%
  gather(gene, expression,
         -pseudotime, -celltypes, -Mouse_ID, 
         -Species_ID, -Age_weeks, -Age_ID)
col_HSCs <- grDevices::colorRampPalette(c("grey95", 
                                          col_cts_hsc["HSCs"], 
                                          "grey5"))
palette_HSCs <- col_HSCs(length(geneset_list_hsc[["HSCs"]][["conserved_signature"]]))


plot_MkEry <- plot_data[,c(colnames(plot_data)[1:6],
                           geneset_list_hsc[["Mk/Ery progs."]][["conserved_signature"]])] %>%
  gather(gene, expression,
         -pseudotime, -celltypes, -Mouse_ID, 
         -Species_ID, -Age_weeks, -Age_ID)
col_MkEry <- grDevices::colorRampPalette(c("grey95", 
                                           col_cts_hsc["Mk/Ery progs."], 
                                           "grey5"))
palette_MkEry <- col_MkEry(length(geneset_list_hsc[["Mk/Ery progs."]][["conserved_signature"]]))


plot_Ery <- plot_data[,c(colnames(plot_data)[1:6],
                           geneset_list_hsc[["Erythroid progs."]][["conserved_signature"]])] %>%
  gather(gene, expression,
         -pseudotime, -celltypes, -Mouse_ID, 
         -Species_ID, -Age_weeks, -Age_ID)
col_Ery <- grDevices::colorRampPalette(c("grey95", 
                                         col_cts_hsc["Erythroid progs."], 
                                         "grey5"))
palette_Ery <- col_Ery(length(geneset_list_hsc[["Erythroid progs."]][["conserved_signature"]]))
```

```{r e_plot}
HSC_1 <- ggplot(plot_HSCs,
       aes(x = rank(pseudotime),
           y = expression,
           color = gene)) +
  geom_smooth(se = F) +
  guides(color = "none") +
  labs(x = "Pseudotime Erythroid", 
       y = "Scaled Expression") +
  scale_color_manual(values = rev(palette_HSCs)) +
  theme_classic() +
  theme(legend.text = element_blank(),
        axis.text = element_blank())

MkEry_2 <- ggplot(plot_MkEry,
       aes(x = rank(pseudotime),
           y = expression,
           color = gene)) +
  geom_smooth(se = F) +
  guides(color = "none") +
  labs(x = "Pseudotime Erythroid", 
       y = NULL) +
  scale_color_manual(values = rev(palette_MkEry)) +
  theme_classic() +
  theme(legend.text = element_blank(),
        axis.text = element_blank())

Ery_3 <- ggplot(plot_Ery,
       aes(x = rank(pseudotime),
           y = expression,
           color = gene)) +
  geom_smooth(se = F) +
  guides(color = "none") +
  labs(x = "Pseudotime Erythroid", 
       y = NULL) +
  scale_color_manual(values = rev(palette_Ery)) +
  theme_classic() +
  theme(legend.text = element_blank(),
        axis.text = element_blank())

```

```{r e_theme_plot}
HSC_1_theme <- HSC_1+ 
  ggplot2::theme(   
    axis.title.x = element_blank(),
    axis.title.y = element_text(
      face = axis_title_face,
      size = axis_title_size,
      color = axis_title_color),
    axis.text.x = element_blank(),
    axis.text.y = element_blank(),
    plot.title = element_text(
      face = plot_title_face,
      size = plot_title_size,
      color = plot_title_color,
      hjust = 0.5))+
  ggplot2::ggtitle("HSCs")

MkEry_2_theme <- MkEry_2+ 
  ggplot2::theme(   
    axis.title.y = element_blank(),
    axis.title.x = element_text(
      face = axis_title_face,
      size = axis_title_size,
      color = axis_title_color),
    axis.text.x = element_blank(),
    axis.text.y = element_blank(),
    plot.title = element_text(
      face = plot_title_face,
      size = plot_title_size,
      color = plot_title_color,
      hjust = 0.5))+
  ggplot2::ggtitle("Mk/Ery progs.")

Ery_3_theme <- Ery_3+ 
  ggplot2::theme(   
    axis.title.y = element_blank(),
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    axis.text.y = element_blank(),
    plot.title = element_text(
      face = plot_title_face,
      size = plot_title_size,
      color = plot_title_color,
      hjust = 0.5))+
  ggplot2::ggtitle("Ery progs.")
```


```{r e_assemble, fig.width = 6, fig.height = 2}
print(
  ggpubr::ggarrange(
    HSC_1,
    MkEry_2,
    Ery_3,
    nrow = 1,
    widths = c(0.36, 0.32, 0.32)
  )
)
```

```{r e_assemble_plot, fig.width = 6, fig.height = 2}
e_plot <- ggpubr::ggarrange(
    HSC_1+theme(axis.title.x = element_blank()),
    MkEry_2,
    Ery_3+theme(axis.title.x = element_blank()),
    nrow = 1,
    align = "h",
    widths = c(0.36, 0.32, 0.32)
  )
```

```{r e_assemble_plot2, fig.width = 10, fig.height = 3}
e_plot <- ggpubr::ggarrange(
    HSC_1_theme,
    MkEry_2_theme,
    Ery_3_theme,
    nrow = 1,
    align = "h",
    widths = c(0.36, 0.32, 0.32)
  )
e_plot
```

## F

G: Scaled expression of signature genes [geneX], [geneY], and [geneZ] across
mouse species and ages (z-score per species and age).

@Veronica

```{r f_prep_1}
metadata <- sce_hsc@colData@listData
sce_hsc_seurat <- SeuratObject::CreateSeuratObject(counts = sce_hsc@assays@data@listData[["logcounts"]],
                           meta.data = metadata)

# assign gene list!!
g <- c(
  "Nkx2-3",
  "Trib2", 
  "Blvrb"
)

metadata$expression <- SeuratObject::FetchData(sce_hsc_seurat, 
                        vars = g)
metadata$celltypes <- as.character(metadata$celltypes)
gene_expression <- metadata$expression %>%
  cbind(do.call(cbind, metadata[c("Species_ID", "Age_ID", "celltypes")]))
```

```{r f_prep_2}
plot_1 <- gene_expression[,c(g[1], "Species_ID", "Age_ID", "celltypes")] 
colnames(plot_1)[1] <- "gene"
plot_1 <- plot_1 %>%
  group_by(Species_ID, Age_ID, celltypes) %>%
  summarize(mean = mean(gene))
plot_1$signature <- lapply(plot_1$celltypes, function(x){
  g[1] %in% geneset_list_hsc[[x]][["conserved_signature"]]
}) %>% unlist()


plot_2 <- gene_expression[,c(g[2], "Species_ID", "Age_ID", "celltypes")] 
colnames(plot_2)[1] <- "gene"
plot_2 <- plot_2 %>%
  group_by(Species_ID, Age_ID, celltypes) %>%
  summarize(mean = mean(gene))
plot_2$signature <- lapply(plot_2$celltypes, function(x){
  g[2] %in% geneset_list_hsc[[x]][["conserved_signature"]]
}) %>% unlist()


plot_3 <- gene_expression[,c(g[3], "Species_ID", "Age_ID", "celltypes")] 
colnames(plot_3)[1] <- "gene"
plot_3 <- plot_3 %>%
  group_by(Species_ID, Age_ID, celltypes) %>%
  summarize(mean = mean(gene))
plot_3$signature <- lapply(plot_3$celltypes, function(x){
  g[3] %in% geneset_list_hsc[[x]][["conserved_signature"]]
}) %>% unlist()
```


```{r f_plot}
marker_1 <- ggplot(plot_1,
       aes(x = factor(celltypes, levels = names(geneset_list_hsc)),
           y = paste(Species_ID, Age_ID),
           fill = mean,
           color = signature,
           size = signature)) +
  geom_tile(width = 0.95, 
            height = 0.95) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, h = 1)) +
  scale_color_manual(guide = "none", values = c(alpha("black", 0), "grey50")) +
  scale_size_manual(guide = "none", values = c(0, 0.4)) +
  labs(x = NULL, y = NULL, title = g[1]) +
  scale_fill_gradient2(low = "blue3", mid = "white", high = "red3",
                       midpoint = mean(plot_1$mean)) +
  theme(legend.position = "none")


marker_2 <- ggplot(plot_2,
       aes(x = factor(celltypes, levels = names(geneset_list_hsc)),
           y = paste(Species_ID, Age_ID),
           fill = mean,
           color = signature,
           size = signature)) +
  geom_tile(width = 0.95, 
            height = 0.95) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, h = 1),
        axis.text.y = element_blank()) +
  scale_color_manual(guide = "none", values = c(alpha("black", 0), "grey50")) +
  scale_size_manual(guide = "none", values = c(0, 0.4)) +
  labs(x = NULL, y = NULL, title = g[2]) +
  scale_fill_gradient2(low = "blue3", mid = "white", high = "red3",
                       midpoint = mean(plot_2$mean)) +
  theme(legend.position = "none")


marker_3 <- ggplot(plot_3,
       aes(x = factor(celltypes, levels = names(geneset_list_hsc)),
           y = paste(Species_ID, Age_ID),
           fill = mean,
           color = signature,
           size = signature)) +
  geom_tile(width = 0.95, 
            height = 0.95) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, h = 1),
        axis.text.y = element_blank()) +
  scale_color_manual(guide = "none", values = c(alpha("black", 0), "grey50")) +
  scale_size_manual(guide = "none", values = c(0, 0.4)) +
  labs(x = NULL, y = NULL, title = g[3]) +
  scale_fill_gradient2(low = "blue3", mid = "white", high = "red3",
                       midpoint = mean(plot_3$mean)) +
  theme(legend.position = "none")


# still need to get colors, legend, etc all correct
```

```{r f_theme}

marker_1_theme <- marker_1+
  ggplot2::theme(
    axis.title.y = element_blank(),
    axis.title.x = element_blank(),
    axis.text.x = element_text(
      face = axis_text_face,
      size = axis_text_size,
      color = axis_text_color,
      hjust = 0,
      vjust = 0.5),
    axis.text.y = element_text(
      face = axis_text_face,
      size = axis_text_size,
      color = axis_text_color,
      vjust = 0.5,
      hjust = 0),
    plot.title = element_text(
      face = "italic",
      size = plot_title_size,
      hjust = 0.5,
      color = plot_title_color))+
  ggplot2::scale_y_discrete(expand = c(0, 0))

marker_2_theme <- marker_2+
  ggplot2::theme(
    axis.title.y = element_blank(),
    axis.title.x = element_blank(),
    axis.text.x = element_text(
      face = axis_text_face,
      size = axis_text_size,
      color = axis_text_color,
      hjust = 0,
      vjust = 0.5),
    axis.text.y = element_blank(),
    plot.title = element_text(
      face = "italic",
      size = plot_title_size,
      hjust = 0.5,
      color = plot_title_color))+
  ggplot2::scale_y_discrete(expand = c(0, 0))

marker_3_theme <- marker_3 +
  ggplot2::theme(
    axis.title.y = element_blank(),
    axis.title.x = element_blank(),
    axis.text.x = element_text(
      face = axis_text_face,
      size = axis_text_size,
      color = axis_text_color,
      hjust = 0,
      vjust = 0.5),
    axis.text.y = element_blank(),
    plot.title = element_text(
      face = "italic",
      size = plot_title_size,
      hjust = 0.5,
      color = plot_title_color))+
  ggplot2::scale_y_discrete(expand = c(0, 0))

```


```{r f_assemble, fig.width = 9, fig.height = 3.5}
print(
  ggpubr::ggarrange(
    marker_1,
    marker_2,
    marker_3,
    nrow = 1,
    widths = c(0.4, 0.3, 0.3)
  )
)
```

```{r f_assemble_plot, fig.width = 9, fig.height = 3.5}
f_theme_plot <- ggpubr::ggarrange(
    marker_1_theme,
    marker_2_theme,
    marker_3_theme,
    nrow = 1,
    widths = c(0.4, 0.3, 0.3)
  )
```

```{r assemble_e_f, fig.width = 10, fig.height = 7}

e_f_plot <- ggpubr::ggarrange(
  e_plot + theme(plot.margin = margin("l" = 1.75, unit = "cm")),
  f_theme_plot + theme(plot.margin = margin("t" = 1, unit = "cm")),
  nrow = 2,
  heights = c(1, 2)
  )
e_f_plot
```

```{r bcd_plot_export}
pdf(output_pdf_ef_plot, width = 11, height = 7)
e_f_plot
dev.off()
```

## F Robustness to Aging

Number of marker genes per HSPC cell type identified when all, only old, 
or only young mouse samples are included.

@Veronica

```{r f_prep_1}
conserved_genes <- data.frame(celltype = rep(names(geneset_list_hsc)),
                              young = 0,
                              both = 0,
                              old = 0,
                              n_markers = 0,
                              young_in_markers = 0,
                              both_in_markers = 0,
                              old_in_markers = 0)
for(ct in names(geneset_list_hsc)){
  young <- markers_conservation_hsc[[grep(paste(ct, "young", sep = "_"),
                                      names(markers_conservation_hsc),
                                      value = T)]] %>% as.data.frame()
  young$gene <- rownames(young)
  young <- young[complete.cases(young),]

  old <- markers_conservation_hsc[[grep(paste(ct, "old", sep = "_"),
                                    names(markers_conservation_hsc),
                                    value = T)]] %>% as.data.frame()
  old$gene <- rownames(old)
  old <- old[complete.cases(old),]

  both <- young$gene[young$gene %in% old$gene]

  mark <- geneset_list_hsc[[grep(ct,
                                 names(geneset_list_hsc),
                                 value = T)]][["conserved_df"]]
  mark <- mark[complete.cases(mark),]

  conserved_genes[which(conserved_genes$celltype == ct),1] <- ct

  conserved_genes[which(conserved_genes$celltype == ct),2:8] <-
    c(nrow(young),
      length(both),
      nrow(old),
      nrow(mark),
      sum(young$gene %in% mark$gene),
      sum(both %in% mark$gene),
      sum(old$gene %in% mark$gene)) %>% as.numeric()
}

plot_hsc <- gather(conserved_genes[,c("celltype", 
                                      "n_markers",
                                      "young_in_markers", 
                                      "old_in_markers",
                                      "both_in_markers")], 
                   "group", "count", -celltype, -both_in_markers)
```

```{r f_plot, fig.width = 3, fig.height = 5}
print(
ggplot(plot_hsc,
       aes(y = factor(celltype, levels = rev(names(geneset_list_hsc))),
           fill = factor(group, levels = c("young_in_markers", 
                                           "n_markers", 
                                           "old_in_markers")))) +
  geom_col(aes(x = count),
           position = "dodge") +
  ggpattern::geom_col_pattern(aes(x = both_in_markers),
           position = "dodge",
           pattern = "stripe",
           pattern_color = "black",
           pattern_spacing = 0.07,
           pattern_density = 0.001,
           pattern_angle = 60) +
  theme_classic() +
  labs(y = NULL, 
       x = "# markers identified", 
       fill = NULL) +
  #theme(axis.text.y = element_text(angle = 90, h = 1)) +
  guides(fill=guide_legend(override.aes=list(pattern="none"))) +
  scale_fill_manual(labels = c("young", "all", "old"),
                      values = c("grey30", "grey60", "grey90"))
)
```

```{r f_theme_plot, fig.height = 6}
# this is the same plot, with themes added and different format
# plus small adjustments: 
# -ggpattern (circles instead of hatches for now)
# -x axis on top and expanded all the way to 0
# -fill colors linked to variable names
# -variable factors reversed so that "young" comes before "old"
# -color and name for "all" changed to "all conserved markers"
# -legend generated separately so that "young" comes before "old"
# - y axis also expanded to 0

plot_hsc$group <- factor(
  plot_hsc$group, 
  levels = base::rev(c("young_in_markers", "n_markers", "old_in_markers")))

plot_hsc$celltype <- factor(
  plot_hsc$celltype, 
  levels = base::rev(names(geneset_list_hsc)))

# TODO: change pattern last by adjusting pattern_density_val/pattern_spacing_val
f_theme_plot <- ggplot2::ggplot(
  plot_hsc,
  aes(y = celltype,
      fill = group))+
  ggplot2::geom_col(
    aes(x = count),
    position = "dodge") +
  ggpattern::geom_col_pattern(
    data = plot_hsc,
    aes(x = both_in_markers),
    position = "dodge",
    pattern = "stripe",
    color = NA,
    pattern_angle = 60,
    pattern_fill = "black",
    pattern_colour = "black",
    pattern_density = pattern_density_val,
    pattern_spacing = pattern_spacing_val)+
  theme_all+
  ggplot2::theme(
    axis.title.y = element_blank(),
    axis.text.y = element_text(
      angle = 0,
      vjust = 0.5,
      hjust = 0),
    legend.position = "none",
    plot.title = element_text(
      hjust = 0.5,
      color = col_cons_long["conserved markers"]))+
  ggplot2::labs(
    y = NULL, 
    x = "# markers identified", 
    fill = NULL) +
  ggplot2::ggtitle("Conserved markers")+
  ggplot2::scale_fill_manual(
    values = c("young_in_markers" = "grey30", 
               "n_markers" = col_cons_long[["conserved markers"]], 
               "old_in_markers" = "grey80"))+
  ggplot2::scale_x_continuous(
    position = "top",
    expand  = expansion(c(0,0)))+
  ggplot2::scale_y_discrete(
    expand = c(0, 0))
f_theme_plot
```

```{r f_legend_plot}

# this is the same plot but adjusted for legend generation only
f_legend_plot <- ggplot2::ggplot(
  plot_hsc,
  aes(y = celltype,
      fill = group))+
  ggplot2::geom_col(
    aes(x = count),
    position = "dodge")+
  ggplot2::theme_classic()+
  ggplot2::theme(
    legend.title = element_blank(),
    legend.text = element_text(
      face = legend_text_face,
      size = legend_text_size,
      color = legend_text_color))+
  ggplot2::guides(fill=guide_legend(override.aes=list(pattern="none"))) +
  ggplot2::scale_fill_manual(
    labels = c("young", "pooled", "old"),
    values = c("grey30", 
               col_cons_long[["conserved markers"]], 
               "grey80"))
f_legend_plot

f_theme_legend <- ggpubr::get_legend(f_legend_plot)
```

```{r c_legend, fig.width = 2, fig.height = 2}
f_legend <- ggpubr::ggarrange(f_theme_legend)
f_legend
```


```{r session_info}
utils::sessionInfo()
```
