---
title: "check mapping specific"
author: "Lea WÃ¶lbert"
date: '2023-04-03'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Report QC for all samples separately (Object_ID).
Some operations from the scripts must be repeated to generate relevant plots,
but any new objects cannot be saved from an .Rmd script.

#### Load libraries, source code 

```{r load,  message = FALSE}

library(SingleCellExperiment, quietly = TRUE) 
library(DropletUtils, quietly = TRUE) 
library(tidyverse, quietly = TRUE) 
library(scater, quietly = TRUE) 
library(scuttle, quietly = TRUE) 
library(scran, quietly = TRUE) 
library(pheatmap, quietly = TRUE) 
set.seed(37)
```

```{r load_objects, message = FALSE}

source(file = "../source/plotting.R")

sce_tog <- readRDS(file = snakemake@input[["sce_tog"]])
markergenes_mapping <-  readRDS(file = snakemake@input[["markergenes_mapping"]])
dge_results <-  read.csv(snakemake@input[["dge_results"]], sep = ";")

nr_hvgs <- snakemake@params[["nr_hvgs"]]

spec_curr <- snakemake@wildcards[["species"]]
fraction_curr <- snakemake@wildcards[["fraction"]]

gene_list <- read.csv(snakemake@input[["gene_list"]], 
  header = TRUE, 
  sep = ";", 
  check.names=FALSE, 
  stringsAsFactors=FALSE, 
  as.is=TRUE, 
  colClasses = "character")

markergenes_mapping_temp <- markergenes_mapping
if(nrow(markergenes_mapping) > 200){
  markergenes_mapping_temp <- markergenes_mapping_temp[1:200,]
}

title <- paste("Mapping,", spec_curr, fraction_curr, sep = " ")

#knitr::knit_exit()
```

# Differentily mapped genes

Show the genes that are most differently mapped between genomes.

```{r}
markergenes_mapping
```

## Heatmaps

```{r, fig.height=20, fig.width=7}

sce_tog_temp <- sce_tog[which(rownames(sce_tog) %in% rownames(markergenes_mapping)),]

mat <- as.matrix(assays(sce_tog_temp)$logcounts)

anno_col <- data.frame(
  features = colnames(sce_tog_temp),
  Genome = colData(sce_tog_temp)$Genome)

anno_col <- column_to_rownames(anno_col, var = "features")

pheatmap(
  mat,
  cluster_cols = FALSE,
  annotation_col = anno_col,
  show_colnames = FALSE,
  main = paste0(title, " DMG")
)

```

# All genes from gene list

Show which genes fromt the gene list (for annotation and sub-clustering)
are affected.

```{r}
genes <- unique(gene_list[,1])
print(genes[which(genes %in% rownames(markergenes_mapping))])
```

```{r}

sce_tog_temp <- sce_tog[which(rownames(sce_tog) %in% genes),]

mat <- as.matrix(assays(sce_tog_temp)$logcounts)

anno_col <- data.frame(
  features = colnames(sce_tog_temp),
  Genome = colData(sce_tog_temp)$Genome)

anno_col <- column_to_rownames(anno_col, var = "features")

pheatmap(
  mat,
  cluster_cols = FALSE,
  annotation_col = anno_col,
  show_colnames = FALSE,
  show_rownames = FALSE,
  main = paste0(title, " gene list")
)

```

## UMAPs

Because they are important, check that UMAPs look similar, too.

```{r}

genevar <- modelGeneVar(sce_tog)
hvg <- getTopHVGs(genevar, n=nr_hvgs)

set.seed(37)
sce_tog_temp <- runPCA(sce_tog, ncomponents=25, subset_row = hvg) 
set.seed(37)
sce_tog_temp <- runUMAP(sce_tog_temp, dimred = 'PCA', external_neighbors=TRUE,  subset_row = hvg)

sce_og <- sce_tog_temp[,which(sce_tog_temp$Genome == "OneGenome")]
sce_fg <- sce_tog_temp[,which(sce_tog_temp$Genome == "FourGenomes")]

sce_og
sce_fg

knitr::knit_exit()

gene_exps <- function(gene){

  plot_1 <- umap_gene(sce_og, gene)+
    ggtitle(paste0(title, " DMG, OG ", gene))
  plot_2 <- umap_gene(sce_fg, gene)+
    ggtitle(paste0(title, " DMG, FG ", gene))
  
  return(list(gene, plot_1, plot_2))
}

plotlist <- lapply(genes, gene_exps)
plotlist
```

# Differentially expressed genes

Show which species-specific DEGs are affected and are therefore not "real" DEGs.

```{r}
#dge_results <- read.csv("/omics/odcf/analysis/OE0538_projects/DO-0008/data/main_analysis/sce_objects/13_DGE_analysis_flayer/PC_0.001_FC_1.3/DGE/res_hsc_mcas_specific_df-11.csv", sep = ";")

degs <- vector()
for(i in 1:ncol(dge_results)){
  degs <- unique(c(degs, dge_results[,i][which(dge_results[,i] != "")]))
}

```

```{r}
degs
print(degs[which(degs %in% rownames(markergenes_mapping))])
```

```{r}

sce_tog_temp <- sce_tog[which(rownames(sce_tog) %in% degs),]

mat <- as.matrix(assays(sce_tog_temp)$logcounts)

anno_col <- data.frame(
  features = colnames(sce_tog_temp),
  Genome = colData(sce_tog_temp)$Genome)

anno_col <- column_to_rownames(anno_col, var = "features")

pheatmap(
  mat,
  cluster_cols = FALSE,
  annotation_col = anno_col,
  show_colnames = FALSE,
  show_rownames = FALSE,
  main = paste0(title, " DEGs")
)

```

# Observed "problematic" genes

They may be part of DEGs, but I have consistently observed these genes
to be very different between species.

```{r}

gene_list_problems_obs <- c(
  "Ly6a",
  "Apoe",
  "Cd63",
  "Cd52",
  "Cd48",
  "Car1",
  "Car2",
  "Gm10131",
  "Gm10282",
  "Gm10320",
  "Ctla2a",
  "Gas5",
  "Myl10",
  "AC149090.1",
  "Gm19590",
  "Ifi203",
  "Lgals1",
  "Lst1",
  "Fkbp11",
  "Ftl1-ps1",
  "Gng11",
  "Hp",
  "Sdf211",
  "Cks1b",
  "Nkg7",
  "Emb",
  "Gm15915",
  "Ifngr1",
  "Ramp1",
  "Nrgn",
  "Serpina3g",
  "Ifi27l2a",
  "Serinc3"
)

print(gene_list_problems_obs[which(
  gene_list_problems_obs %in% rownames(markergenes_mapping))])
```

```{r}

sce_tog_temp <- sce_tog[which(rownames(sce_tog) %in% gene_list_problems_obs),]

mat <- as.matrix(assays(sce_tog_temp)$logcounts)

anno_col <- data.frame(
  features = colnames(sce_tog_temp),
  Genome = colData(sce_tog_temp)$Genome)

anno_col <- column_to_rownames(anno_col, var = "features")

pheatmap(
  mat,
  cluster_cols = FALSE,
  annotation_col = anno_col,
  show_colnames = FALSE,
  show_rownames = TRUE,
  main = paste0(title, " observed problems")
)

```
