---
title: "emf_celltype_report"
author: "Lea WÃ¶lbert"
date: '2023-11-07'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### Load objects

```{r seed, message = FALSE}
set.seed(37)
```

```{r load, message = FALSE}
library(tidyverse)
library(scater)
```

```{r source, message = FALSE}
source(file = snakemake@params[["plotting"]])
```

```{r load_objects}
sce <- readRDS(snakemake@input[["sce_input"]])
celltype <- snakemake@wildcards[["celltype"]]

# alter celltypes to fit the wildcards for subsetting
sce$celltypes_alt <- sce$celltypes
sce$celltypes_alt <- gsub(" ", "_", sce$celltypes_alt)
sce$celltypes_alt <- gsub("/", "_", sce$celltypes_alt)
sce$celltypes_alt <- gsub("[.]", "", sce$celltypes_alt)

sce_sep <- sce[,which(sce$celltypes_alt == celltype)]
ct_curr <- sce_sep$celltypes[1]

cons_emf_list <- readRDS(snakemake@input[["cons_EMF_list"]])
cons_emf <- cons_emf_list[[ct_curr]]

sce_ct <- sce_sep
sce <- readRDS(snakemake@input[["sce_input"]])

celltype_res_list_shared <- readRDS(snakemake@input[["celltype_res_list_shared"]])

fraction_curr <- snakemake@wildcards[["fraction"]]
print(ct_curr)
print(names(cons_emf))
```

```{r colors, meassage = FALSE}
colors_path <- snakemake@params[["colors_path"]]
source(snakemake@params[["colors"]])

if(fraction_curr == "hsc"){
  col_cts <- col_cts_hsc
}else if(fraction_curr == "str"){
  col_cts <- col_cts_str
}
```

# Proportions of shared marker genes per species

```{r}
ggdf1 <- as.data.frame(cons_emf$marker_cons_ct)

ggdf1 <- pivot_longer(ggdf1, cols = c(2, 3, 4, 5),
                      names_to = "species", values_to = "foldchange")
ggdf1$species <- factor(ggdf1$species, 
                        levels = c("mmus", "mcas", "mspr", "mcar"))
ggdf1$marker <- vector(length = nrow(ggdf1))
ggdf1$marker[!is.na(ggdf1$foldchange)] <- TRUE

ggdf1_cons <- ggdf1[!is.na(ggdf1$foldchange),]
head(ggdf1_cons)

```

Genes with conserved marker gene function

```{r}
ggplot(ggdf1_cons, aes(fill = cons_marker, x = species))+
  geom_bar(position = "stack")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Genes with conserved marker function", 
                    values = c("TRUE" = "orange", "FALSE" = "grey60"))+
  ggtitle(ct_curr)
  
ggplot(ggdf1_cons, aes(fill = cons_marker, x = species))+
  geom_bar(position = "fill")+
  theme_classic()+
  ylab("Proportion of detected marker genes")+
  scale_fill_manual("Genes with conserved marker function", 
                    values = c("TRUE" = "orange", "FALSE" = "grey60"))+
  ggtitle(ct_curr)
```

Genes that are non-differentially expressed between all four species

```{r}
ggplot(ggdf1_cons, aes(fill = ndge, x = species))+
  geom_bar(position = "stack")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Genes with conserved expression levels", 
                    values = c("TRUE" = "orange", "FALSE" = "grey60"))+
  ggtitle(ct_curr)
  
ggplot(ggdf1_cons, aes(fill = ndge, x = species))+
  geom_bar(position = "fill")+
  theme_classic()+
  ylab("Proportion of detected marker genes")+
  scale_fill_manual("Genes with conserved expression levels", 
                    values = c("TRUE" = "orange", "FALSE" = "grey60"))+
  ggtitle(ct_curr)
```

Genes that are both = core conserved

```{r}
ggplot(ggdf1_cons, aes(fill = cons_emf, x = species))+
  geom_bar(position = "stack")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  ggtitle(ct_curr)+
  scale_fill_manual("Core genes", 
                    values = c("TRUE" = "orange", "FALSE" = "grey60"))

ggplot(ggdf1_cons, aes(fill = cons_emf, x = species))+
  geom_bar(position = "fill")+
  theme_classic()+
  ylab("Proportion of detected marker genes")+
  ggtitle(ct_curr)+
  scale_fill_manual("Core genes", 
                    values = c("TRUE" = "orange", "FALSE" = "grey60"))
```

Proportions per species all together

```{r}
ggdf2 <- ggdf1[ggdf1$marker == TRUE,]

# get overlap of categories as well, otherwise, genes will appear multiple times
ggdf2$marker_only <- vector(length = nrow(ggdf2))
ggdf2$marker_and_ndge <- vector(length = nrow(ggdf2))
ggdf2$conserved_only <- vector(length = nrow(ggdf2))
ggdf2$cons_emferved <- ggdf2$cons_emf

ggdf2$marker_only[which(ggdf2$marker == TRUE &
                          ggdf2$cons_marker == FALSE &
                          ggdf2$cons_emf == FALSE &
                          ggdf2$ndge == FALSE)] <- TRUE

ggdf2$marker_and_ndge[which(ggdf2$marker == TRUE &
                              ggdf2$cons_marker == FALSE &
                              ggdf2$cons_emf == FALSE &
                              ggdf2$ndge == TRUE)] <- TRUE

ggdf2$conserved_only[which(ggdf2$marker == TRUE &
                             ggdf2$cons_marker == TRUE &
                             ggdf2$cons_emf == FALSE &
                             ggdf2$ndge == FALSE)] <- TRUE

ggdf2 <-  as.data.frame(pivot_longer(ggdf2, cols = c(8, 9, 10, 3), 
                                     names_to = "mode", values_to = "is_true"))
ggdf2$mode <- factor(ggdf2$mode, 
                           levels = c("marker_only", "marker_and_ndge",
                                      "conserved_only", "cons_emf"))
ggdf2 <- ggdf2[ggdf2$is_true == TRUE,]
table(ggdf2$species, ggdf2$is_true)
```


```{r}
ggplot(ggdf2, aes(fill = mode, x = species))+
  geom_bar(position = "stack")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Conservation level", 
                    values = c("marker_only" = "grey60",
                               "marker_and_ndge" = "thistle4",
                               "conserved_only" = "darkorange3",
                               "cons_emf" = "orange"))+
  ggtitle(ct_curr)

ggplot(ggdf2, aes(fill = mode, x = species))+
  geom_bar(position = "fill")+
  theme_classic()+
  ylab("Nr of detected marker genes")+
  scale_fill_manual("Conservation level", 
                    values = c("marker_only" = "grey60",
                               "marker_and_ndge" = "thistle4",
                               "conserved_only" = "darkorange3",
                               "cons_emf" = "orange"))+
  ggtitle(ct_curr)

if(ct_curr == "Activated MPPs"){knitr::knit_exit()}
```

# Check the core genes

Do they make sense?

```{r}
print(cons_emf$df_cons)
```

```{r}
print(cons_emf$marker_cons_ct)

print(cons_emf$df_core)
```

Prepare Plots

```{r}

gene_exps <- function(gene){
  
  sce_ct$Species_ID <- factor(sce_ct$Species_ID,
                                   levels = c("mmus", "mcas", "mspr", "mcar"))

  sce$Species_ID <- factor(sce$Species_ID,
                           levels = c("mmus", "mcas", "mspr", "mcar"))
  
  # Gene expression UMAP in cluster X
  plot_1 <- umap_gene(sce_ct, gene)

  # Gene expression UMAP in cluster X, separated by species
  plot_2 <- umap_gene(sce_ct, gene)+
    facet_grid(cols = vars(sce_ct$Species_ID),
               rows = vars(sce_ct$Age_ID))
  
  ggdf <- as.data.frame(colData(sce_ct))
  logc <- assays(sce_ct)[["logcounts"]]

  ggdf$Species_ID <- factor(ggdf$Species_ID,
                            levels = c("mmus", "mcas", "mspr", "mcar"))
  
  plot_3 <- ggplot(ggdf, aes(x = celltypes,
                             y = logc[rownames(logc) == gene,]))+
    ggbeeswarm::geom_quasirandom(size = 0.2, color = "grey60")+
    theme_classic()+
    geom_boxplot(color = "black", alpha = 0)+
    ylab(paste0(gene, " expression"))+
    facet_grid(cols = vars(sce_ct$Species_ID),
               rows = vars(sce_ct$Age_ID))
  
  ggdf_full <- as.data.frame(colData(sce))
  logc_full <- assays(sce)[["logcounts"]]
  
  plot_4 <- umap_gene(sce, gene)
  
  plot_5 <- ggplot(ggdf_full, aes(x = celltypes,
                           y = logc_full[rownames(logc_full) == gene,], 
                           color = celltypes))+
    ggbeeswarm::geom_quasirandom(size = 0.2)+
    theme_classic()+
    scale_color_manual(values = col_cts)+
    geom_boxplot(color = "black", alpha = 0)+
    ylab(paste0(gene, " expression"))+
    theme(axis.text.x = element_text(angle = 90))

  plot_6 <- umap_gene(sce, gene)+
    facet_grid(cols = vars(sce$Species_ID))
  
  return(list(gene, plot_1, plot_2, plot_3, plot_4, plot_5, plot_6))
}

```

```{r}
if(nrow(cons_emf$df_core >= 5)){
  plotlist <- lapply(cons_emf$df_core$gene[1:5], gene_exps)
  print(plotlist)
}
```

# What is the distribution of nDGEs in all genes?

```{r}
shared_ct <- celltype_res_list_shared[[ct_curr]]$all
part_shared_ct <- celltype_res_list_shared[[ct_curr]]$at_least_three
shared_ct[1]
```

```{r}
ggdf <- as.data.frame(rowData(sce_sep))
ggdf$ndge <- rep("none", nrow(ggdf))
ggdf$ndge[ggdf$Symbol %in% part_shared_ct] <- "ndge_three_of_four"
ggdf$ndge[ggdf$Symbol %in% shared_ct] <- "ndge_all"
ggdf$celltype <- rep(ct_curr, nrow(ggdf))

ggplot(ggdf, aes(x = celltype, fill = ndge))+
  geom_bar(position = "stack")+
  theme_classic()+
  scale_fill_manual(values = c("none" = "grey70",
                    "ndge_three_of_four" = "thistle4",
                    "ndge_all" = "thistle"))
```

